{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pybiomart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "server = pybiomart.Server(host='http://www.ensembl.org')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>display_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSEMBL_MART_ENSEMBL</td>\n",
       "      <td>Ensembl Genes 96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSEMBL_MART_MOUSE</td>\n",
       "      <td>Mouse strains 96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSEMBL_MART_SEQUENCE</td>\n",
       "      <td>Sequence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSEMBL_MART_ONTOLOGY</td>\n",
       "      <td>Ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSEMBL_MART_GENOMIC</td>\n",
       "      <td>Genomic features 96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSEMBL_MART_SNP</td>\n",
       "      <td>Ensembl Variation 96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSEMBL_MART_FUNCGEN</td>\n",
       "      <td>Ensembl Regulation 96</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    name           display_name\n",
       "0   ENSEMBL_MART_ENSEMBL       Ensembl Genes 96\n",
       "1     ENSEMBL_MART_MOUSE       Mouse strains 96\n",
       "2  ENSEMBL_MART_SEQUENCE               Sequence\n",
       "3  ENSEMBL_MART_ONTOLOGY               Ontology\n",
       "4   ENSEMBL_MART_GENOMIC    Genomic features 96\n",
       "5       ENSEMBL_MART_SNP   Ensembl Variation 96\n",
       "6   ENSEMBL_MART_FUNCGEN  Ensembl Regulation 96"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "server.list_marts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "mart = server['ENSEMBL_MART_ENSEMBL']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>display_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>choffmanni_gene_ensembl</td>\n",
       "      <td>Sloth genes (choHof1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ohsok_gene_ensembl</td>\n",
       "      <td>Japanese medaka HSOK genes (ASM223469v1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>xtropicalis_gene_ensembl</td>\n",
       "      <td>Xenopus genes (JGI 4.2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rbieti_gene_ensembl</td>\n",
       "      <td>Black snub-nosed monkey genes (ASM169854v1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>aowenii_gene_ensembl</td>\n",
       "      <td>Little spotted kiwi genes (aptOwe1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>dordii_gene_ensembl</td>\n",
       "      <td>Kangaroo rat genes (Dord_2.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ppardus_gene_ensembl</td>\n",
       "      <td>Leopard genes (PanPar1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>mcaroli_gene_ensembl</td>\n",
       "      <td>Ryukyu mouse genes (CAROLI_EIJ_v1.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>mspretus_gene_ensembl</td>\n",
       "      <td>Algerian mouse genes (SPRET_EiJ_v1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ptephrosceles_gene_ensembl</td>\n",
       "      <td>Ugandan red Colobus genes (ASM277652v2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>anancymaae_gene_ensembl</td>\n",
       "      <td>Ma's night monkey genes (Anan_2.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>aplatyrhynchos_gene_ensembl</td>\n",
       "      <td>Duck genes (CAU_duck1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ttruncatus_gene_ensembl</td>\n",
       "      <td>Dolphin genes (turTru1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>nscutatus_gene_ensembl</td>\n",
       "      <td>Mainland tiger snake genes (TS10Xv2-PRI)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>bmutus_gene_ensembl</td>\n",
       "      <td>Wild yak genes (BosGru_v2.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>jjaculus_gene_ensembl</td>\n",
       "      <td>Lesser Egyptian jerboa genes (JacJac1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>oanatinus_gene_ensembl</td>\n",
       "      <td>Platypus genes (OANA5)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>amexicanus_gene_ensembl</td>\n",
       "      <td>Mexican tetra genes (Astyanax_mexicanus-2.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>neugenii_gene_ensembl</td>\n",
       "      <td>Wallaby genes (Meug_1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>mnemestrina_gene_ensembl</td>\n",
       "      <td>Pig-tailed macaque genes (Mnem_1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>smerianae_gene_ensembl</td>\n",
       "      <td>Argentine black and white tegu genes (HLtupMer3)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>oprinceps_gene_ensembl</td>\n",
       "      <td>Pika genes (OchPri2.0-Ens)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>aocellaris_gene_ensembl</td>\n",
       "      <td>Clown anemonefish genes (AmpOce1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>trubripes_gene_ensembl</td>\n",
       "      <td>Fugu genes (FUGU5)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ptroglodytes_gene_ensembl</td>\n",
       "      <td>Chimpanzee genes (Pan_tro_3.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>nvison_gene_ensembl</td>\n",
       "      <td>American mink genes (NNQGG.v01)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>mlucifugus_gene_ensembl</td>\n",
       "      <td>Microbat genes (Myoluc2.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ocuniculus_gene_ensembl</td>\n",
       "      <td>Rabbit genes (OryCun2.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>umaritimus_gene_ensembl</td>\n",
       "      <td>Polar bear genes (UrsMar_1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>kmarmoratus_gene_ensembl</td>\n",
       "      <td>Mangrove rivulus genes (ASM164957v1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>pmajor_gene_ensembl</td>\n",
       "      <td>Great Tit genes (Parus_major1.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>csavignyi_gene_ensembl</td>\n",
       "      <td>C.savignyi genes (CSAV 2.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>pcinereus_gene_ensembl</td>\n",
       "      <td>Koala genes (phaCin_unsw_v4.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>scerevisiae_gene_ensembl</td>\n",
       "      <td>Saccharomyces cerevisiae genes (R64-1-1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>lafricana_gene_ensembl</td>\n",
       "      <td>Elephant genes (Loxafr3.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>loculatus_gene_ensembl</td>\n",
       "      <td>Spotted gar genes (LepOcu1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>hburtoni_gene_ensembl</td>\n",
       "      <td>Burton's mouthbrooder genes (AstBur1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>dnovemcinctus_gene_ensembl</td>\n",
       "      <td>Armadillo genes (Dasnov3.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>ccapucinus_gene_ensembl</td>\n",
       "      <td>Capuchin genes (Cebus_imitator-1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>zalbicollis_gene_ensembl</td>\n",
       "      <td>White-throated sparrow genes (Zonotrichia_albi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>apolyacanthus_gene_ensembl</td>\n",
       "      <td>Spiny chromis genes (ASM210954v1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>mleucophaeus_gene_ensembl</td>\n",
       "      <td>Drill genes (Mleu.le_1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>odegus_gene_ensembl</td>\n",
       "      <td>Degu genes (OctDeg1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>csabaeus_gene_ensembl</td>\n",
       "      <td>Vervet-AGM genes (ChlSab1.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>arowi_gene_ensembl</td>\n",
       "      <td>Okarito brown kiwi genes (aptRow1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>vpacos_gene_ensembl</td>\n",
       "      <td>Alpaca genes (vicPac1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>nbrichardi_gene_ensembl</td>\n",
       "      <td>Lyretail cichlid genes (NeoBri1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>eburgeri_gene_ensembl</td>\n",
       "      <td>Hagfish genes (Eburgeri_3.2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>panubis_gene_ensembl</td>\n",
       "      <td>Olive baboon genes (Panu_3.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>ahaastii_gene_ensembl</td>\n",
       "      <td>Great spotted kiwi genes (aptHaa1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>xcouchianus_gene_ensembl</td>\n",
       "      <td>Monterrey platyfish genes (Xiphophorus_couchia...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>cdingo_gene_ensembl</td>\n",
       "      <td>Dingo genes (ASM325472v1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>munguiculatus_gene_ensembl</td>\n",
       "      <td>Mongolian gerbil genes (MunDraft-v1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>csyrichta_gene_ensembl</td>\n",
       "      <td>Tarsier genes (Tarsius_syrichta-2.0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>mauratus_gene_ensembl</td>\n",
       "      <td>Golden Hamster genes (MesAur1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>elucius_gene_ensembl</td>\n",
       "      <td>Northern pike genes (Eluc_V3)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>mmusculus_gene_ensembl</td>\n",
       "      <td>Mouse genes (GRCm38.p6)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>mzebra_gene_ensembl</td>\n",
       "      <td>Zebra mbuna genes (M_zebra_UMD2a)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>pmarinus_gene_ensembl</td>\n",
       "      <td>Lamprey genes (Pmarinus_7.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>platipinna_gene_ensembl</td>\n",
       "      <td>Sailfin molly genes (P_latipinna-1.0)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>184 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            name  \\\n",
       "0        choffmanni_gene_ensembl   \n",
       "1             ohsok_gene_ensembl   \n",
       "2       xtropicalis_gene_ensembl   \n",
       "3            rbieti_gene_ensembl   \n",
       "4           aowenii_gene_ensembl   \n",
       "5            dordii_gene_ensembl   \n",
       "6           ppardus_gene_ensembl   \n",
       "7           mcaroli_gene_ensembl   \n",
       "8          mspretus_gene_ensembl   \n",
       "9     ptephrosceles_gene_ensembl   \n",
       "10       anancymaae_gene_ensembl   \n",
       "11   aplatyrhynchos_gene_ensembl   \n",
       "12       ttruncatus_gene_ensembl   \n",
       "13        nscutatus_gene_ensembl   \n",
       "14           bmutus_gene_ensembl   \n",
       "15         jjaculus_gene_ensembl   \n",
       "16        oanatinus_gene_ensembl   \n",
       "17       amexicanus_gene_ensembl   \n",
       "18         neugenii_gene_ensembl   \n",
       "19      mnemestrina_gene_ensembl   \n",
       "20        smerianae_gene_ensembl   \n",
       "21        oprinceps_gene_ensembl   \n",
       "22       aocellaris_gene_ensembl   \n",
       "23        trubripes_gene_ensembl   \n",
       "24     ptroglodytes_gene_ensembl   \n",
       "25           nvison_gene_ensembl   \n",
       "26       mlucifugus_gene_ensembl   \n",
       "27       ocuniculus_gene_ensembl   \n",
       "28       umaritimus_gene_ensembl   \n",
       "29      kmarmoratus_gene_ensembl   \n",
       "..                           ...   \n",
       "154          pmajor_gene_ensembl   \n",
       "155       csavignyi_gene_ensembl   \n",
       "156       pcinereus_gene_ensembl   \n",
       "157     scerevisiae_gene_ensembl   \n",
       "158       lafricana_gene_ensembl   \n",
       "159       loculatus_gene_ensembl   \n",
       "160        hburtoni_gene_ensembl   \n",
       "161   dnovemcinctus_gene_ensembl   \n",
       "162      ccapucinus_gene_ensembl   \n",
       "163     zalbicollis_gene_ensembl   \n",
       "164   apolyacanthus_gene_ensembl   \n",
       "165    mleucophaeus_gene_ensembl   \n",
       "166          odegus_gene_ensembl   \n",
       "167        csabaeus_gene_ensembl   \n",
       "168           arowi_gene_ensembl   \n",
       "169          vpacos_gene_ensembl   \n",
       "170      nbrichardi_gene_ensembl   \n",
       "171        eburgeri_gene_ensembl   \n",
       "172         panubis_gene_ensembl   \n",
       "173        ahaastii_gene_ensembl   \n",
       "174     xcouchianus_gene_ensembl   \n",
       "175          cdingo_gene_ensembl   \n",
       "176   munguiculatus_gene_ensembl   \n",
       "177       csyrichta_gene_ensembl   \n",
       "178        mauratus_gene_ensembl   \n",
       "179         elucius_gene_ensembl   \n",
       "180       mmusculus_gene_ensembl   \n",
       "181          mzebra_gene_ensembl   \n",
       "182        pmarinus_gene_ensembl   \n",
       "183      platipinna_gene_ensembl   \n",
       "\n",
       "                                          display_name  \n",
       "0                                Sloth genes (choHof1)  \n",
       "1             Japanese medaka HSOK genes (ASM223469v1)  \n",
       "2                              Xenopus genes (JGI 4.2)  \n",
       "3          Black snub-nosed monkey genes (ASM169854v1)  \n",
       "4                  Little spotted kiwi genes (aptOwe1)  \n",
       "5                        Kangaroo rat genes (Dord_2.0)  \n",
       "6                            Leopard genes (PanPar1.0)  \n",
       "7                 Ryukyu mouse genes (CAROLI_EIJ_v1.1)  \n",
       "8                  Algerian mouse genes (SPRET_EiJ_v1)  \n",
       "9              Ugandan red Colobus genes (ASM277652v2)  \n",
       "10                  Ma's night monkey genes (Anan_2.0)  \n",
       "11                            Duck genes (CAU_duck1.0)  \n",
       "12                             Dolphin genes (turTru1)  \n",
       "13            Mainland tiger snake genes (TS10Xv2-PRI)  \n",
       "14                        Wild yak genes (BosGru_v2.0)  \n",
       "15            Lesser Egyptian jerboa genes (JacJac1.0)  \n",
       "16                              Platypus genes (OANA5)  \n",
       "17        Mexican tetra genes (Astyanax_mexicanus-2.0)  \n",
       "18                            Wallaby genes (Meug_1.0)  \n",
       "19                 Pig-tailed macaque genes (Mnem_1.0)  \n",
       "20    Argentine black and white tegu genes (HLtupMer3)  \n",
       "21                          Pika genes (OchPri2.0-Ens)  \n",
       "22                 Clown anemonefish genes (AmpOce1.0)  \n",
       "23                                  Fugu genes (FUGU5)  \n",
       "24                      Chimpanzee genes (Pan_tro_3.0)  \n",
       "25                     American mink genes (NNQGG.v01)  \n",
       "26                          Microbat genes (Myoluc2.0)  \n",
       "27                            Rabbit genes (OryCun2.0)  \n",
       "28                       Polar bear genes (UrsMar_1.0)  \n",
       "29                Mangrove rivulus genes (ASM164957v1)  \n",
       "..                                                 ...  \n",
       "154                   Great Tit genes (Parus_major1.1)  \n",
       "155                        C.savignyi genes (CSAV 2.0)  \n",
       "156                     Koala genes (phaCin_unsw_v4.1)  \n",
       "157           Saccharomyces cerevisiae genes (R64-1-1)  \n",
       "158                         Elephant genes (Loxafr3.0)  \n",
       "159                        Spotted gar genes (LepOcu1)  \n",
       "160            Burton's mouthbrooder genes (AstBur1.0)  \n",
       "161                        Armadillo genes (Dasnov3.0)  \n",
       "162                Capuchin genes (Cebus_imitator-1.0)  \n",
       "163  White-throated sparrow genes (Zonotrichia_albi...  \n",
       "164                  Spiny chromis genes (ASM210954v1)  \n",
       "165                          Drill genes (Mleu.le_1.0)  \n",
       "166                             Degu genes (OctDeg1.0)  \n",
       "167                       Vervet-AGM genes (ChlSab1.1)  \n",
       "168                 Okarito brown kiwi genes (aptRow1)  \n",
       "169                             Alpaca genes (vicPac1)  \n",
       "170                 Lyretail cichlid genes (NeoBri1.0)  \n",
       "171                       Hagfish genes (Eburgeri_3.2)  \n",
       "172                      Olive baboon genes (Panu_3.0)  \n",
       "173                 Great spotted kiwi genes (aptHaa1)  \n",
       "174  Monterrey platyfish genes (Xiphophorus_couchia...  \n",
       "175                          Dingo genes (ASM325472v1)  \n",
       "176             Mongolian gerbil genes (MunDraft-v1.0)  \n",
       "177             Tarsier genes (Tarsius_syrichta-2.0.1)  \n",
       "178                   Golden Hamster genes (MesAur1.0)  \n",
       "179                      Northern pike genes (Eluc_V3)  \n",
       "180                            Mouse genes (GRCm38.p6)  \n",
       "181                  Zebra mbuna genes (M_zebra_UMD2a)  \n",
       "182                       Lamprey genes (Pmarinus_7.0)  \n",
       "183              Sailfin molly genes (P_latipinna-1.0)  \n",
       "\n",
       "[184 rows x 2 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mart.list_datasets()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = mart['hsapiens_gene_ensembl']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Response [200]>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset.get()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyensembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on method register in module pyensembl.species:\n",
      "\n",
      "register(latin_name, synonyms, reference_assemblies) method of builtins.type instance\n",
      "    Create a Species object from the given arguments and enter into\n",
      "    all the dicts used to look the species up by its fields.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(pyensembl.species.Species.register)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'species': [{'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'African green monkey', 'strain': None, 'assembly': 'ChlSab1.1', 'display_name': 'Vervet-AGM', 'groups': ['rnaseq', 'core', 'variation', 'otherfeatures'], 'aliases': ['chlorocebus sabeus', 'chlorocebus_sabeus', 'chlsab', 'csabaeus', 'agm', 'csab', '60711', 'vervet monkey', 'african green monkey', 'chlorocebus sabaeus', 'vervet-agm'], 'name': 'chlorocebus_sabaeus', 'taxon_id': '60711', 'accession': 'GCA_000409795.2', 'release': 96}, {'strain': None, 'assembly': 'LonStrDom1', 'groups': ['otherfeatures', 'core', 'rnaseq'], 'display_name': 'Bengalese finch', 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Bengalese finch', 'taxon_id': '299123', 'accession': 'GCA_002197715.1', 'release': 96, 'aliases': [], 'name': 'lonchura_striata_domestica'}, {'display_name': 'Drill', 'groups': ['funcgen', 'core', 'otherfeatures'], 'assembly': 'Mleu.le_1.0', 'strain': None, 'strain_collection': None, 'common_name': 'Drill', 'division': 'EnsemblVertebrates', 'release': 96, 'taxon_id': '9568', 'accession': 'GCA_000951045.1', 'name': 'mandrillus_leucophaeus', 'aliases': []}, {'accession': 'GCA_002776525.2', 'taxon_id': '591936', 'release': 96, 'aliases': [], 'name': 'piliocolobus_tephrosceles', 'assembly': 'ASM277652v2', 'strain': None, 'display_name': 'Ugandan red Colobus', 'groups': ['funcgen', 'rnaseq', 'core'], 'division': 'EnsemblVertebrates', 'common_name': 'Ugandan red Colobus', 'strain_collection': None}, {'name': 'cebus_capucinus', 'aliases': [], 'taxon_id': '1737458', 'accession': 'GCA_001604975.1', 'release': 96, 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'White-headed capuchin', 'strain': None, 'assembly': 'Cebus_imitator-1.0', 'groups': ['core', 'funcgen', 'rnaseq', 'otherfeatures'], 'display_name': 'Capuchin'}, {'division': 'EnsemblVertebrates', 'common_name': 'gelada', 'strain_collection': None, 'strain': None, 'assembly': 'Tgel_1.0', 'display_name': 'Gelada', 'groups': ['core', 'rnaseq', 'funcgen'], 'name': 'theropithecus_gelada', 'aliases': [], 'accession': 'GCA_003255815.1', 'taxon_id': '9565', 'release': 96}, {'release': 96, 'accession': 'GCA_003597395.1', 'taxon_id': '106734', 'name': 'chelonoidis_abingdonii', 'aliases': [], 'groups': ['core', 'rnaseq'], 'display_name': 'Abingdon island giant tortoise', 'assembly': 'ASM359739v1', 'strain': None, 'common_name': 'Abingdon island giant tortoise', 'strain_collection': None, 'division': 'EnsemblVertebrates'}, {'groups': ['core', 'rnaseq', 'otherfeatures'], 'display_name': 'Tiger', 'assembly': 'PanTig1.0', 'strain': None, 'strain_collection': None, 'common_name': 'Tiger', 'division': 'EnsemblVertebrates', 'release': 96, 'taxon_id': '74533', 'accession': 'GCA_000464555.1', 'name': 'panthera_tigris_altaica', 'aliases': ['tiger']}, {'division': 'EnsemblVertebrates', 'common_name': 'house mouse', 'strain_collection': 'mouse', 'strain': 'reference (CL57BL6)', 'assembly': 'GRCm38', 'groups': ['core', 'cdna', 'rnaseq', 'funcgen', 'otherfeatures', 'variation'], 'display_name': 'Mouse', 'aliases': ['mmus', 'mus musculus', '10090', 'musmus', 'm_musculus', 'mouse', 'house mouse', 'mus', 'ensmm', 'mmusculus'], 'name': 'mus_musculus', 'accession': 'GCA_000001635.8', 'taxon_id': '10090', 'release': 96}, {'division': 'EnsemblVertebrates', 'common_name': 'shortfin molly', 'strain_collection': None, 'assembly': 'P_mexicana-1.0', 'strain': None, 'display_name': 'Shortfin molly', 'groups': ['otherfeatures', 'rnaseq', 'core'], 'aliases': [], 'name': 'poecilia_mexicana', 'accession': 'GCA_001443325.1', 'taxon_id': '48701', 'release': 96}, {'aliases': ['nleu', 'gibbon', 'nomleu', '61853', 'nleucogenys', 'nomascus leucogenys', 'northern white-cheeked gibbon'], 'name': 'nomascus_leucogenys', 'taxon_id': '61853', 'accession': 'GCA_000146795.3', 'release': 96, 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Northern white-cheeked gibbon', 'assembly': 'Nleu_3.0', 'strain': None, 'groups': ['funcgen', 'core', 'variation', 'otherfeatures'], 'display_name': 'Gibbon'}, {'assembly': 'Sscrofa11.1', 'strain': None, 'display_name': 'Pig', 'groups': ['core', 'funcgen', 'rnaseq', 'otherfeatures', 'variation'], 'division': 'EnsemblVertebrates', 'common_name': 'pig', 'strain_collection': None, 'accession': 'GCA_000003025.6', 'taxon_id': '9823', 'release': 96, 'aliases': ['9823', 'wild boar', 'pigs', 'pig', 'swine', 'sscr', 'sscrofa', 'sus scrofa'], 'name': 'sus_scrofa'}, {'taxon_id': '99837', 'accession': 'GCA_002406435.1', 'release': 96, 'aliases': [], 'name': 'spermophilus_dauricus', 'strain': None, 'assembly': 'ASM240643v1', 'groups': ['core'], 'display_name': 'Daurian ground squirrel', 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Daurian ground squirrel'}, {'division': 'EnsemblVertebrates', 'common_name': 'Domestic sheep', 'strain_collection': None, 'assembly': 'Oar_v3.1', 'strain': None, 'display_name': 'Sheep', 'groups': ['variation', 'otherfeatures', 'rnaseq', 'core'], 'name': 'ovis_aries', 'aliases': ['ovisaries', 'ovis aries', 'oaries', 'oar', 'sheep', '9940', 'oari', 'domestic sheep', 'oviari'], 'accession': 'GCA_000298735.1', 'taxon_id': '9940', 'release': 96}, {'assembly': 'CavAp1.0', 'strain': None, 'display_name': 'Brazilian guinea pig', 'groups': ['core'], 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Brazilian guinea pig', 'taxon_id': '37548', 'accession': 'GCA_000688575.1', 'release': 96, 'name': 'cavia_aperea', 'aliases': ['cavia aperea', 'cape', '37548', 'brazilian guinea pig', 'caperea', 'cavape']}, {'assembly': 'CSAV2.0', 'strain': None, 'display_name': 'C.savignyi', 'groups': ['otherfeatures', 'core'], 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Sea squirt Ciona savignyi', 'taxon_id': '51511', 'accession': None, 'release': 96, 'aliases': ['sea squirt ciona savignyi', '51511', 'csav', 'c.savignyi', 'ciosav', 'ciona savignyi', 'csavignyi'], 'name': 'ciona_savignyi'}, {'assembly': 'DBA_2J_v1', 'strain': 'DBA/2J', 'groups': ['core', 'funcgen'], 'display_name': 'Mouse DBA/2J', 'division': 'EnsemblVertebrates', 'common_name': 'house mouse', 'strain_collection': 'mouse', 'accession': 'GCA_001624505.1', 'taxon_id': '10090', 'release': 96, 'aliases': [], 'name': 'mus_musculus_dba2j'}, {'name': 'pteropus_vampyrus', 'aliases': ['large flying fox', 'pvampyrus', '132908', 'pteropus vampyrus', 'megabat', 'pteropusvampyrus', 'pvam', 'ptevam'], 'taxon_id': '132908', 'accession': None, 'release': 96, 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'large flying fox', 'strain': None, 'assembly': 'pteVam1', 'display_name': 'Megabat', 'groups': ['core']}, {'common_name': \"Coquerel's sifaka\", 'strain_collection': None, 'division': 'EnsemblVertebrates', 'display_name': \"Coquerel's sifaka\", 'groups': ['core', 'funcgen', 'otherfeatures'], 'assembly': 'Pcoq_1.0', 'strain': None, 'aliases': [], 'name': 'propithecus_coquereli', 'release': 96, 'accession': 'GCA_000956105.1', 'taxon_id': '379532'}, {'division': 'EnsemblVertebrates', 'common_name': 'donkey', 'strain_collection': None, 'assembly': 'ASM303372v1', 'strain': None, 'groups': ['core', 'rnaseq'], 'display_name': 'Donkey', 'aliases': ['ass', 'donkey'], 'name': 'equus_asinus_asinus', 'accession': 'GCA_003033725.1', 'taxon_id': '83772', 'release': 96}, {'aliases': ['dmel', 'd.mel', 'd. melanogaster'], 'name': 'drosophila_melanogaster', 'release': 96, 'taxon_id': '7227', 'accession': 'GCA_000001215.4', 'strain_collection': None, 'common_name': 'fruit fly', 'division': 'EnsemblVertebrates', 'display_name': 'Drosophila melanogaster', 'groups': ['funcgen', 'core'], 'strain': None, 'assembly': 'BDGP6.22'}, {'accession': 'GCA_000164805.2', 'taxon_id': '1868482', 'release': 96, 'aliases': ['csyr', 'tarsius_syrichta', 'tarsius syrichta', 'tsyrichta', 'carsyr', 'tarsiussyrichta', 'carlito syrichta', '9478', 'tsyr', 'tarsyr', 'carlitosyrichta', 'carlito', 'tarsier', 'philippine tarsier', 'csyrichta'], 'name': 'carlito_syrichta', 'strain': None, 'assembly': 'Tarsius_syrichta-2.0.1', 'groups': ['core', 'funcgen', 'otherfeatures'], 'display_name': 'Tarsier', 'division': 'EnsemblVertebrates', 'common_name': 'Philippine tarsier', 'strain_collection': None}, {'aliases': ['naked mole-rat female', 'heterocephalus glaber female', 'hglaber_female'], 'name': 'heterocephalus_glaber_female', 'release': 96, 'taxon_id': '10181', 'accession': 'GCA_000247695.1', 'strain_collection': None, 'common_name': 'naked mole-rat', 'division': 'EnsemblVertebrates', 'display_name': 'Naked mole-rat female', 'groups': ['otherfeatures', 'rnaseq', 'core'], 'assembly': 'HetGla_female_1.0', 'strain': None}, {'display_name': 'Midas cichlid', 'groups': ['core'], 'assembly': 'Midas_v5', 'strain': None, 'strain_collection': None, 'common_name': 'Midas cichlid', 'division': 'EnsemblVertebrates', 'release': 96, 'taxon_id': '61819', 'accession': 'GCA_000751415.1', 'aliases': [], 'name': 'amphilophus_citrinellus'}, {'name': 'prolemur_simus', 'aliases': [], 'accession': 'GCA_003258685.1', 'taxon_id': '1328070', 'release': 96, 'division': 'EnsemblVertebrates', 'common_name': 'greater bamboo lemur', 'strain_collection': None, 'strain': None, 'assembly': 'Prosim_1.0', 'display_name': 'Greater bamboo lemur', 'groups': ['funcgen', 'core']}, {'division': 'EnsemblVertebrates', 'common_name': 'naked mole-rat', 'strain_collection': None, 'assembly': 'HetGla_1.0', 'strain': None, 'display_name': 'Naked mole-rat male', 'groups': ['otherfeatures', 'core', 'rnaseq'], 'name': 'heterocephalus_glaber_male', 'aliases': ['hgla', 'heterocephalus glaber male', 'hetgla', 'hglaber_male', 'naked mole-rat male', '10181', 'naked mole-rat'], 'accession': 'GCA_000230445.1', 'taxon_id': '10181', 'release': 96}, {'name': 'manacus_vitellinus', 'aliases': [], 'accession': 'GCA_001715985.2', 'taxon_id': '328815', 'release': 96, 'division': 'EnsemblVertebrates', 'common_name': 'golden-collared manakin', 'strain_collection': None, 'strain': None, 'assembly': 'ASM171598v2', 'groups': ['otherfeatures', 'rnaseq', 'core'], 'display_name': 'Golden-collared manakin'}, {'groups': ['otherfeatures', 'variation', 'cdna', 'core', 'funcgen', 'rnaseq'], 'display_name': 'Human', 'strain': None, 'assembly': 'GRCh38', 'common_name': 'human', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'release': 96, 'accession': 'GCA_000001405.27', 'taxon_id': '9606', 'aliases': ['9606', 'h_sapiens', 'homsap', 'hsapiens', 'homo', 'human', 'homo sapiens', 'enshs', 'hsap'], 'name': 'homo_sapiens'}, {'taxon_id': '9135', 'accession': 'GCA_000534875.1', 'release': 96, 'name': 'serinus_canaria', 'aliases': [], 'assembly': 'SCA1', 'strain': None, 'display_name': 'Common canary', 'groups': ['otherfeatures', 'core', 'rnaseq'], 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'common canary'}, {'groups': ['otherfeatures', 'core', 'rnaseq'], 'display_name': 'Alpine marmot', 'strain': None, 'assembly': 'marMar2.1', 'strain_collection': None, 'common_name': 'Alpine marmot', 'division': 'EnsemblVertebrates', 'release': 96, 'taxon_id': '9994', 'accession': 'GCA_001458135.1', 'aliases': [], 'name': 'marmota_marmota_marmota'}, {'accession': 'GCA_000165445.3', 'taxon_id': '30608', 'release': 96, 'aliases': ['micmur', 'mmur', 'mouse_lemur', 'grey mouse lemur', '30608', 'microcebus murinus', 'mmurinus', 'mouse lemur', 'gray mouse lemur'], 'name': 'microcebus_murinus', 'assembly': 'Mmur_3.0', 'strain': None, 'groups': ['funcgen', 'rnaseq', 'core', 'otherfeatures'], 'display_name': 'Mouse Lemur', 'division': 'EnsemblVertebrates', 'common_name': 'gray mouse lemur', 'strain_collection': None}, {'division': 'EnsemblVertebrates', 'common_name': 'bottlenosed dolphin', 'strain_collection': None, 'assembly': 'turTru1', 'strain': None, 'display_name': 'Dolphin', 'groups': ['core'], 'aliases': ['turtru', 'dolphin', 'ttruncatus', 'tursiops truncatus', 'tursiopstruncatus', '9739', 'bottlenose dolphin', 'bottlenosed dolphin', 'ttru'], 'name': 'tursiops_truncatus', 'accession': None, 'taxon_id': '9739', 'release': 96}, {'strain': None, 'assembly': 'VulVul2.2', 'groups': ['core', 'rnaseq', 'otherfeatures'], 'display_name': 'Red fox', 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'red fox', 'taxon_id': '9627', 'accession': 'GCA_003160815.1', 'release': 96, 'name': 'vulpes_vulpes', 'aliases': ['silver fox', 'fox']}, {'assembly': 'ASM223471v1', 'strain': None, 'display_name': 'Japanese medaka HNI', 'groups': ['core', 'rnaseq'], 'division': 'EnsemblVertebrates', 'common_name': 'Japanese medaka HNI', 'strain_collection': None, 'accession': 'GCA_002234715.1', 'taxon_id': '8090', 'release': 96, 'name': 'oryzias_latipes_hni', 'aliases': []}, {'common_name': 'Spotted gar', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'display_name': 'Spotted gar', 'groups': ['otherfeatures', 'rnaseq', 'core'], 'assembly': 'LepOcu1', 'strain': None, 'name': 'lepisosteus_oculatus', 'aliases': ['locu', '7918', 'lepocu', 'lepisosteus oculatus', 'loculatus', 'spotted gar'], 'release': 96, 'accession': 'GCA_000242695.1', 'taxon_id': '7918'}, {'groups': ['otherfeatures', 'rnaseq', 'core'], 'display_name': 'Tiger tail seahorse', 'assembly': 'H_comes_QL1_v1', 'strain': None, 'common_name': 'tiger tail seahorse', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'release': 96, 'accession': 'GCA_001891065.1', 'taxon_id': '109280', 'name': 'hippocampus_comes', 'aliases': []}, {'common_name': 'Northern American deer mouse', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'display_name': 'Northern American deer mouse', 'groups': ['core'], 'strain': None, 'assembly': 'HU_Pman_2.1', 'aliases': ['pman', 'northern american deer mouse', 'peromyscus maniculatus bairdii', 'perman', '230844', 'pmaniculatus_bairdii'], 'name': 'peromyscus_maniculatus_bairdii', 'release': 96, 'accession': 'GCA_003704035.1', 'taxon_id': '230844'}, {'aliases': ['mus musculus castaneus casteij', '10091', 'mus_musculus_castaneus_casteij', 'south eastern house mouse', 'mmusculus_castaneus_casteij', 'mouse casteij'], 'name': 'mus_musculus_casteij', 'release': 96, 'accession': 'GCA_001624445.1', 'taxon_id': '10091', 'common_name': 'south eastern house mouse', 'strain_collection': 'mouse', 'division': 'EnsemblVertebrates', 'groups': ['core', 'funcgen'], 'display_name': 'Mouse CAST/EiJ', 'strain': 'CAST/EiJ', 'assembly': 'CAST_EiJ_v1'}, {'release': 96, 'accession': None, 'taxon_id': '59729', 'aliases': ['finch', 'taegut', 'tguttata', 'zebra finch', 'zebrafinch', 'taeniopygia guttata', '59729', 'tgut', 'taeniopygiaguttata', 'taeniopygia'], 'name': 'taeniopygia_guttata', 'groups': ['core', 'otherfeatures', 'variation'], 'display_name': 'Zebra Finch', 'assembly': 'taeGut3.2.4', 'strain': None, 'common_name': 'zebra finch', 'strain_collection': None, 'division': 'EnsemblVertebrates'}, {'name': 'capra_hircus', 'aliases': [], 'release': 96, 'accession': 'GCA_001704415.1', 'taxon_id': '9925', 'common_name': 'Goat', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'groups': ['core', 'funcgen', 'rnaseq', 'otherfeatures', 'variation'], 'display_name': 'Goat', 'assembly': 'ARS1', 'strain': None}, {'name': 'colobus_angolensis_palliatus', 'aliases': [], 'release': 96, 'accession': 'GCA_000951035.1', 'taxon_id': '336983', 'common_name': 'Angola colobus', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'display_name': 'Angola colobus', 'groups': ['otherfeatures', 'funcgen', 'core'], 'strain': None, 'assembly': 'Cang.pa_1.0'}, {'groups': ['otherfeatures', 'rnaseq', 'core'], 'display_name': 'Blue-crowned manakin', 'strain': None, 'assembly': 'Lepidothrix_coronata-1.0', 'strain_collection': None, 'common_name': 'blue-crowned manakin', 'division': 'EnsemblVertebrates', 'release': 96, 'taxon_id': '321398', 'accession': 'GCA_001604755.1', 'name': 'lepidothrix_coronata', 'aliases': []}, {'common_name': 'vole', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'display_name': 'Prairie vole', 'groups': ['otherfeatures', 'core'], 'strain': None, 'assembly': 'MicOch1.0', 'name': 'microtus_ochrogaster', 'aliases': ['vole', '79684', 'moch', 'microtus ochrogaster', 'mochrogaster', 'micoch', 'm_ochrogaster', 'pvole', 'prairie vole'], 'release': 96, 'accession': 'GCA_000317375.1', 'taxon_id': '79684'}, {'strain': None, 'assembly': 'gorGor4', 'groups': ['otherfeatures', 'core', 'funcgen', 'rnaseq'], 'display_name': 'Gorilla', 'division': 'EnsemblVertebrates', 'common_name': 'Western Lowland Gorilla', 'strain_collection': None, 'accession': 'GCA_000151905.3', 'taxon_id': '9595', 'release': 96, 'aliases': ['9595', '9593', 'western gorilla', 'gorilla_gorilla_gorilla', 'gorilla gorilla gorilla', 'gorilla gorilla', 'ggorilla', 'gorilla', 'ggor', 'gorgor'], 'name': 'gorilla_gorilla'}, {'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Lesser Egyptian jerboa', 'strain': None, 'assembly': 'JacJac1.0', 'display_name': 'Lesser Egyptian jerboa', 'groups': ['core', 'otherfeatures'], 'aliases': ['lesser egyptian jerboa', 'jjaculus', 'jaculus jaculus', 'jacjac', '51337', 'jjac'], 'name': 'jaculus_jaculus', 'taxon_id': '51337', 'accession': 'GCA_000280705.1', 'release': 96}, {'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'tuatara', 'assembly': 'ASM311381v1', 'strain': None, 'groups': ['rnaseq', 'core', 'otherfeatures'], 'display_name': 'Tuatara', 'name': 'sphenodon_punctatus', 'aliases': [], 'taxon_id': '8508', 'accession': 'GCA_003113815.1', 'release': 96}, {'assembly': 'Eburgeri_3.2', 'strain': None, 'display_name': 'Hagfish', 'groups': ['rnaseq', 'core'], 'division': 'EnsemblVertebrates', 'common_name': 'Inshore hagfish', 'strain_collection': None, 'accession': 'GCA_900186335.2', 'taxon_id': '7764', 'release': 96, 'aliases': ['hagfish'], 'name': 'eptatretus_burgeri'}, {'release': 96, 'taxon_id': '8840', 'accession': 'GCA_002743455.1', 'name': 'anas_platyrhynchos_platyrhynchos', 'aliases': [], 'display_name': 'Duck', 'groups': ['core', 'funcgen', 'rnaseq'], 'assembly': 'CAU_duck1.0', 'strain': None, 'strain_collection': None, 'common_name': 'common mallard', 'division': 'EnsemblVertebrates'}, {'division': 'EnsemblVertebrates', 'strain_collection': 'mouse', 'common_name': 'house mouse', 'strain': 'FVB/NJ', 'assembly': 'FVB_NJ_v1', 'display_name': 'Mouse FVB/NJ', 'groups': ['core', 'funcgen'], 'aliases': [], 'name': 'mus_musculus_fvbnj', 'taxon_id': '10090', 'accession': 'GCA_001624535.1', 'release': 96}, {'assembly': '129S1_SvImJ_v1', 'strain': '129S1/SvImJ', 'display_name': 'Mouse 129S1/SvImJ', 'groups': ['funcgen', 'core'], 'division': 'EnsemblVertebrates', 'strain_collection': 'mouse', 'common_name': 'house mouse', 'taxon_id': '10090', 'accession': 'GCA_001624185.1', 'release': 96, 'aliases': [], 'name': 'mus_musculus_129s1svimj'}, {'strain': None, 'assembly': 'proCap1', 'display_name': 'Hyrax', 'groups': ['core'], 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'cape rock hyrax', 'taxon_id': '9813', 'accession': None, 'release': 96, 'aliases': ['pcap', 'hyrax', 'cape rock hyrax', '9813', 'pcapensis', 'procaviacapensis', 'rock_hyrax', 'procap', 'procavia capensis'], 'name': 'procavia_capensis'}, {'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'green anole', 'strain': None, 'assembly': 'AnoCar2.0', 'display_name': 'Anole lizard', 'groups': ['rnaseq', 'core', 'otherfeatures'], 'aliases': ['anoliscarolinensis', 'acarolinensis', 'green anole', 'acar', '28377', 'anolis_lizard', 'anolis', 'green anole lizard', 'anolis carolinensis', 'anocar', 'anole lizard'], 'name': 'anolis_carolinensis', 'taxon_id': '28377', 'accession': 'GCA_000090745.1', 'release': 96}, {'release': 96, 'accession': 'GCA_003344425.1', 'taxon_id': '9643', 'name': 'ursus_americanus', 'aliases': [], 'groups': ['rnaseq', 'core'], 'display_name': 'American black bear', 'strain': None, 'assembly': 'ASM334442v1', 'common_name': 'American black bear', 'strain_collection': None, 'division': 'EnsemblVertebrates'}, {'strain_collection': None, 'common_name': 'white-tufted-ear marmoset', 'division': 'EnsemblVertebrates', 'groups': ['rnaseq', 'funcgen', 'core'], 'display_name': 'Marmoset', 'assembly': 'ASM275486v1', 'strain': None, 'aliases': ['cjac', '9483', 'cjacchus', 'callithrix jacchus', 'white-tufted-ear marmoset', 'caljac', 'cj321', 'marmoset', 'cjacchus321'], 'name': 'callithrix_jacchus', 'release': 96, 'taxon_id': '9483', 'accession': 'GCA_002754865.1'}, {'division': 'EnsemblVertebrates', 'common_name': 'Western painted turtle', 'strain_collection': None, 'assembly': 'Chrysemys_picta_bellii-3.0.3', 'strain': None, 'display_name': 'Painted turtle', 'groups': ['rnaseq', 'core', 'otherfeatures'], 'name': 'chrysemys_picta_bellii', 'aliases': ['western painted turtle', 'painted turtle'], 'accession': 'GCA_000241765.2', 'taxon_id': '8478', 'release': 96}, {'aliases': ['arma', 'dnovemcinctus', 'dasypus novemcinctu', 'nine-banded armadillo', 'armadillo', '9361', 'dasypus novemcinctus', 'dasnov', 'dnov'], 'name': 'dasypus_novemcinctus', 'taxon_id': '9361', 'accession': 'GCA_000208655.2', 'release': 96, 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'nine-banded armadillo', 'strain': None, 'assembly': 'Dasnov3.0', 'groups': ['rnaseq', 'core', 'otherfeatures'], 'display_name': 'Armadillo'}, {'common_name': 'small Madagascar hedgehog', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'display_name': 'Lesser hedgehog tenrec', 'groups': ['core'], 'strain': None, 'assembly': 'TENREC', 'aliases': ['echinops telfairi', '9371', 'etelfairi', 'lesser_hedgehog', 'madagascar_hedgehog', 'small madagascar hedgehog', 'lesser hedgehog tenrec', 'etel', 'echtel', 'tenrec'], 'name': 'echinops_telfairi', 'release': 96, 'accession': None, 'taxon_id': '9371'}, {'release': 96, 'accession': 'GCA_001624745.1', 'taxon_id': '10090', 'aliases': [], 'name': 'mus_musculus_nzohlltj', 'display_name': 'Mouse NZO/HlLtJ', 'groups': ['funcgen', 'core'], 'strain': 'NZO/HlLtJ', 'assembly': 'NZO_HlLtJ_v1', 'common_name': 'house mouse', 'strain_collection': 'mouse', 'division': 'EnsemblVertebrates'}, {'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'rhesus monkey', 'strain': None, 'assembly': 'Mmul_8.0.1', 'display_name': 'Macaque', 'groups': ['variation', 'otherfeatures', 'rnaseq', 'funcgen', 'core'], 'name': 'macaca_mulatta', 'aliases': ['macaca mulatta', 'macaque', '9544', 'mmulatta', 'rhesus', 'rhesus_monkey', 'mmul', 'macmul', 'rhesus monkey'], 'taxon_id': '9544', 'accession': 'GCA_000772875.3', 'release': 96}, {'name': 'mus_musculus_c3hhej', 'aliases': [], 'accession': 'GCA_001632575.1', 'taxon_id': '10090', 'release': 96, 'division': 'EnsemblVertebrates', 'common_name': 'house mouse', 'strain_collection': 'mouse', 'assembly': 'C3H_HeJ_v1', 'strain': 'C3H/HeJ', 'display_name': 'Mouse C3H/HeJ', 'groups': ['core', 'funcgen']}, {'assembly': 'OtoGar3', 'strain': None, 'display_name': 'Bushbaby', 'groups': ['otherfeatures', 'core'], 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'small-eared galago', 'taxon_id': '30611', 'accession': 'GCA_000181295.3', 'release': 96, 'name': 'otolemur_garnettii', 'aliases': ['galago', 'ogar', 'bushbaby', 'otogar', 'ogarnettii', '30611', 'small-eared galago', 'otolemur garnettii']}, {'division': 'EnsemblVertebrates', 'common_name': 'koala', 'strain_collection': None, 'strain': None, 'assembly': 'phaCin_unsw_v4.1', 'display_name': 'Koala', 'groups': ['core'], 'aliases': [], 'name': 'phascolarctos_cinereus', 'accession': 'GCA_002099425.1', 'taxon_id': '38626', 'release': 96}, {'display_name': 'Alpaca', 'groups': ['core'], 'strain': None, 'assembly': 'vicPac1', 'strain_collection': None, 'common_name': 'alpaca', 'division': 'EnsemblVertebrates', 'release': 96, 'taxon_id': '30538', 'accession': None, 'name': 'vicugna_pacos', 'aliases': ['vpacos', 'vicpac', 'vicugnapacos', 'vicugna pacos', 'vpac', '30538', 'alpaca']}, {'strain': None, 'assembly': 'ASM169857v1', 'display_name': 'Ocean sunfish', 'groups': ['core'], 'division': 'EnsemblVertebrates', 'common_name': 'ocean sunfish', 'strain_collection': None, 'accession': 'GCA_001698575.1', 'taxon_id': '94237', 'release': 96, 'aliases': [], 'name': 'mola_mola'}, {'assembly': 'MUSP714', 'strain': None, 'groups': ['rnaseq', 'core'], 'display_name': 'Steppe mouse', 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'steppe mouse', 'taxon_id': '10103', 'accession': 'GCA_003336285.1', 'release': 96, 'aliases': [], 'name': 'mus_spicilegus'}, {'common_name': 'Ryukyu mouse', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'display_name': 'Ryukyu mouse', 'groups': ['core'], 'assembly': 'CAROLI_EIJ_v1.1', 'strain': 'CAROLI_EIJ', 'name': 'mus_caroli', 'aliases': ['muscar', 'ryukyu mouse', 'mcar', 'mus caroli', '10089', 'mcaroli'], 'release': 96, 'accession': 'GCA_900094665.2', 'taxon_id': '10089'}, {'release': 96, 'accession': 'GCA_000224145.1', 'taxon_id': '7719', 'aliases': ['cint', 'ciona', 'cintestinalis', 'ciona intestinalis', 'cioint', '7719', 'sea squirt ciona intestinalis', 'c.intestinalis'], 'name': 'ciona_intestinalis', 'groups': ['core', 'funcgen', 'otherfeatures'], 'display_name': 'C.intestinalis', 'assembly': 'KH', 'strain': None, 'common_name': 'Sea squirt Ciona intestinalis', 'strain_collection': None, 'division': 'EnsemblVertebrates'}, {'name': 'dromaius_novaehollandiae', 'aliases': [], 'release': 96, 'taxon_id': '8790', 'accession': 'GCA_003342905.1', 'strain_collection': None, 'common_name': 'emu', 'division': 'EnsemblVertebrates', 'display_name': 'Emu', 'groups': ['otherfeatures', 'core', 'rnaseq'], 'strain': None, 'assembly': 'droNov1'}, {'common_name': \"Ma's night monkey\", 'strain_collection': None, 'division': 'EnsemblVertebrates', 'display_name': \"Ma's night monkey\", 'groups': ['core', 'rnaseq', 'funcgen', 'otherfeatures'], 'assembly': 'Anan_2.0', 'strain': None, 'name': 'aotus_nancymaae', 'aliases': [], 'release': 96, 'accession': 'GCA_000952055.2', 'taxon_id': '37293'}, {'common_name': 'red-bellied piranha', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'groups': ['core', 'rnaseq', 'otherfeatures'], 'display_name': 'Red-bellied piranha', 'strain': None, 'assembly': 'Pygocentrus_nattereri-1.0.2', 'aliases': [], 'name': 'pygocentrus_nattereri', 'release': 96, 'accession': 'GCA_001682695.1', 'taxon_id': '42514'}, {'accession': 'GCA_000238955.5', 'taxon_id': '106582', 'release': 96, 'aliases': [], 'name': 'maylandia_zebra', 'assembly': 'M_zebra_UMD2a', 'strain': None, 'display_name': 'Zebra mbuna', 'groups': ['rnaseq', 'core', 'otherfeatures'], 'division': 'EnsemblVertebrates', 'common_name': 'zebra mbuna', 'strain_collection': None}, {'aliases': ['broado5', 'mdom', 'gray short-tailed opossum', '13616', 'monodelphis domestica', 'opossum', 'mondom', 'mdomestica'], 'name': 'monodelphis_domestica', 'taxon_id': '13616', 'accession': 'GCF_000002295.2', 'release': 96, 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'gray short-tailed opossum', 'strain': None, 'assembly': 'monDom5', 'display_name': 'Opossum', 'groups': ['variation', 'rnaseq', 'core']}, {'name': 'oryzias_melastigma', 'aliases': [], 'taxon_id': '30732', 'accession': 'GCA_002922805.1', 'release': 96, 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Indian medaka', 'strain': None, 'assembly': 'Om_v0.7.RACA', 'groups': ['core', 'rnaseq'], 'display_name': 'Indian medaka'}, {'aliases': [], 'name': 'stegastes_partitus', 'release': 96, 'accession': 'GCA_000690725.1', 'taxon_id': '144197', 'common_name': 'bicolor damselfish', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'groups': ['core', 'rnaseq', 'otherfeatures'], 'display_name': 'Bicolor damselfish', 'strain': None, 'assembly': 'Stegastes_partitus-1.0.2'}, {'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'pink-footed goose', 'assembly': 'ASM259213v1', 'strain': None, 'display_name': 'Pink-footed goose', 'groups': ['core'], 'name': 'anser_brachyrhynchus', 'aliases': [], 'taxon_id': '132585', 'accession': 'GCA_002592135.1', 'release': 96}, {'name': 'myotis_lucifugus', 'aliases': ['mluc', 'mlucifugus', 'little_brown_bat', 'microbat', 'little brown bat', '59463', 'myotis lucifugus', 'myoluc'], 'release': 96, 'accession': 'GCA_000147115.1', 'taxon_id': '59463', 'common_name': 'little brown bat', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'display_name': 'Microbat', 'groups': ['core', 'otherfeatures'], 'strain': None, 'assembly': 'Myoluc2.0'}, {'strain': None, 'assembly': 'Felis_catus_9.0', 'groups': ['variation', 'otherfeatures', 'rnaseq', 'funcgen', 'core'], 'display_name': 'Cat', 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'domestic cat', 'taxon_id': '9685', 'accession': 'GCA_000181335.4', 'release': 96, 'aliases': ['9685', 'felis catus', 'felcat', 'domestic cat', 'cat', 'fcat', 'fcatus'], 'name': 'felis_catus'}, {'name': 'junco_hyemalis', 'aliases': [], 'accession': 'GCA_003829775.1', 'taxon_id': '40217', 'release': 96, 'division': 'EnsemblVertebrates', 'common_name': 'dark-eyed junco', 'strain_collection': None, 'assembly': 'ASM382977v1', 'strain': None, 'groups': ['core', 'rnaseq'], 'display_name': 'Dark-eyed junco'}, {'name': 'mus_musculus_aj', 'aliases': [], 'accession': 'GCA_001624215.1', 'taxon_id': '10090', 'release': 96, 'division': 'EnsemblVertebrates', 'common_name': 'house mouse', 'strain_collection': 'mouse', 'strain': 'A/J', 'assembly': 'A_J_v1', 'display_name': 'Mouse A/J', 'groups': ['core', 'funcgen']}, {'aliases': [], 'name': 'cricetulus_griseus_picr', 'release': 96, 'taxon_id': '10029', 'accession': 'GCA_003668045.1', 'strain_collection': None, 'common_name': 'Chinese hamster', 'division': 'EnsemblVertebrates', 'display_name': 'Chinese hamster PICR', 'groups': ['core', 'rnaseq', 'funcgen'], 'assembly': 'CriGri-PICR', 'strain': None}, {'release': 96, 'taxon_id': '9555', 'accession': 'GCA_000264685.2', 'aliases': ['papio cynocephalus anubis', '9555', 'olive baboon', 'panu', 'papio hamadryas doguera', 'papio anubis', 'kenya baboon', 'panubis', 'papio doguera', 'papanu', 'anubis baboon', 'doguera baboon', 'papio hamadryas anubis'], 'name': 'papio_anubis', 'groups': ['otherfeatures', 'core', 'rnaseq', 'funcgen'], 'display_name': 'Olive baboon', 'assembly': 'Panu_3.0', 'strain': None, 'strain_collection': None, 'common_name': 'Olive baboon', 'division': 'EnsemblVertebrates'}, {'name': 'rhinopithecus_bieti', 'aliases': [], 'release': 96, 'taxon_id': '61621', 'accession': 'GCA_001698545.1', 'strain_collection': None, 'common_name': 'Black snub-nosed monkey', 'division': 'EnsemblVertebrates', 'groups': ['funcgen', 'rnaseq', 'core', 'otherfeatures'], 'display_name': 'Black snub-nosed monkey', 'assembly': 'ASM169854v1', 'strain': None}, {'taxon_id': '39442', 'accession': 'GCA_001624775.1', 'release': 96, 'aliases': ['mmusculus_musculus_pwkphj', '39442', 'eastern european house mouse', 'mus_musculus_musculus_pwkphj', 'mus musculus musculus pwkphj', 'mouse pwkphj'], 'name': 'mus_musculus_pwkphj', 'assembly': 'PWK_PhJ_v1', 'strain': 'PWK/PhJ', 'display_name': 'Mouse PWK/PhJ', 'groups': ['funcgen', 'core'], 'division': 'EnsemblVertebrates', 'strain_collection': 'mouse', 'common_name': 'eastern european house mouse'}, {'display_name': 'Bolivian squirrel monkey', 'groups': ['otherfeatures', 'core', 'rnaseq', 'funcgen'], 'assembly': 'SaiBol1.0', 'strain': None, 'common_name': 'Bolivian squirrel monkey', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'release': 96, 'accession': 'GCA_000235385.1', 'taxon_id': '39432', 'aliases': [], 'name': 'saimiri_boliviensis_boliviensis'}, {'aliases': [], 'name': 'salvator_merianae', 'release': 96, 'accession': 'GCA_003586115.1', 'taxon_id': '96440', 'common_name': 'Argentine black and white tegu', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'display_name': 'Argentine black and white tegu', 'groups': ['core', 'rnaseq'], 'strain': None, 'assembly': 'HLtupMer3'}, {'strain': None, 'assembly': 'BallGen_V1', 'groups': ['otherfeatures', 'core', 'rnaseq'], 'display_name': 'Ballan wrasse', 'division': 'EnsemblVertebrates', 'common_name': 'ballan wrasse', 'strain_collection': None, 'accession': 'GCA_900080235.1', 'taxon_id': '56723', 'release': 96, 'name': 'labrus_bergylta', 'aliases': []}, {'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Tasmanian devil', 'strain': None, 'assembly': 'DEVIL7.0', 'groups': ['core', 'otherfeatures'], 'display_name': 'Tasmanian devil', 'name': 'sarcophilus_harrisii', 'aliases': ['sharrisii', 'sarhar', 'devil', 'tasmanian_devil', 'shar', 'sarcophilus harrisii', 'tasmanian devil', '9305', 'tdevil'], 'taxon_id': '9305', 'accession': 'GCA_000189315.1', 'release': 96}, {'release': 96, 'taxon_id': '10141', 'accession': 'GCA_000151735.1', 'aliases': ['guinea pig', 'cporcellus', 'guinea_pig', 'cavpor', '10141', 'cpor', 'domestic guinea pig', 'cavia porcellus'], 'name': 'cavia_porcellus', 'groups': ['rnaseq', 'funcgen', 'core', 'otherfeatures'], 'display_name': 'Guinea Pig', 'assembly': 'Cavpor3.0', 'strain': None, 'strain_collection': None, 'common_name': 'domestic guinea pig', 'division': 'EnsemblVertebrates'}, {'name': 'crocodylus_porosus', 'aliases': [], 'taxon_id': '8502', 'accession': 'GCA_001723895.1', 'release': 96, 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Australian saltwater crocodile', 'strain': None, 'assembly': 'CroPor_comp1', 'groups': ['rnaseq', 'core', 'otherfeatures'], 'display_name': 'Australian saltwater crocodile'}, {'release': 96, 'taxon_id': '10090', 'accession': 'GCA_001632555.1', 'aliases': [], 'name': 'mus_musculus_c57bl6nj', 'display_name': 'Mouse C57BL/6NJ', 'groups': ['core', 'funcgen'], 'assembly': 'C57BL_6NJ_v1', 'strain': 'C57BL/6NJ', 'strain_collection': 'mouse', 'common_name': 'house mouse', 'division': 'EnsemblVertebrates'}, {'release': 96, 'accession': 'GCA_001444195.1', 'taxon_id': '32473', 'name': 'xiphophorus_couchianus', 'aliases': [], 'groups': ['core'], 'display_name': 'Monterrey platyfish', 'assembly': 'Xiphophorus_couchianus-4.0.1', 'strain': None, 'common_name': 'Monterrey platyfish', 'strain_collection': None, 'division': 'EnsemblVertebrates'}, {'strain_collection': None, 'common_name': 'American pika', 'division': 'EnsemblVertebrates', 'display_name': 'Pika', 'groups': ['core'], 'assembly': 'OchPri2.0-Ens', 'strain': None, 'aliases': ['ochpri', 'pika', 'oprinceps', 'opri', '9978', 'american pika', 'ochotona princeps'], 'name': 'ochotona_princeps', 'release': 96, 'taxon_id': '9978', 'accession': None}, {'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Chinese softshell turtle', 'strain': None, 'assembly': 'PelSin_1.0', 'display_name': 'Chinese softshell turtle', 'groups': ['otherfeatures', 'rnaseq', 'core'], 'name': 'pelodiscus_sinensis', 'aliases': ['pelodiscus sinensis', '13735', 'chinese softshell turtle', 'psinensis', 'p_sinensis', 'chinese soft shell turtle', 'softshell turtle'], 'taxon_id': '13735', 'accession': 'GCA_000230535.1', 'release': 96}, {'name': 'zonotrichia_albicollis', 'aliases': [], 'release': 96, 'taxon_id': '44394', 'accession': 'GCA_000385455.1', 'strain_collection': None, 'common_name': 'white-throated sparrow', 'division': 'EnsemblVertebrates', 'display_name': 'White-throated sparrow', 'groups': ['otherfeatures', 'core', 'rnaseq'], 'assembly': 'Zonotrichia_albicollis-1.0.1', 'strain': None}, {'name': 'mustela_putorius_furo', 'aliases': ['ferret', 'domestic ferret', 'mputorius_furo', 'mustela putorius furo', 'musput', '9669', 'mput'], 'taxon_id': '9669', 'accession': 'GCA_000215625.1', 'release': 96, 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Domestic ferret', 'assembly': 'MusPutFur1.0', 'strain': None, 'display_name': 'Ferret', 'groups': ['rnaseq', 'core', 'otherfeatures']}, {'strain': None, 'assembly': 'ASM342692v1', 'groups': ['core', 'rnaseq'], 'display_name': 'Arctic ground squirrel', 'division': 'EnsemblVertebrates', 'common_name': 'Arctic ground squirrel', 'strain_collection': None, 'accession': 'GCA_003426925.1', 'taxon_id': '9999', 'release': 96, 'name': 'urocitellus_parryii', 'aliases': []}, {'taxon_id': '7897', 'accession': 'GCA_000225785.1', 'release': 96, 'name': 'latimeria_chalumnae', 'aliases': ['coelacanth', 'lcha', 'lchalumnae', 'living fossil', 'latimeria chalumnae', 'lamineria', 'latcha', '7897'], 'strain': None, 'assembly': 'LatCha1', 'display_name': 'Coelacanth', 'groups': ['core', 'otherfeatures'], 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'coelacanth'}, {'taxon_id': '43179', 'accession': 'GCA_000236235.1', 'release': 96, 'name': 'ictidomys_tridecemlineatus', 'aliases': ['stridecemlineatus', '43179', 'thirteen-lined ground squirrel', 'spetri', 'ictidomys tridecemlineatus', 'stri', 'squirrel', 'spermophilus_tridecemlineatus'], 'strain': None, 'assembly': 'SpeTri2.0', 'groups': ['otherfeatures', 'core', 'rnaseq'], 'display_name': 'Squirrel', 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'thirteen-lined ground squirrel'}, {'display_name': 'Western mosquitofish', 'groups': ['core', 'rnaseq'], 'assembly': 'ASM309773v1', 'strain': None, 'strain_collection': None, 'common_name': 'western mosquitofish', 'division': 'EnsemblVertebrates', 'release': 96, 'taxon_id': '33528', 'accession': 'GCA_003097735.1', 'name': 'gambusia_affinis', 'aliases': []}, {'strain_collection': None, 'common_name': 'blue tit', 'division': 'EnsemblVertebrates', 'groups': ['rnaseq', 'core', 'otherfeatures'], 'display_name': 'Blue tit', 'strain': None, 'assembly': 'cyaCae2', 'name': 'cyanistes_caeruleus', 'aliases': [], 'release': 96, 'taxon_id': '156563', 'accession': 'GCA_002901205.1'}, {'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Golden snub-nosed monkey', 'strain': None, 'assembly': 'Rrox_v1', 'groups': ['funcgen', 'rnaseq', 'core', 'otherfeatures'], 'display_name': 'Golden snub-nosed monkey', 'aliases': [], 'name': 'rhinopithecus_roxellana', 'taxon_id': '61622', 'accession': 'GCA_000769185.1', 'release': 96}, {'release': 96, 'taxon_id': '1026970', 'accession': 'GCA_000622305.1', 'name': 'nannospalax_galili', 'aliases': ['nannospalax galili', 'upper galilee mountains blind mole rat', '1026970', 'ngal', 'nangal', 'ngalili'], 'groups': ['core', 'funcgen', 'rnaseq', 'otherfeatures'], 'display_name': 'Upper Galilee mountains blind mole rat', 'assembly': 'S.galili_v1.0', 'strain': None, 'strain_collection': None, 'common_name': 'Upper Galilee mountains blind mole rat', 'division': 'EnsemblVertebrates'}, {'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'southern platyfish', 'assembly': 'X_maculatus-5.0-male', 'strain': None, 'display_name': 'Platyfish', 'groups': ['otherfeatures', 'rnaseq', 'core'], 'aliases': ['xipmac', 'xmac', 'southern platyfish', 'xiphophorus maculatus', 'xmaculatus', 'platyfish', '8083'], 'name': 'xiphophorus_maculatus', 'taxon_id': '8083', 'accession': 'GCA_002775205.2', 'release': 96}, {'groups': ['variation', 'otherfeatures', 'funcgen', 'core'], 'display_name': 'Saccharomyces cerevisiae', 'strain': 'S288C', 'assembly': 'R64-1-1', 'strain_collection': None, 'common_name': \"baker's yeast\", 'division': 'EnsemblVertebrates', 'release': 96, 'taxon_id': '4932', 'accession': 'GCA_000146045.2', 'aliases': ['saccer', 's_cerevisiae', \"saccharomyces cerevisiae (baker's yeast)\", '4932', 'scer', 'scerevisiae', \"baker's yeast\", 'saccharomyces cerevisiae s288c', 'saccharomyces cerevisiae'], 'name': 'saccharomyces_cerevisiae'}, {'division': 'EnsemblVertebrates', 'common_name': 'rabbit', 'strain_collection': None, 'strain': None, 'assembly': 'OryCun2.0', 'display_name': 'Rabbit', 'groups': ['otherfeatures', 'funcgen', 'rnaseq', 'core'], 'aliases': ['rabbit', 'ocun', 'oryctolagus cuniculus', '9986', 'ocuniculus', 'orycun'], 'name': 'oryctolagus_cuniculus', 'accession': 'GCA_000003625.1', 'taxon_id': '9986', 'release': 96}, {'taxon_id': '6239', 'accession': 'GCA_000002985.3', 'release': 96, 'aliases': ['caenorhabditis elegans', 'c.elegans', 'worm', '6239', 'caenorhabditis_elegans_prjna13758', 'caeele', 'elegans', 'cele', 'celegans'], 'name': 'caenorhabditis_elegans', 'strain': 'N2', 'assembly': 'WBcel235', 'display_name': 'Caenorhabditis elegans', 'groups': ['funcgen', 'core'], 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'C.elegans'}, {'strain_collection': None, 'common_name': 'Chinese hamster', 'division': 'EnsemblVertebrates', 'display_name': 'Chinese hamster CriGri', 'groups': ['funcgen', 'rnaseq', 'core', 'otherfeatures'], 'assembly': 'CriGri_1.0', 'strain': None, 'aliases': ['chinese hamster crigri', 'cgriseus_crigri', 'cricetulus griseus crigri'], 'name': 'cricetulus_griseus_crigri', 'release': 96, 'taxon_id': '10029', 'accession': 'GCA_000223135.1'}, {'name': 'numida_meleagris', 'aliases': [], 'release': 96, 'accession': 'GCA_002078875.2', 'taxon_id': '8996', 'common_name': 'helmeted guineafowl', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'groups': ['rnaseq', 'core', 'otherfeatures'], 'display_name': 'Helmeted guineafowl', 'strain': None, 'assembly': 'NumMel1.0'}, {'assembly': 'ailMel1', 'strain': None, 'groups': ['core', 'otherfeatures'], 'display_name': 'Panda', 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'giant panda', 'taxon_id': '9646', 'accession': 'GCA_000004335.1', 'release': 96, 'aliases': ['amelanoleuca', 'ailuropoda melanoleuca', 'giant panda', 'ailmel', 'panda', 'amel', '9646', 'ailmel1'], 'name': 'ailuropoda_melanoleuca'}, {'release': 96, 'taxon_id': '161767', 'accession': 'GCA_003047355.1', 'aliases': [], 'name': 'amphiprion_percula', 'display_name': 'Orange clownfish', 'groups': ['core', 'rnaseq'], 'assembly': 'Nemo_v1', 'strain': None, 'strain_collection': None, 'common_name': 'orange clownfish', 'division': 'EnsemblVertebrates'}, {'taxon_id': '8364', 'accession': 'GCA_000004195.1', 'release': 96, 'aliases': ['xtropicalis', 'xtro', 'xentro', 'xenopus', 'xenopus tropicalis', 'western clawed frog', 'x.tropicalis', '8364'], 'name': 'xenopus_tropicalis', 'strain': None, 'assembly': 'JGI_4.2', 'display_name': 'Xenopus', 'groups': ['core', 'funcgen', 'otherfeatures'], 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'western clawed frog'}, {'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Great spotted kiwi', 'strain': None, 'assembly': 'aptHaa1', 'groups': ['rnaseq', 'core'], 'display_name': 'Great spotted kiwi', 'aliases': [], 'name': 'apteryx_haastii', 'taxon_id': '8823', 'accession': 'GCA_003342985.1', 'release': 96}, {'assembly': 'PAHARI_EIJ_v1.1', 'strain': 'PAHARI_EIJ', 'display_name': 'Shrew mouse', 'groups': ['core'], 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Shrew mouse', 'taxon_id': '10093', 'accession': 'GCA_900095145.2', 'release': 96, 'name': 'mus_pahari', 'aliases': ['shrew mouse', 'mpahari', '10093', 'mus pahari', 'muspah', 'mpah']}, {'release': 96, 'taxon_id': '93934', 'accession': 'GCA_001577835.1', 'name': 'coturnix_japonica', 'aliases': [], 'display_name': 'Japanese quail', 'groups': ['otherfeatures', 'rnaseq', 'core'], 'strain': None, 'assembly': 'Coturnix_japonica_2.0', 'strain_collection': None, 'common_name': 'Japanese quail', 'division': 'EnsemblVertebrates'}, {'release': 96, 'accession': None, 'taxon_id': '8049', 'aliases': ['atlantic_cod', 'gmorhua', 'atlantic cod', '8049', 'gadus morhua', 'gadmor', 'cod', 'gmor'], 'name': 'gadus_morhua', 'groups': ['core'], 'display_name': 'Cod', 'strain': None, 'assembly': 'gadMor1', 'common_name': 'Atlantic cod', 'strain_collection': None, 'division': 'EnsemblVertebrates'}, {'release': 96, 'taxon_id': '9598', 'accession': 'GCA_000001515.5', 'name': 'pan_troglodytes', 'aliases': ['common chimpanzee', 'ptro', 'ptroglodytes', 'chimp', 'pan troglodytes', '9598', 'chimpanzee', 'pantro'], 'display_name': 'Chimpanzee', 'groups': ['funcgen', 'rnaseq', 'core', 'variation', 'otherfeatures'], 'strain': None, 'assembly': 'Pan_tro_3.0', 'strain_collection': None, 'common_name': 'chimpanzee', 'division': 'EnsemblVertebrates'}, {'aliases': [], 'name': 'ictalurus_punctatus', 'release': 96, 'accession': 'GCA_001660625.1', 'taxon_id': '7998', 'common_name': 'channel catfish', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'groups': ['otherfeatures', 'core', 'funcgen', 'rnaseq'], 'display_name': 'Channel catfish', 'assembly': 'IpCoco_1.2', 'strain': None}, {'display_name': 'Wallaby', 'groups': ['core'], 'assembly': 'Meug_1.0', 'strain': None, 'strain_collection': None, 'common_name': 'tammar wallaby', 'division': 'EnsemblVertebrates', 'release': 96, 'taxon_id': '9315', 'accession': 'GCA_000004035.1', 'name': 'notamacropus_eugenii', 'aliases': ['macropus', 'neug', 'notamacropus', 'meugenii', 'macropus eugenii', 'noteug', 'macropuseugenii', 'meug', 'notamacropuseugenii', 'maceug', '9315', 'tammar wallaby', 'wallaby', 'notamacropus eugenii', 'macropus_eugenii', 'neugenii']}, {'strain': None, 'assembly': 'Sdu_1.0', 'display_name': 'Greater amberjack', 'groups': ['core', 'rnaseq', 'otherfeatures'], 'division': 'EnsemblVertebrates', 'common_name': 'greater amberjack', 'strain_collection': None, 'accession': 'GCA_002260705.1', 'taxon_id': '41447', 'release': 96, 'name': 'seriola_dumerili', 'aliases': []}, {'name': 'mus_musculus_wsbeij', 'aliases': ['mus musculus domesticus wsbeij', 'mouse wsbeij', 'mmusculus_domesticus_wsbeij', 'mus_musculus_domesticus_wsbeij', 'western european house mouse', '10092'], 'accession': 'GCA_001624835.1', 'taxon_id': '10092', 'release': 96, 'division': 'EnsemblVertebrates', 'common_name': 'western european house mouse', 'strain_collection': 'mouse', 'assembly': 'WSB_EiJ_v1', 'strain': 'WSB/EiJ', 'groups': ['funcgen', 'core'], 'display_name': 'Mouse WSB/EiJ'}, {'name': 'gallus_gallus', 'aliases': ['gallus gallus', 'ggallus', '9031', 'ggal', 'galgal', 'chicken', 'g_gallus'], 'accession': 'GCA_000002315.5', 'taxon_id': '9031', 'release': 96, 'division': 'EnsemblVertebrates', 'common_name': 'chicken', 'strain_collection': None, 'strain': None, 'assembly': 'GRCg6a', 'groups': ['rnaseq', 'funcgen', 'core', 'variation', 'otherfeatures'], 'display_name': 'Chicken'}, {'strain': None, 'assembly': 'Caty_1.0', 'groups': ['funcgen', 'rnaseq', 'core', 'otherfeatures'], 'display_name': 'Sooty mangabey', 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Sooty mangabey', 'taxon_id': '9531', 'accession': 'GCA_000955945.1', 'release': 96, 'aliases': [], 'name': 'cercocebus_atys'}, {'accession': 'GCA_000001895.4', 'taxon_id': '10116', 'release': 96, 'name': 'rattus_norvegicus', 'aliases': ['ratnor', 'rat', 'rattus norvegicus', 'r_norvegicus', 'rattus', 'rnorvegicus', 'rnor', 'norway rat', '10116'], 'assembly': 'Rnor_6.0', 'strain': None, 'display_name': 'Rat', 'groups': ['core', 'funcgen', 'rnaseq', 'otherfeatures', 'variation'], 'division': 'EnsemblVertebrates', 'common_name': 'Norway rat', 'strain_collection': None}, {'display_name': 'Okarito brown kiwi', 'groups': ['otherfeatures', 'rnaseq', 'core'], 'strain': None, 'assembly': 'aptRow1', 'common_name': 'Okarito brown kiwi', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'release': 96, 'accession': 'GCA_003343035.1', 'taxon_id': '308060', 'aliases': [], 'name': 'apteryx_rowi'}, {'release': 96, 'accession': None, 'taxon_id': '42254', 'aliases': ['shrew', '42254', 'ground_shrew', 'saraneus', 'sorex araneus', 'european shrew', 'sara', 'european_shrew', 'sorara'], 'name': 'sorex_araneus', 'display_name': 'Shrew', 'groups': ['core'], 'assembly': 'COMMON_SHREW1', 'strain': None, 'common_name': 'European shrew', 'strain_collection': None, 'division': 'EnsemblVertebrates'}, {'aliases': [], 'name': 'calidris_pugnax', 'release': 96, 'accession': 'GCA_001431845.1', 'taxon_id': '198806', 'common_name': 'ruff', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'display_name': 'Ruff', 'groups': ['otherfeatures', 'rnaseq', 'core'], 'strain': None, 'assembly': 'ASM143184v1'}, {'assembly': 'pvi1.1', 'strain': None, 'display_name': 'Central bearded dragon', 'groups': ['otherfeatures', 'rnaseq', 'core'], 'division': 'EnsemblVertebrates', 'common_name': 'central bearded dragon', 'strain_collection': None, 'accession': 'GCA_900067755.1', 'taxon_id': '103695', 'release': 96, 'name': 'pogona_vitticeps', 'aliases': []}, {'release': 96, 'taxon_id': '244447', 'accession': 'GCA_000523025.1', 'aliases': [], 'name': 'cynoglossus_semilaevis', 'groups': ['rnaseq', 'core', 'otherfeatures'], 'display_name': 'Tongue sole', 'assembly': 'Cse_v1.0', 'strain': None, 'strain_collection': None, 'common_name': 'tongue sole', 'division': 'EnsemblVertebrates'}, {'common_name': 'northern tree shrew', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'groups': ['core'], 'display_name': 'Tree Shrew', 'assembly': 'TREESHREW', 'strain': None, 'name': 'tupaia_belangeri', 'aliases': ['37347', 'tree_shrew', 'tbel', 'northern tree shrew', 'tbelangeri', 'tree shrew', 'tupbel', 'tupaia belangeri'], 'release': 96, 'accession': None, 'taxon_id': '37347'}, {'name': 'astatotilapia_calliptera', 'aliases': [], 'release': 96, 'accession': 'GCA_900246225.3', 'taxon_id': '8154', 'common_name': 'eastern happy', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'groups': ['rnaseq', 'core'], 'display_name': 'Eastern happy', 'strain': None, 'assembly': 'fAstCal1.2'}, {'groups': ['funcgen', 'core'], 'display_name': 'Mouse NOD/ShiLtJ', 'strain': 'NOD/ShiLtJ', 'assembly': 'NOD_ShiLtJ_v1', 'strain_collection': 'mouse', 'common_name': 'house mouse', 'division': 'EnsemblVertebrates', 'release': 96, 'taxon_id': '10090', 'accession': 'GCA_001624675.1', 'name': 'mus_musculus_nodshiltj', 'aliases': []}, {'aliases': [], 'name': 'pundamilia_nyererei', 'release': 96, 'taxon_id': '303518', 'accession': 'GCA_000239375.1', 'strain_collection': None, 'common_name': 'Makobe Island cichlid', 'division': 'EnsemblVertebrates', 'display_name': 'Makobe Island cichlid', 'groups': ['rnaseq', 'core', 'otherfeatures'], 'assembly': 'PunNye1.0', 'strain': None}, {'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Collared flycatcher', 'assembly': 'FicAlb_1.4', 'strain': None, 'groups': ['otherfeatures', 'rnaseq', 'core'], 'display_name': 'Flycatcher', 'aliases': ['ficalb', '59894', 'ficedula albicollis', 'f_albicollis', 'flycatcher', 'falbicollis', 'falb', 'collared flycatcher'], 'name': 'ficedula_albicollis', 'taxon_id': '59894', 'accession': 'GCA_000247815.1', 'release': 96}, {'strain_collection': None, 'common_name': 'yellowtail amberjack', 'division': 'EnsemblVertebrates', 'display_name': 'Yellowtail amberjack', 'groups': ['rnaseq', 'core', 'otherfeatures'], 'assembly': 'Sedor1', 'strain': None, 'aliases': [], 'name': 'seriola_lalandi_dorsalis', 'release': 96, 'taxon_id': '1841481', 'accession': 'GCA_002814215.1'}, {'release': 96, 'accession': 'GCA_000633615.2', 'taxon_id': '8081', 'aliases': [], 'name': 'poecilia_reticulata', 'groups': ['otherfeatures', 'rnaseq', 'core'], 'display_name': 'Guppy', 'assembly': 'Guppy_female_1.0_MT', 'strain': None, 'common_name': 'guppy', 'strain_collection': None, 'division': 'EnsemblVertebrates'}, {'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'western European hedgehog', 'strain': None, 'assembly': 'HEDGEHOG', 'display_name': 'Hedgehog', 'groups': ['core'], 'aliases': ['erieur', 'western european hedgehog', 'eeuropaeus', 'western_european_hedgehog', 'erinaceus europaeus', 'eeur', '9365', 'hedgehog'], 'name': 'erinaceus_europaeus', 'taxon_id': '9365', 'accession': None, 'release': 96}, {'division': 'EnsemblVertebrates', 'common_name': 'dingo', 'strain_collection': None, 'assembly': 'ASM325472v1', 'strain': None, 'groups': ['otherfeatures', 'core', 'rnaseq'], 'display_name': 'Dingo', 'name': 'canis_lupus_dingo', 'aliases': [], 'accession': 'GCA_003254725.1', 'taxon_id': '286419', 'release': 96}, {'aliases': [], 'name': 'pan_paniscus', 'accession': 'GCA_000258655.2', 'taxon_id': '9597', 'release': 96, 'division': 'EnsemblVertebrates', 'common_name': 'bonobo', 'strain_collection': None, 'strain': None, 'assembly': 'panpan1.1', 'groups': ['core', 'funcgen', 'rnaseq', 'otherfeatures'], 'display_name': 'Bonobo'}, {'release': 96, 'taxon_id': '8078', 'accession': 'GCA_000826765.1', 'name': 'fundulus_heteroclitus', 'aliases': [], 'groups': ['otherfeatures', 'funcgen', 'rnaseq', 'core'], 'display_name': 'Mummichog', 'strain': None, 'assembly': 'Fundulus_heteroclitus-3.0.2', 'strain_collection': None, 'common_name': 'mummichog', 'division': 'EnsemblVertebrates'}, {'release': 96, 'accession': 'GCF_000002275.2', 'taxon_id': '9258', 'aliases': ['ornithorhynchus anatinus', '9258', 'oanatinus', 'platypus', 'oana', 'ornana'], 'name': 'ornithorhynchus_anatinus', 'display_name': 'Platypus', 'groups': ['funcgen', 'rnaseq', 'core', 'variation', 'otherfeatures'], 'assembly': 'OANA5', 'strain': None, 'common_name': 'platypus', 'strain_collection': None, 'division': 'EnsemblVertebrates'}, {'division': 'EnsemblVertebrates', 'common_name': 'sheepshead minnow', 'strain_collection': None, 'assembly': 'C_variegatus-1.0', 'strain': None, 'display_name': 'Sheepshead minnow', 'groups': ['rnaseq', 'funcgen', 'core', 'otherfeatures'], 'aliases': [], 'name': 'cyprinodon_variegatus', 'accession': 'GCA_000732505.1', 'taxon_id': '28743', 'release': 96}, {'release': 96, 'taxon_id': '10160', 'accession': 'GCA_000260255.1', 'aliases': ['octdeg', 'degu', 'octodon degus', '10160', 'odegus', 'odeg'], 'name': 'octodon_degus', 'groups': ['otherfeatures', 'core'], 'display_name': 'Degu', 'assembly': 'OctDeg1.0', 'strain': None, 'strain_collection': None, 'common_name': 'Degu', 'division': 'EnsemblVertebrates'}, {'taxon_id': '9691', 'accession': 'GCA_001857705.1', 'release': 96, 'aliases': ['leopard'], 'name': 'panthera_pardus', 'strain': None, 'assembly': 'PanPar1.0', 'display_name': 'Leopard', 'groups': ['core', 'otherfeatures'], 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'leopard'}, {'common_name': 'Japanese medaka HdrR', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'groups': ['rnaseq', 'core', 'otherfeatures'], 'display_name': 'Japanese medaka HdrR', 'assembly': 'ASM223467v1', 'strain': None, 'aliases': ['orylat', 'japanese medaka', 'medaka', '8090', 'oryzias latipes', 'olat', 'olatipes'], 'name': 'oryzias_latipes', 'release': 96, 'accession': 'GCA_002234675.1', 'taxon_id': '8090'}, {'strain_collection': 'mouse', 'common_name': 'house mouse', 'division': 'EnsemblVertebrates', 'groups': ['funcgen', 'core'], 'display_name': 'Mouse LP/J', 'assembly': 'LP_J_v1', 'strain': 'LP/J', 'aliases': [], 'name': 'mus_musculus_lpj', 'release': 96, 'taxon_id': '10090', 'accession': 'GCA_001632615.1'}, {'release': 96, 'taxon_id': '8128', 'accession': 'GCA_000188235.1', 'name': 'oreochromis_niloticus', 'aliases': ['nile tilapia', '8128', 'oreochromis niloticus', 'oniloticus', 'tilapia', 'o. niloticus', 'orenil', 'onil'], 'display_name': 'Tilapia', 'groups': ['rnaseq', 'core', 'otherfeatures'], 'assembly': 'Orenil1.0', 'strain': None, 'strain_collection': None, 'common_name': 'Nile tilapia', 'division': 'EnsemblVertebrates'}, {'display_name': 'Clown anemonefish', 'groups': ['rnaseq', 'core', 'otherfeatures'], 'assembly': 'AmpOce1.0', 'strain': None, 'common_name': 'clown anemonefish', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'release': 96, 'accession': 'GCA_002776465.1', 'taxon_id': '80972', 'name': 'amphiprion_ocellaris', 'aliases': []}, {'name': 'mus_spretus', 'aliases': ['mus_spretus_spret_eij', 'mspretus', 'western mediterranean mouse', 'spreteij'], 'accession': 'GCA_001624865.1', 'taxon_id': '10096', 'release': 96, 'division': 'EnsemblVertebrates', 'common_name': 'algerian mouse', 'strain_collection': None, 'assembly': 'SPRET_EiJ_v1', 'strain': 'SPRET/EiJ', 'groups': ['funcgen', 'core'], 'display_name': 'Algerian mouse'}, {'accession': None, 'taxon_id': '99883', 'release': 96, 'name': 'tetraodon_nigroviridis', 'aliases': ['tetraodon', 'tetraodon nigroviridis', '99883', 'spotted green pufferfish', 'tetnig', 'tnig', 'tnigroviridis', 'fresh water pufferfish'], 'strain': None, 'assembly': 'TETRAODON8', 'display_name': 'Tetraodon', 'groups': ['otherfeatures', 'variation', 'core'], 'division': 'EnsemblVertebrates', 'common_name': 'spotted green pufferfish', 'strain_collection': None}, {'aliases': [], 'name': 'esox_lucius', 'release': 96, 'accession': 'GCA_000721915.3', 'taxon_id': '8010', 'common_name': 'northern pike', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'groups': ['otherfeatures', 'core', 'rnaseq'], 'display_name': 'Northern pike', 'assembly': 'Eluc_V3', 'strain': None}, {'groups': ['otherfeatures', 'core', 'rnaseq'], 'display_name': 'Amazon molly', 'assembly': 'PoeFor_5.1.2', 'strain': None, 'common_name': 'Amazon molly', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'release': 96, 'accession': 'GCA_000485575.1', 'taxon_id': '48698', 'name': 'poecilia_formosa', 'aliases': ['pformosa', 'poefor', 'amazon molly', 'poecilia formosa', '48698', 'pfor']}, {'release': 96, 'taxon_id': '13146', 'accession': 'GCA_000238935.1', 'aliases': [], 'name': 'melopsittacus_undulatus', 'display_name': 'Budgerigar', 'groups': ['otherfeatures', 'rnaseq', 'core'], 'assembly': 'Melopsittacus_undulatus_6.3', 'strain': None, 'strain_collection': None, 'common_name': 'Budgie', 'division': 'EnsemblVertebrates'}, {'accession': None, 'taxon_id': '69293', 'release': 96, 'aliases': ['69293', 'gaculeatus', 'gacu', 'gasacu', 'three spined stickleback', 'three-spined stickleback', 'gasterosteus aculeatus', 'stickleback'], 'name': 'gasterosteus_aculeatus', 'strain': None, 'assembly': 'BROADS1', 'display_name': 'Stickleback', 'groups': ['core', 'otherfeatures'], 'division': 'EnsemblVertebrates', 'common_name': 'three-spined stickleback', 'strain_collection': None}, {'accession': 'GCA_003342965.1', 'taxon_id': '8824', 'release': 96, 'name': 'apteryx_owenii', 'aliases': [], 'strain': None, 'assembly': 'aptOwe1', 'groups': ['rnaseq', 'core'], 'display_name': 'Little spotted kiwi', 'division': 'EnsemblVertebrates', 'common_name': 'little spotted kiwi', 'strain_collection': None}, {'name': 'neovison_vison', 'aliases': [], 'accession': 'GCA_900108605.1', 'taxon_id': '452646', 'release': 96, 'division': 'EnsemblVertebrates', 'common_name': 'American mink', 'strain_collection': None, 'strain': None, 'assembly': 'NNQGG.v01', 'display_name': 'American mink', 'groups': ['rnaseq', 'core']}, {'groups': ['core', 'otherfeatures'], 'display_name': 'Chilean tinamou', 'assembly': 'notPer1', 'strain': None, 'strain_collection': None, 'common_name': 'birds', 'division': 'EnsemblVertebrates', 'release': 96, 'taxon_id': '30464', 'accession': 'GCA_003342845.1', 'aliases': [], 'name': 'nothoprocta_perdicaria'}, {'taxon_id': '7955', 'accession': 'GCA_000002035.4', 'release': 96, 'name': 'danio_rerio', 'aliases': ['danrer', '7955', 'zfish', 'danio rerio', 'drerio', 'danio', 'drer', 'd_rerio', 'zebrafish'], 'strain': None, 'assembly': 'GRCz11', 'display_name': 'Zebrafish', 'groups': ['otherfeatures', 'variation', 'core', 'funcgen', 'rnaseq'], 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'zebrafish'}, {'strain_collection': 'mouse', 'common_name': 'house mouse', 'division': 'EnsemblVertebrates', 'groups': ['funcgen', 'core'], 'display_name': 'Mouse AKR/J', 'assembly': 'AKR_J_v1', 'strain': 'AKR/J', 'aliases': [], 'name': 'mus_musculus_akrj', 'release': 96, 'taxon_id': '10090', 'accession': 'GCA_001624295.1'}, {'release': 96, 'taxon_id': '52904', 'accession': 'GCA_003186165.1', 'name': 'scophthalmus_maximus', 'aliases': [], 'groups': ['core', 'funcgen', 'rnaseq'], 'display_name': 'Turbot', 'strain': None, 'assembly': 'ASM318616v1', 'strain_collection': None, 'common_name': 'turbot', 'division': 'EnsemblVertebrates'}, {'groups': ['core', 'rnaseq', 'otherfeatures'], 'display_name': 'Wild yak', 'strain': None, 'assembly': 'BosGru_v2.0', 'common_name': 'wild yak', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'release': 96, 'accession': 'GCA_000298355.1', 'taxon_id': '72004', 'name': 'bos_mutus', 'aliases': []}, {'aliases': ['c_familiaris', 'canis_lupus_familiaris', 'canfam', 'canis familiaris', 'cfamiliaris', 'cfam', 'canis', '9615', 'canis lupus familiaris', 'dog'], 'name': 'canis_familiaris', 'release': 96, 'taxon_id': '9615', 'accession': 'GCA_000002285.2', 'strain_collection': None, 'common_name': 'dog', 'division': 'EnsemblVertebrates', 'display_name': 'Dog', 'groups': ['core', 'funcgen', 'rnaseq', 'otherfeatures', 'variation'], 'strain': None, 'assembly': 'CanFam3.1'}, {'release': 96, 'taxon_id': '38772', 'accession': 'GCA_002896415.1', 'aliases': [], 'name': 'gopherus_agassizii', 'groups': ['core', 'rnaseq'], 'display_name': \"Agassiz's desert tortoise\", 'assembly': 'ASM289641v1', 'strain': None, 'strain_collection': None, 'common_name': \"Agassiz's desert tortoise\", 'division': 'EnsemblVertebrates'}, {'aliases': [], 'name': 'anabas_testudineus', 'release': 96, 'taxon_id': '64144', 'accession': 'GCA_900324465.1', 'strain_collection': None, 'common_name': 'climbing perch', 'division': 'EnsemblVertebrates', 'display_name': 'Climbing perch', 'groups': ['rnaseq', 'core'], 'strain': None, 'assembly': 'fAnaTes1.1'}, {'display_name': 'Mainland tiger snake', 'groups': ['rnaseq', 'core', 'otherfeatures'], 'assembly': 'TS10Xv2-PRI', 'strain': None, 'common_name': 'mainland tiger snake', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'release': 96, 'accession': 'GCA_900518725.1', 'taxon_id': '8663', 'name': 'notechis_scutatus', 'aliases': []}, {'strain': None, 'assembly': 'PM.fa', 'display_name': 'Periophthalmus magnuspinnatus', 'groups': ['core'], 'division': 'EnsemblVertebrates', 'common_name': 'Periophthalmus magnuspinnatus', 'strain_collection': None, 'accession': 'GCA_000787105.1', 'taxon_id': '409849', 'release': 96, 'name': 'periophthalmus_magnuspinnatus', 'aliases': []}, {'release': 96, 'taxon_id': '10029', 'accession': 'GCA_900186095.1', 'name': 'cricetulus_griseus_chok1gshd', 'aliases': ['cgriseus_chok1gshd', 'cgri', 'cricetulus griseus chok1gshd', 'chinese hamster chok1gs', 'crigri'], 'display_name': 'Chinese hamster CHOK1GS', 'groups': ['rnaseq', 'funcgen', 'core', 'otherfeatures'], 'assembly': 'CHOK1GS_HDv1', 'strain': None, 'strain_collection': None, 'common_name': 'Chinese hamster', 'division': 'EnsemblVertebrates'}, {'groups': ['otherfeatures', 'rnaseq', 'funcgen', 'core'], 'display_name': 'Crab-eating macaque', 'assembly': 'Macaca_fascicularis_5.0', 'strain': None, 'strain_collection': None, 'common_name': 'Crab-eating macaque', 'division': 'EnsemblVertebrates', 'release': 96, 'taxon_id': '9541', 'accession': 'GCA_000364345.1', 'name': 'macaca_fascicularis', 'aliases': []}, {'strain_collection': None, 'common_name': \"Burton's mouthbrooder\", 'division': 'EnsemblVertebrates', 'display_name': \"Burton's mouthbrooder\", 'groups': ['rnaseq', 'core', 'otherfeatures'], 'strain': None, 'assembly': 'AstBur1.0', 'aliases': [], 'name': 'haplochromis_burtoni', 'release': 96, 'taxon_id': '8153', 'accession': 'GCA_000239415.1'}, {'assembly': 'PPYG2', 'strain': None, 'groups': ['otherfeatures', 'variation', 'core', 'rnaseq'], 'display_name': 'Orangutan', 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Bornean orangutan', 'taxon_id': '9601', 'accession': None, 'release': 96, 'aliases': ['ponpyg', 'ponabe', '9601', 'orang', 'ppygmaeus', 'pongo_pygmaeus', 'pabe', 'pabelii', 'sumatran orangutan', 'ppyg', 'pongo pygmaeus', 'orang_utan', 'orangutan', 'bornean orangutan', 'pongo abelii'], 'name': 'pongo_abelii'}, {'release': 96, 'taxon_id': '9545', 'accession': 'GCA_000956065.1', 'name': 'macaca_nemestrina', 'aliases': [], 'groups': ['otherfeatures', 'rnaseq', 'funcgen', 'core'], 'display_name': 'Pig-tailed macaque', 'strain': None, 'assembly': 'Mnem_1.0', 'strain_collection': None, 'common_name': 'Pig-tailed macaque', 'division': 'EnsemblVertebrates'}, {'name': 'dipodomys_ordii', 'aliases': ['ords kangaroo rat', 'dipodomysordii', 'kangaroo_rat', \"ord's kangaroo rat\", 'dipodomys ordii', '10020', 'dord', 'dipord', 'kangaroo rat', 'dordii'], 'release': 96, 'taxon_id': '10020', 'accession': 'GCA_000151885.2', 'strain_collection': None, 'common_name': \"Ord's kangaroo rat\", 'division': 'EnsemblVertebrates', 'display_name': 'Kangaroo rat', 'groups': ['core', 'otherfeatures'], 'assembly': 'Dord_2.0', 'strain': None}, {'taxon_id': '7757', 'accession': None, 'release': 96, 'name': 'petromyzon_marinus', 'aliases': ['7757', 'pmarinus', 'sea lamprey', 'petromyzon marinus', 'petmar', 'lamprey', 'pmar'], 'strain': None, 'assembly': 'Pmarinus_7.0', 'groups': ['otherfeatures', 'core'], 'display_name': 'Lamprey', 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'sea lamprey'}, {'aliases': [], 'name': 'scleropages_formosus', 'accession': 'GCA_001624265.1', 'taxon_id': '113540', 'release': 96, 'division': 'EnsemblVertebrates', 'common_name': 'Asian bonytongue', 'strain_collection': None, 'assembly': 'ASM162426v1', 'strain': None, 'groups': ['otherfeatures', 'core', 'rnaseq'], 'display_name': 'Asian bonytongue'}, {'release': 96, 'accession': 'GCA_900324485.1', 'taxon_id': '205130', 'aliases': [], 'name': 'mastacembelus_armatus', 'display_name': 'Zig-zag eel', 'groups': ['rnaseq', 'core'], 'assembly': 'fMasArm1.1', 'strain': None, 'common_name': 'zig-zag eel', 'strain_collection': None, 'division': 'EnsemblVertebrates'}, {'release': 96, 'taxon_id': '425635', 'accession': 'GCA_003697955.1', 'name': 'calidris_pygmaea', 'aliases': [], 'groups': ['rnaseq', 'core'], 'display_name': 'Spoon-billed sandpiper', 'assembly': 'ASM369795v1', 'strain': None, 'strain_collection': None, 'common_name': 'Spoon-billed sandpiper', 'division': 'EnsemblVertebrates'}, {'taxon_id': '9103', 'accession': 'GCA_000146605.1', 'release': 96, 'aliases': ['mgal', 'common turkey', '9103', 'mga', 'mgallopavo', 'melgal', 'turkey', 'wild turkey', 'meleagris gallopavo'], 'name': 'meleagris_gallopavo', 'strain': None, 'assembly': 'Turkey_2.01', 'display_name': 'Turkey', 'groups': ['core', 'otherfeatures', 'variation'], 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'domestic turkey'}, {'taxon_id': '10090', 'accession': 'GCA_001624475.1', 'release': 96, 'name': 'mus_musculus_cbaj', 'aliases': [], 'strain': 'CBA/J', 'assembly': 'CBA_J_v1', 'display_name': 'Mouse CBA/J', 'groups': ['funcgen', 'core'], 'division': 'EnsemblVertebrates', 'strain_collection': 'mouse', 'common_name': 'house mouse'}, {'aliases': [], 'name': 'meriones_unguiculatus', 'release': 96, 'accession': 'GCA_002204375.1', 'taxon_id': '10047', 'common_name': 'Mongolian gerbil', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'display_name': 'Mongolian gerbil', 'groups': ['otherfeatures', 'core', 'rnaseq'], 'strain': None, 'assembly': 'MunDraft-v1.0'}, {'aliases': [], 'name': 'kryptolebias_marmoratus', 'release': 96, 'accession': 'GCA_001649575.1', 'taxon_id': '37003', 'common_name': 'mangrove rivulus', 'strain_collection': None, 'division': 'EnsemblVertebrates', 'groups': ['otherfeatures', 'core', 'rnaseq'], 'display_name': 'Mangrove rivulus', 'assembly': 'ASM164957v1', 'strain': None}, {'assembly': 'loxAfr3', 'strain': None, 'display_name': 'Elephant', 'groups': ['core'], 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'African savanna elephant', 'taxon_id': '9785', 'accession': 'GCA_000001905.1', 'release': 96, 'name': 'loxodonta_africana', 'aliases': ['9785', 'elephant', 'african savanna elephant', 'loxafr', 'loxodonta africana', 'african_elephant', 'lafricana', 'lafr']}, {'strain': None, 'assembly': 'ChiLan1.0', 'groups': ['otherfeatures', 'rnaseq', 'core'], 'display_name': 'Long-tailed chinchilla', 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Long-tailed chinchilla', 'taxon_id': '34839', 'accession': 'GCA_000276665.1', 'release': 96, 'aliases': ['chinchilla lanigera', 'chilan', 'clanigera', 'clan', '34839', 'long-tailed chinchilla'], 'name': 'chinchilla_lanigera'}, {'release': 96, 'accession': 'GCA_001522545.2', 'taxon_id': '9157', 'name': 'parus_major', 'aliases': [], 'display_name': 'Great Tit', 'groups': ['otherfeatures', 'core', 'rnaseq'], 'strain': None, 'assembly': 'Parus_major1.1', 'common_name': 'Great Tit', 'strain_collection': None, 'division': 'EnsemblVertebrates'}, {'assembly': 'P_latipinna-1.0', 'strain': None, 'groups': ['otherfeatures', 'rnaseq', 'core'], 'display_name': 'Sailfin molly', 'division': 'EnsemblVertebrates', 'common_name': 'sailfin molly', 'strain_collection': None, 'accession': 'GCA_001443285.1', 'taxon_id': '48699', 'release': 96, 'aliases': [], 'name': 'poecilia_latipinna'}, {'name': 'oryzias_latipes_hsok', 'aliases': [], 'accession': 'GCA_002234695.1', 'taxon_id': '8090', 'release': 96, 'division': 'EnsemblVertebrates', 'common_name': 'Japanese medaka HSOK', 'strain_collection': None, 'strain': None, 'assembly': 'ASM223469v1', 'groups': ['rnaseq', 'core'], 'display_name': 'Japanese medaka HSOK'}, {'display_name': 'Cow', 'groups': ['core', 'rnaseq', 'funcgen', 'otherfeatures', 'variation'], 'strain': None, 'assembly': 'ARS-UCD1.2', 'strain_collection': None, 'common_name': 'cattle', 'division': 'EnsemblVertebrates', 'release': 96, 'taxon_id': '9913', 'accession': 'GCA_002263795.2', 'name': 'bos_taurus', 'aliases': ['bostau', 'bovine', 'btaurus', '9913', 'cow', 'cattle', 'bos taurus', 'domestic cattle', 'domestic cow', 'btau']}, {'release': 96, 'accession': None, 'taxon_id': '9358', 'aliases': ['choloepus', 'choffmanni', 'chohof', \"hoffmann's two-fingered sloth\", 'choloepus hoffmanni', '9358', 'choloepushoffmanni', 'chof', 'two-toed sloth', 'sloth'], 'name': 'choloepus_hoffmanni', 'groups': ['core'], 'display_name': 'Sloth', 'assembly': 'choHof1', 'strain': None, 'common_name': \"Hoffmann's two-fingered sloth\", 'strain_collection': None, 'division': 'EnsemblVertebrates'}, {'release': 96, 'taxon_id': '29073', 'accession': 'GCA_000687225.1', 'name': 'ursus_maritimus', 'aliases': [], 'groups': ['core', 'rnaseq', 'otherfeatures'], 'display_name': 'Polar bear', 'assembly': 'UrsMar_1.0', 'strain': None, 'strain_collection': None, 'common_name': 'Polar bear', 'division': 'EnsemblVertebrates'}, {'strain_collection': None, 'common_name': 'swamp eel', 'division': 'EnsemblVertebrates', 'display_name': 'Swamp eel', 'groups': ['otherfeatures', 'core', 'rnaseq'], 'assembly': 'M_albus_1.0', 'strain': None, 'aliases': [], 'name': 'monopterus_albus', 'release': 96, 'taxon_id': '43700', 'accession': 'GCA_001952655.1'}, {'aliases': [], 'name': 'castor_canadensis', 'release': 96, 'taxon_id': '51338', 'accession': 'GCA_001984765.1', 'strain_collection': None, 'common_name': 'American beaver', 'division': 'EnsemblVertebrates', 'groups': ['rnaseq', 'core', 'otherfeatures'], 'display_name': 'American beaver', 'assembly': 'C.can_genome_v1.0', 'strain': None}, {'strain_collection': None, 'common_name': 'Mexican tetra', 'division': 'EnsemblVertebrates', 'groups': ['otherfeatures', 'core', 'rnaseq'], 'display_name': 'Mexican tetra', 'assembly': 'Astyanax_mexicanus-2.0', 'strain': None, 'name': 'astyanax_mexicanus', 'aliases': ['7994', 'amexicanus', 'astmex', 'amex', 'cave fish', 'astyanax mexicanus'], 'release': 96, 'taxon_id': '7994', 'accession': 'GCA_000372685.2'}, {'release': 96, 'accession': 'GCA_002109545.1', 'taxon_id': '80966', 'aliases': [], 'name': 'acanthochromis_polyacanthus', 'display_name': 'Spiny chromis', 'groups': ['core', 'rnaseq', 'otherfeatures'], 'strain': None, 'assembly': 'ASM210954v1', 'common_name': 'spiny chromis', 'strain_collection': None, 'division': 'EnsemblVertebrates'}, {'accession': 'GCA_000743615.1', 'taxon_id': '885580', 'release': 96, 'aliases': ['fdamarensi', 'fdam', 'fukomys_damarensi', '885580', 'fukomys damarensi', 'fukdam', 'damara mole rat'], 'name': 'fukomys_damarensis', 'strain': None, 'assembly': 'DMR_v1.0', 'groups': ['core', 'rnaseq', 'otherfeatures'], 'display_name': 'Damara mole rat', 'division': 'EnsemblVertebrates', 'common_name': 'Damara mole rat', 'strain_collection': None}, {'division': 'EnsemblVertebrates', 'common_name': 'Paramormyrops kingsleyae', 'strain_collection': None, 'assembly': 'PKINGS_0.1', 'strain': None, 'display_name': 'Paramormyrops kingsleyae', 'groups': ['core', 'rnaseq'], 'aliases': [], 'name': 'paramormyrops_kingsleyae', 'accession': 'GCA_002872115.1', 'taxon_id': '1676925', 'release': 96}, {'assembly': 'Bison_UMD1.0', 'strain': None, 'display_name': 'American bison', 'groups': ['core', 'otherfeatures'], 'division': 'EnsemblVertebrates', 'common_name': 'American bison', 'strain_collection': None, 'accession': 'GCA_000754665.1', 'taxon_id': '43346', 'release': 96, 'name': 'bison_bison_bison', 'aliases': []}, {'taxon_id': '32507', 'accession': 'GCA_000239395.1', 'release': 96, 'name': 'neolamprologus_brichardi', 'aliases': [], 'assembly': 'NeoBri1.0', 'strain': None, 'groups': ['core', 'rnaseq', 'otherfeatures'], 'display_name': 'Lyretail cichlid', 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'lyretail cichlid'}, {'division': 'EnsemblVertebrates', 'common_name': 'horse', 'strain_collection': None, 'strain': None, 'assembly': 'EquCab3.0', 'display_name': 'Horse', 'groups': ['otherfeatures', 'variation', 'core', 'funcgen', 'rnaseq'], 'name': 'equus_caballus', 'aliases': ['ecaballus', 'ecab', 'equcab', '9796', 'equus caballus', 'horse', 'equuscaballus'], 'accession': 'GCA_002863925.1', 'taxon_id': '9796', 'release': 96}, {'release': 96, 'accession': 'GCA_001632525.1', 'taxon_id': '10090', 'aliases': [], 'name': 'mus_musculus_balbcj', 'display_name': 'Mouse BALB/cJ', 'groups': ['core', 'funcgen'], 'assembly': 'BALB_cJ_v1', 'strain': 'BALB/cJ', 'common_name': 'house mouse', 'strain_collection': 'mouse', 'division': 'EnsemblVertebrates'}, {'strain': None, 'assembly': 'MesAur1.0', 'groups': ['otherfeatures', 'core', 'funcgen'], 'display_name': 'Golden Hamster', 'division': 'EnsemblVertebrates', 'strain_collection': None, 'common_name': 'Golden Hamster', 'taxon_id': '10036', 'accession': 'GCA_000349665.1', 'release': 96, 'name': 'mesocricetus_auratus', 'aliases': ['10036', 'mesocricetus auratus', 'mesaur', 'maur', 'golden hamster', 'mauratus']}, {'release': 96, 'accession': 'GCA_000180615.2', 'taxon_id': '31033', 'name': 'takifugu_rubripes', 'aliases': ['takifugu', 'trubripes', 'takifugu rubripes', 't_rubripes', 'takrub', 'torafugu', 'fugu', '31033', 'trub'], 'groups': ['otherfeatures', 'core', 'rnaseq'], 'display_name': 'Fugu', 'strain': None, 'assembly': 'FUGU5', 'common_name': 'torafugu', 'strain_collection': None, 'division': 'EnsemblVertebrates'}]}\n"
     ]
    }
   ],
   "source": [
    "import requests, sys\n",
    " \n",
    "server = \"http://rest.ensembl.org\"\n",
    "ext = \"/info/species?\"\n",
    " \n",
    "r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    " \n",
    "if not r.ok:\n",
    "  r.raise_for_status()\n",
    "  sys.exit()\n",
    " \n",
    "decoded = r.json()\n",
    "print(repr(decoded))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(r.json())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame.from_dict(decoded)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>aliases</th>\n",
       "      <th>assembly</th>\n",
       "      <th>common_name</th>\n",
       "      <th>display_name</th>\n",
       "      <th>division</th>\n",
       "      <th>groups</th>\n",
       "      <th>name</th>\n",
       "      <th>release</th>\n",
       "      <th>strain</th>\n",
       "      <th>strain_collection</th>\n",
       "      <th>taxon_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_000409795.2</td>\n",
       "      <td>[chlorocebus sabeus, chlorocebus_sabeus, chlsa...</td>\n",
       "      <td>ChlSab1.1</td>\n",
       "      <td>African green monkey</td>\n",
       "      <td>Vervet-AGM</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[rnaseq, core, variation, otherfeatures]</td>\n",
       "      <td>chlorocebus_sabaeus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>60711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_002197715.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>LonStrDom1</td>\n",
       "      <td>Bengalese finch</td>\n",
       "      <td>Bengalese finch</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, core, rnaseq]</td>\n",
       "      <td>lonchura_striata_domestica</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>299123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000951045.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>Mleu.le_1.0</td>\n",
       "      <td>Drill</td>\n",
       "      <td>Drill</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[funcgen, core, otherfeatures]</td>\n",
       "      <td>mandrillus_leucophaeus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_002776525.2</td>\n",
       "      <td>[]</td>\n",
       "      <td>ASM277652v2</td>\n",
       "      <td>Ugandan red Colobus</td>\n",
       "      <td>Ugandan red Colobus</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[funcgen, rnaseq, core]</td>\n",
       "      <td>piliocolobus_tephrosceles</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>591936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_001604975.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>Cebus_imitator-1.0</td>\n",
       "      <td>White-headed capuchin</td>\n",
       "      <td>Capuchin</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, funcgen, rnaseq, otherfeatures]</td>\n",
       "      <td>cebus_capucinus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1737458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GCA_003255815.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>Tgel_1.0</td>\n",
       "      <td>gelada</td>\n",
       "      <td>Gelada</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, rnaseq, funcgen]</td>\n",
       "      <td>theropithecus_gelada</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GCA_003597395.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>ASM359739v1</td>\n",
       "      <td>Abingdon island giant tortoise</td>\n",
       "      <td>Abingdon island giant tortoise</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, rnaseq]</td>\n",
       "      <td>chelonoidis_abingdonii</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>106734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GCA_000464555.1</td>\n",
       "      <td>[tiger]</td>\n",
       "      <td>PanTig1.0</td>\n",
       "      <td>Tiger</td>\n",
       "      <td>Tiger</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, rnaseq, otherfeatures]</td>\n",
       "      <td>panthera_tigris_altaica</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>74533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GCA_000001635.8</td>\n",
       "      <td>[mmus, mus musculus, 10090, musmus, m_musculus...</td>\n",
       "      <td>GRCm38</td>\n",
       "      <td>house mouse</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, cdna, rnaseq, funcgen, otherfeatures, v...</td>\n",
       "      <td>mus_musculus</td>\n",
       "      <td>96</td>\n",
       "      <td>reference (CL57BL6)</td>\n",
       "      <td>mouse</td>\n",
       "      <td>10090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GCA_001443325.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>P_mexicana-1.0</td>\n",
       "      <td>shortfin molly</td>\n",
       "      <td>Shortfin molly</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, rnaseq, core]</td>\n",
       "      <td>poecilia_mexicana</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>48701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GCA_000146795.3</td>\n",
       "      <td>[nleu, gibbon, nomleu, 61853, nleucogenys, nom...</td>\n",
       "      <td>Nleu_3.0</td>\n",
       "      <td>Northern white-cheeked gibbon</td>\n",
       "      <td>Gibbon</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[funcgen, core, variation, otherfeatures]</td>\n",
       "      <td>nomascus_leucogenys</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>61853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GCA_000003025.6</td>\n",
       "      <td>[9823, wild boar, pigs, pig, swine, sscr, sscr...</td>\n",
       "      <td>Sscrofa11.1</td>\n",
       "      <td>pig</td>\n",
       "      <td>Pig</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, funcgen, rnaseq, otherfeatures, variation]</td>\n",
       "      <td>sus_scrofa</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GCA_002406435.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>ASM240643v1</td>\n",
       "      <td>Daurian ground squirrel</td>\n",
       "      <td>Daurian ground squirrel</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core]</td>\n",
       "      <td>spermophilus_dauricus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>99837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GCA_000298735.1</td>\n",
       "      <td>[ovisaries, ovis aries, oaries, oar, sheep, 99...</td>\n",
       "      <td>Oar_v3.1</td>\n",
       "      <td>Domestic sheep</td>\n",
       "      <td>Sheep</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[variation, otherfeatures, rnaseq, core]</td>\n",
       "      <td>ovis_aries</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GCA_000688575.1</td>\n",
       "      <td>[cavia aperea, cape, 37548, brazilian guinea p...</td>\n",
       "      <td>CavAp1.0</td>\n",
       "      <td>Brazilian guinea pig</td>\n",
       "      <td>Brazilian guinea pig</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core]</td>\n",
       "      <td>cavia_aperea</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>37548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>None</td>\n",
       "      <td>[sea squirt ciona savignyi, 51511, csav, c.sav...</td>\n",
       "      <td>CSAV2.0</td>\n",
       "      <td>Sea squirt Ciona savignyi</td>\n",
       "      <td>C.savignyi</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, core]</td>\n",
       "      <td>ciona_savignyi</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>51511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GCA_001624505.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>DBA_2J_v1</td>\n",
       "      <td>house mouse</td>\n",
       "      <td>Mouse DBA/2J</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, funcgen]</td>\n",
       "      <td>mus_musculus_dba2j</td>\n",
       "      <td>96</td>\n",
       "      <td>DBA/2J</td>\n",
       "      <td>mouse</td>\n",
       "      <td>10090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>None</td>\n",
       "      <td>[large flying fox, pvampyrus, 132908, pteropus...</td>\n",
       "      <td>pteVam1</td>\n",
       "      <td>large flying fox</td>\n",
       "      <td>Megabat</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core]</td>\n",
       "      <td>pteropus_vampyrus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>132908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>GCA_000956105.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>Pcoq_1.0</td>\n",
       "      <td>Coquerel's sifaka</td>\n",
       "      <td>Coquerel's sifaka</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, funcgen, otherfeatures]</td>\n",
       "      <td>propithecus_coquereli</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>379532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GCA_003033725.1</td>\n",
       "      <td>[ass, donkey]</td>\n",
       "      <td>ASM303372v1</td>\n",
       "      <td>donkey</td>\n",
       "      <td>Donkey</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, rnaseq]</td>\n",
       "      <td>equus_asinus_asinus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>83772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>GCA_000001215.4</td>\n",
       "      <td>[dmel, d.mel, d. melanogaster]</td>\n",
       "      <td>BDGP6.22</td>\n",
       "      <td>fruit fly</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[funcgen, core]</td>\n",
       "      <td>drosophila_melanogaster</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>7227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>GCA_000164805.2</td>\n",
       "      <td>[csyr, tarsius_syrichta, tarsius syrichta, tsy...</td>\n",
       "      <td>Tarsius_syrichta-2.0.1</td>\n",
       "      <td>Philippine tarsier</td>\n",
       "      <td>Tarsier</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, funcgen, otherfeatures]</td>\n",
       "      <td>carlito_syrichta</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1868482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>GCA_000247695.1</td>\n",
       "      <td>[naked mole-rat female, heterocephalus glaber ...</td>\n",
       "      <td>HetGla_female_1.0</td>\n",
       "      <td>naked mole-rat</td>\n",
       "      <td>Naked mole-rat female</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, rnaseq, core]</td>\n",
       "      <td>heterocephalus_glaber_female</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>10181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>GCA_000751415.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>Midas_v5</td>\n",
       "      <td>Midas cichlid</td>\n",
       "      <td>Midas cichlid</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core]</td>\n",
       "      <td>amphilophus_citrinellus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>61819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>GCA_003258685.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>Prosim_1.0</td>\n",
       "      <td>greater bamboo lemur</td>\n",
       "      <td>Greater bamboo lemur</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[funcgen, core]</td>\n",
       "      <td>prolemur_simus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1328070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>GCA_000230445.1</td>\n",
       "      <td>[hgla, heterocephalus glaber male, hetgla, hgl...</td>\n",
       "      <td>HetGla_1.0</td>\n",
       "      <td>naked mole-rat</td>\n",
       "      <td>Naked mole-rat male</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, core, rnaseq]</td>\n",
       "      <td>heterocephalus_glaber_male</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>10181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>GCA_001715985.2</td>\n",
       "      <td>[]</td>\n",
       "      <td>ASM171598v2</td>\n",
       "      <td>golden-collared manakin</td>\n",
       "      <td>Golden-collared manakin</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, rnaseq, core]</td>\n",
       "      <td>manacus_vitellinus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>328815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>GCA_000001405.27</td>\n",
       "      <td>[9606, h_sapiens, homsap, hsapiens, homo, huma...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>human</td>\n",
       "      <td>Human</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, variation, cdna, core, funcgen...</td>\n",
       "      <td>homo_sapiens</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>GCA_000534875.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>SCA1</td>\n",
       "      <td>common canary</td>\n",
       "      <td>Common canary</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, core, rnaseq]</td>\n",
       "      <td>serinus_canaria</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>GCA_001458135.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>marMar2.1</td>\n",
       "      <td>Alpine marmot</td>\n",
       "      <td>Alpine marmot</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, core, rnaseq]</td>\n",
       "      <td>marmota_marmota_marmota</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>GCA_000956065.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>Mnem_1.0</td>\n",
       "      <td>Pig-tailed macaque</td>\n",
       "      <td>Pig-tailed macaque</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, rnaseq, funcgen, core]</td>\n",
       "      <td>macaca_nemestrina</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>GCA_000151885.2</td>\n",
       "      <td>[ords kangaroo rat, dipodomysordii, kangaroo_r...</td>\n",
       "      <td>Dord_2.0</td>\n",
       "      <td>Ord's kangaroo rat</td>\n",
       "      <td>Kangaroo rat</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, otherfeatures]</td>\n",
       "      <td>dipodomys_ordii</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>10020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>None</td>\n",
       "      <td>[7757, pmarinus, sea lamprey, petromyzon marin...</td>\n",
       "      <td>Pmarinus_7.0</td>\n",
       "      <td>sea lamprey</td>\n",
       "      <td>Lamprey</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, core]</td>\n",
       "      <td>petromyzon_marinus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>7757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>GCA_001624265.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>ASM162426v1</td>\n",
       "      <td>Asian bonytongue</td>\n",
       "      <td>Asian bonytongue</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, core, rnaseq]</td>\n",
       "      <td>scleropages_formosus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>113540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>GCA_900324485.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>fMasArm1.1</td>\n",
       "      <td>zig-zag eel</td>\n",
       "      <td>Zig-zag eel</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[rnaseq, core]</td>\n",
       "      <td>mastacembelus_armatus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>205130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>GCA_003697955.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>ASM369795v1</td>\n",
       "      <td>Spoon-billed sandpiper</td>\n",
       "      <td>Spoon-billed sandpiper</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[rnaseq, core]</td>\n",
       "      <td>calidris_pygmaea</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>425635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>GCA_000146605.1</td>\n",
       "      <td>[mgal, common turkey, 9103, mga, mgallopavo, m...</td>\n",
       "      <td>Turkey_2.01</td>\n",
       "      <td>domestic turkey</td>\n",
       "      <td>Turkey</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, otherfeatures, variation]</td>\n",
       "      <td>meleagris_gallopavo</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>GCA_001624475.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>CBA_J_v1</td>\n",
       "      <td>house mouse</td>\n",
       "      <td>Mouse CBA/J</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[funcgen, core]</td>\n",
       "      <td>mus_musculus_cbaj</td>\n",
       "      <td>96</td>\n",
       "      <td>CBA/J</td>\n",
       "      <td>mouse</td>\n",
       "      <td>10090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>GCA_002204375.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>MunDraft-v1.0</td>\n",
       "      <td>Mongolian gerbil</td>\n",
       "      <td>Mongolian gerbil</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, core, rnaseq]</td>\n",
       "      <td>meriones_unguiculatus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>10047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>GCA_001649575.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>ASM164957v1</td>\n",
       "      <td>mangrove rivulus</td>\n",
       "      <td>Mangrove rivulus</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, core, rnaseq]</td>\n",
       "      <td>kryptolebias_marmoratus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>37003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>GCA_000001905.1</td>\n",
       "      <td>[9785, elephant, african savanna elephant, lox...</td>\n",
       "      <td>loxAfr3</td>\n",
       "      <td>African savanna elephant</td>\n",
       "      <td>Elephant</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core]</td>\n",
       "      <td>loxodonta_africana</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>GCA_000276665.1</td>\n",
       "      <td>[chinchilla lanigera, chilan, clanigera, clan,...</td>\n",
       "      <td>ChiLan1.0</td>\n",
       "      <td>Long-tailed chinchilla</td>\n",
       "      <td>Long-tailed chinchilla</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, rnaseq, core]</td>\n",
       "      <td>chinchilla_lanigera</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>34839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>GCA_001522545.2</td>\n",
       "      <td>[]</td>\n",
       "      <td>Parus_major1.1</td>\n",
       "      <td>Great Tit</td>\n",
       "      <td>Great Tit</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, core, rnaseq]</td>\n",
       "      <td>parus_major</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>GCA_001443285.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>P_latipinna-1.0</td>\n",
       "      <td>sailfin molly</td>\n",
       "      <td>Sailfin molly</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, rnaseq, core]</td>\n",
       "      <td>poecilia_latipinna</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>48699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>GCA_002234695.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>ASM223469v1</td>\n",
       "      <td>Japanese medaka HSOK</td>\n",
       "      <td>Japanese medaka HSOK</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[rnaseq, core]</td>\n",
       "      <td>oryzias_latipes_hsok</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>8090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>GCA_002263795.2</td>\n",
       "      <td>[bostau, bovine, btaurus, 9913, cow, cattle, b...</td>\n",
       "      <td>ARS-UCD1.2</td>\n",
       "      <td>cattle</td>\n",
       "      <td>Cow</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, rnaseq, funcgen, otherfeatures, variation]</td>\n",
       "      <td>bos_taurus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>None</td>\n",
       "      <td>[choloepus, choffmanni, chohof, hoffmann's two...</td>\n",
       "      <td>choHof1</td>\n",
       "      <td>Hoffmann's two-fingered sloth</td>\n",
       "      <td>Sloth</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core]</td>\n",
       "      <td>choloepus_hoffmanni</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>GCA_000687225.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>UrsMar_1.0</td>\n",
       "      <td>Polar bear</td>\n",
       "      <td>Polar bear</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, rnaseq, otherfeatures]</td>\n",
       "      <td>ursus_maritimus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>29073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>GCA_001952655.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>M_albus_1.0</td>\n",
       "      <td>swamp eel</td>\n",
       "      <td>Swamp eel</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, core, rnaseq]</td>\n",
       "      <td>monopterus_albus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>43700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>GCA_001984765.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>C.can_genome_v1.0</td>\n",
       "      <td>American beaver</td>\n",
       "      <td>American beaver</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[rnaseq, core, otherfeatures]</td>\n",
       "      <td>castor_canadensis</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>51338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>GCA_000372685.2</td>\n",
       "      <td>[7994, amexicanus, astmex, amex, cave fish, as...</td>\n",
       "      <td>Astyanax_mexicanus-2.0</td>\n",
       "      <td>Mexican tetra</td>\n",
       "      <td>Mexican tetra</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, core, rnaseq]</td>\n",
       "      <td>astyanax_mexicanus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>7994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>GCA_002109545.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>ASM210954v1</td>\n",
       "      <td>spiny chromis</td>\n",
       "      <td>Spiny chromis</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, rnaseq, otherfeatures]</td>\n",
       "      <td>acanthochromis_polyacanthus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>80966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>GCA_000743615.1</td>\n",
       "      <td>[fdamarensi, fdam, fukomys_damarensi, 885580, ...</td>\n",
       "      <td>DMR_v1.0</td>\n",
       "      <td>Damara mole rat</td>\n",
       "      <td>Damara mole rat</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, rnaseq, otherfeatures]</td>\n",
       "      <td>fukomys_damarensis</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>885580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>GCA_002872115.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>PKINGS_0.1</td>\n",
       "      <td>Paramormyrops kingsleyae</td>\n",
       "      <td>Paramormyrops kingsleyae</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, rnaseq]</td>\n",
       "      <td>paramormyrops_kingsleyae</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1676925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>GCA_000754665.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>Bison_UMD1.0</td>\n",
       "      <td>American bison</td>\n",
       "      <td>American bison</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, otherfeatures]</td>\n",
       "      <td>bison_bison_bison</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>43346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>GCA_000239395.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>NeoBri1.0</td>\n",
       "      <td>lyretail cichlid</td>\n",
       "      <td>Lyretail cichlid</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, rnaseq, otherfeatures]</td>\n",
       "      <td>neolamprologus_brichardi</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>32507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>GCA_002863925.1</td>\n",
       "      <td>[ecaballus, ecab, equcab, 9796, equus caballus...</td>\n",
       "      <td>EquCab3.0</td>\n",
       "      <td>horse</td>\n",
       "      <td>Horse</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, variation, core, funcgen, rnaseq]</td>\n",
       "      <td>equus_caballus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>GCA_001632525.1</td>\n",
       "      <td>[]</td>\n",
       "      <td>BALB_cJ_v1</td>\n",
       "      <td>house mouse</td>\n",
       "      <td>Mouse BALB/cJ</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[core, funcgen]</td>\n",
       "      <td>mus_musculus_balbcj</td>\n",
       "      <td>96</td>\n",
       "      <td>BALB/cJ</td>\n",
       "      <td>mouse</td>\n",
       "      <td>10090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>GCA_000349665.1</td>\n",
       "      <td>[10036, mesocricetus auratus, mesaur, maur, go...</td>\n",
       "      <td>MesAur1.0</td>\n",
       "      <td>Golden Hamster</td>\n",
       "      <td>Golden Hamster</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, core, funcgen]</td>\n",
       "      <td>mesocricetus_auratus</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>10036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>GCA_000180615.2</td>\n",
       "      <td>[takifugu, trubripes, takifugu rubripes, t_rub...</td>\n",
       "      <td>FUGU5</td>\n",
       "      <td>torafugu</td>\n",
       "      <td>Fugu</td>\n",
       "      <td>EnsemblVertebrates</td>\n",
       "      <td>[otherfeatures, core, rnaseq]</td>\n",
       "      <td>takifugu_rubripes</td>\n",
       "      <td>96</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>31033</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>199 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            accession                                            aliases  \\\n",
       "0     GCA_000409795.2  [chlorocebus sabeus, chlorocebus_sabeus, chlsa...   \n",
       "1     GCA_002197715.1                                                 []   \n",
       "2     GCA_000951045.1                                                 []   \n",
       "3     GCA_002776525.2                                                 []   \n",
       "4     GCA_001604975.1                                                 []   \n",
       "5     GCA_003255815.1                                                 []   \n",
       "6     GCA_003597395.1                                                 []   \n",
       "7     GCA_000464555.1                                            [tiger]   \n",
       "8     GCA_000001635.8  [mmus, mus musculus, 10090, musmus, m_musculus...   \n",
       "9     GCA_001443325.1                                                 []   \n",
       "10    GCA_000146795.3  [nleu, gibbon, nomleu, 61853, nleucogenys, nom...   \n",
       "11    GCA_000003025.6  [9823, wild boar, pigs, pig, swine, sscr, sscr...   \n",
       "12    GCA_002406435.1                                                 []   \n",
       "13    GCA_000298735.1  [ovisaries, ovis aries, oaries, oar, sheep, 99...   \n",
       "14    GCA_000688575.1  [cavia aperea, cape, 37548, brazilian guinea p...   \n",
       "15               None  [sea squirt ciona savignyi, 51511, csav, c.sav...   \n",
       "16    GCA_001624505.1                                                 []   \n",
       "17               None  [large flying fox, pvampyrus, 132908, pteropus...   \n",
       "18    GCA_000956105.1                                                 []   \n",
       "19    GCA_003033725.1                                      [ass, donkey]   \n",
       "20    GCA_000001215.4                     [dmel, d.mel, d. melanogaster]   \n",
       "21    GCA_000164805.2  [csyr, tarsius_syrichta, tarsius syrichta, tsy...   \n",
       "22    GCA_000247695.1  [naked mole-rat female, heterocephalus glaber ...   \n",
       "23    GCA_000751415.1                                                 []   \n",
       "24    GCA_003258685.1                                                 []   \n",
       "25    GCA_000230445.1  [hgla, heterocephalus glaber male, hetgla, hgl...   \n",
       "26    GCA_001715985.2                                                 []   \n",
       "27   GCA_000001405.27  [9606, h_sapiens, homsap, hsapiens, homo, huma...   \n",
       "28    GCA_000534875.1                                                 []   \n",
       "29    GCA_001458135.1                                                 []   \n",
       "..                ...                                                ...   \n",
       "169   GCA_000956065.1                                                 []   \n",
       "170   GCA_000151885.2  [ords kangaroo rat, dipodomysordii, kangaroo_r...   \n",
       "171              None  [7757, pmarinus, sea lamprey, petromyzon marin...   \n",
       "172   GCA_001624265.1                                                 []   \n",
       "173   GCA_900324485.1                                                 []   \n",
       "174   GCA_003697955.1                                                 []   \n",
       "175   GCA_000146605.1  [mgal, common turkey, 9103, mga, mgallopavo, m...   \n",
       "176   GCA_001624475.1                                                 []   \n",
       "177   GCA_002204375.1                                                 []   \n",
       "178   GCA_001649575.1                                                 []   \n",
       "179   GCA_000001905.1  [9785, elephant, african savanna elephant, lox...   \n",
       "180   GCA_000276665.1  [chinchilla lanigera, chilan, clanigera, clan,...   \n",
       "181   GCA_001522545.2                                                 []   \n",
       "182   GCA_001443285.1                                                 []   \n",
       "183   GCA_002234695.1                                                 []   \n",
       "184   GCA_002263795.2  [bostau, bovine, btaurus, 9913, cow, cattle, b...   \n",
       "185              None  [choloepus, choffmanni, chohof, hoffmann's two...   \n",
       "186   GCA_000687225.1                                                 []   \n",
       "187   GCA_001952655.1                                                 []   \n",
       "188   GCA_001984765.1                                                 []   \n",
       "189   GCA_000372685.2  [7994, amexicanus, astmex, amex, cave fish, as...   \n",
       "190   GCA_002109545.1                                                 []   \n",
       "191   GCA_000743615.1  [fdamarensi, fdam, fukomys_damarensi, 885580, ...   \n",
       "192   GCA_002872115.1                                                 []   \n",
       "193   GCA_000754665.1                                                 []   \n",
       "194   GCA_000239395.1                                                 []   \n",
       "195   GCA_002863925.1  [ecaballus, ecab, equcab, 9796, equus caballus...   \n",
       "196   GCA_001632525.1                                                 []   \n",
       "197   GCA_000349665.1  [10036, mesocricetus auratus, mesaur, maur, go...   \n",
       "198   GCA_000180615.2  [takifugu, trubripes, takifugu rubripes, t_rub...   \n",
       "\n",
       "                   assembly                     common_name  \\\n",
       "0                 ChlSab1.1            African green monkey   \n",
       "1                LonStrDom1                 Bengalese finch   \n",
       "2               Mleu.le_1.0                           Drill   \n",
       "3               ASM277652v2             Ugandan red Colobus   \n",
       "4        Cebus_imitator-1.0           White-headed capuchin   \n",
       "5                  Tgel_1.0                          gelada   \n",
       "6               ASM359739v1  Abingdon island giant tortoise   \n",
       "7                 PanTig1.0                           Tiger   \n",
       "8                    GRCm38                     house mouse   \n",
       "9            P_mexicana-1.0                  shortfin molly   \n",
       "10                 Nleu_3.0   Northern white-cheeked gibbon   \n",
       "11              Sscrofa11.1                             pig   \n",
       "12              ASM240643v1         Daurian ground squirrel   \n",
       "13                 Oar_v3.1                  Domestic sheep   \n",
       "14                 CavAp1.0            Brazilian guinea pig   \n",
       "15                  CSAV2.0       Sea squirt Ciona savignyi   \n",
       "16                DBA_2J_v1                     house mouse   \n",
       "17                  pteVam1                large flying fox   \n",
       "18                 Pcoq_1.0               Coquerel's sifaka   \n",
       "19              ASM303372v1                          donkey   \n",
       "20                 BDGP6.22                       fruit fly   \n",
       "21   Tarsius_syrichta-2.0.1              Philippine tarsier   \n",
       "22        HetGla_female_1.0                  naked mole-rat   \n",
       "23                 Midas_v5                   Midas cichlid   \n",
       "24               Prosim_1.0            greater bamboo lemur   \n",
       "25               HetGla_1.0                  naked mole-rat   \n",
       "26              ASM171598v2         golden-collared manakin   \n",
       "27                   GRCh38                           human   \n",
       "28                     SCA1                   common canary   \n",
       "29                marMar2.1                   Alpine marmot   \n",
       "..                      ...                             ...   \n",
       "169                Mnem_1.0              Pig-tailed macaque   \n",
       "170                Dord_2.0              Ord's kangaroo rat   \n",
       "171            Pmarinus_7.0                     sea lamprey   \n",
       "172             ASM162426v1                Asian bonytongue   \n",
       "173              fMasArm1.1                     zig-zag eel   \n",
       "174             ASM369795v1          Spoon-billed sandpiper   \n",
       "175             Turkey_2.01                 domestic turkey   \n",
       "176                CBA_J_v1                     house mouse   \n",
       "177           MunDraft-v1.0                Mongolian gerbil   \n",
       "178             ASM164957v1                mangrove rivulus   \n",
       "179                 loxAfr3        African savanna elephant   \n",
       "180               ChiLan1.0          Long-tailed chinchilla   \n",
       "181          Parus_major1.1                       Great Tit   \n",
       "182         P_latipinna-1.0                   sailfin molly   \n",
       "183             ASM223469v1            Japanese medaka HSOK   \n",
       "184              ARS-UCD1.2                          cattle   \n",
       "185                 choHof1   Hoffmann's two-fingered sloth   \n",
       "186              UrsMar_1.0                      Polar bear   \n",
       "187             M_albus_1.0                       swamp eel   \n",
       "188       C.can_genome_v1.0                 American beaver   \n",
       "189  Astyanax_mexicanus-2.0                   Mexican tetra   \n",
       "190             ASM210954v1                   spiny chromis   \n",
       "191                DMR_v1.0                 Damara mole rat   \n",
       "192              PKINGS_0.1        Paramormyrops kingsleyae   \n",
       "193            Bison_UMD1.0                  American bison   \n",
       "194               NeoBri1.0                lyretail cichlid   \n",
       "195               EquCab3.0                           horse   \n",
       "196              BALB_cJ_v1                     house mouse   \n",
       "197               MesAur1.0                  Golden Hamster   \n",
       "198                   FUGU5                        torafugu   \n",
       "\n",
       "                       display_name            division  \\\n",
       "0                        Vervet-AGM  EnsemblVertebrates   \n",
       "1                   Bengalese finch  EnsemblVertebrates   \n",
       "2                             Drill  EnsemblVertebrates   \n",
       "3               Ugandan red Colobus  EnsemblVertebrates   \n",
       "4                          Capuchin  EnsemblVertebrates   \n",
       "5                            Gelada  EnsemblVertebrates   \n",
       "6    Abingdon island giant tortoise  EnsemblVertebrates   \n",
       "7                             Tiger  EnsemblVertebrates   \n",
       "8                             Mouse  EnsemblVertebrates   \n",
       "9                    Shortfin molly  EnsemblVertebrates   \n",
       "10                           Gibbon  EnsemblVertebrates   \n",
       "11                              Pig  EnsemblVertebrates   \n",
       "12          Daurian ground squirrel  EnsemblVertebrates   \n",
       "13                            Sheep  EnsemblVertebrates   \n",
       "14             Brazilian guinea pig  EnsemblVertebrates   \n",
       "15                       C.savignyi  EnsemblVertebrates   \n",
       "16                     Mouse DBA/2J  EnsemblVertebrates   \n",
       "17                          Megabat  EnsemblVertebrates   \n",
       "18                Coquerel's sifaka  EnsemblVertebrates   \n",
       "19                           Donkey  EnsemblVertebrates   \n",
       "20          Drosophila melanogaster  EnsemblVertebrates   \n",
       "21                          Tarsier  EnsemblVertebrates   \n",
       "22            Naked mole-rat female  EnsemblVertebrates   \n",
       "23                    Midas cichlid  EnsemblVertebrates   \n",
       "24             Greater bamboo lemur  EnsemblVertebrates   \n",
       "25              Naked mole-rat male  EnsemblVertebrates   \n",
       "26          Golden-collared manakin  EnsemblVertebrates   \n",
       "27                            Human  EnsemblVertebrates   \n",
       "28                    Common canary  EnsemblVertebrates   \n",
       "29                    Alpine marmot  EnsemblVertebrates   \n",
       "..                              ...                 ...   \n",
       "169              Pig-tailed macaque  EnsemblVertebrates   \n",
       "170                    Kangaroo rat  EnsemblVertebrates   \n",
       "171                         Lamprey  EnsemblVertebrates   \n",
       "172                Asian bonytongue  EnsemblVertebrates   \n",
       "173                     Zig-zag eel  EnsemblVertebrates   \n",
       "174          Spoon-billed sandpiper  EnsemblVertebrates   \n",
       "175                          Turkey  EnsemblVertebrates   \n",
       "176                     Mouse CBA/J  EnsemblVertebrates   \n",
       "177                Mongolian gerbil  EnsemblVertebrates   \n",
       "178                Mangrove rivulus  EnsemblVertebrates   \n",
       "179                        Elephant  EnsemblVertebrates   \n",
       "180          Long-tailed chinchilla  EnsemblVertebrates   \n",
       "181                       Great Tit  EnsemblVertebrates   \n",
       "182                   Sailfin molly  EnsemblVertebrates   \n",
       "183            Japanese medaka HSOK  EnsemblVertebrates   \n",
       "184                             Cow  EnsemblVertebrates   \n",
       "185                           Sloth  EnsemblVertebrates   \n",
       "186                      Polar bear  EnsemblVertebrates   \n",
       "187                       Swamp eel  EnsemblVertebrates   \n",
       "188                 American beaver  EnsemblVertebrates   \n",
       "189                   Mexican tetra  EnsemblVertebrates   \n",
       "190                   Spiny chromis  EnsemblVertebrates   \n",
       "191                 Damara mole rat  EnsemblVertebrates   \n",
       "192        Paramormyrops kingsleyae  EnsemblVertebrates   \n",
       "193                  American bison  EnsemblVertebrates   \n",
       "194                Lyretail cichlid  EnsemblVertebrates   \n",
       "195                           Horse  EnsemblVertebrates   \n",
       "196                   Mouse BALB/cJ  EnsemblVertebrates   \n",
       "197                  Golden Hamster  EnsemblVertebrates   \n",
       "198                            Fugu  EnsemblVertebrates   \n",
       "\n",
       "                                                groups  \\\n",
       "0             [rnaseq, core, variation, otherfeatures]   \n",
       "1                        [otherfeatures, core, rnaseq]   \n",
       "2                       [funcgen, core, otherfeatures]   \n",
       "3                              [funcgen, rnaseq, core]   \n",
       "4               [core, funcgen, rnaseq, otherfeatures]   \n",
       "5                              [core, rnaseq, funcgen]   \n",
       "6                                       [core, rnaseq]   \n",
       "7                        [core, rnaseq, otherfeatures]   \n",
       "8    [core, cdna, rnaseq, funcgen, otherfeatures, v...   \n",
       "9                        [otherfeatures, rnaseq, core]   \n",
       "10           [funcgen, core, variation, otherfeatures]   \n",
       "11   [core, funcgen, rnaseq, otherfeatures, variation]   \n",
       "12                                              [core]   \n",
       "13            [variation, otherfeatures, rnaseq, core]   \n",
       "14                                              [core]   \n",
       "15                               [otherfeatures, core]   \n",
       "16                                     [core, funcgen]   \n",
       "17                                              [core]   \n",
       "18                      [core, funcgen, otherfeatures]   \n",
       "19                                      [core, rnaseq]   \n",
       "20                                     [funcgen, core]   \n",
       "21                      [core, funcgen, otherfeatures]   \n",
       "22                       [otherfeatures, rnaseq, core]   \n",
       "23                                              [core]   \n",
       "24                                     [funcgen, core]   \n",
       "25                       [otherfeatures, core, rnaseq]   \n",
       "26                       [otherfeatures, rnaseq, core]   \n",
       "27   [otherfeatures, variation, cdna, core, funcgen...   \n",
       "28                       [otherfeatures, core, rnaseq]   \n",
       "29                       [otherfeatures, core, rnaseq]   \n",
       "..                                                 ...   \n",
       "169             [otherfeatures, rnaseq, funcgen, core]   \n",
       "170                              [core, otherfeatures]   \n",
       "171                              [otherfeatures, core]   \n",
       "172                      [otherfeatures, core, rnaseq]   \n",
       "173                                     [rnaseq, core]   \n",
       "174                                     [rnaseq, core]   \n",
       "175                   [core, otherfeatures, variation]   \n",
       "176                                    [funcgen, core]   \n",
       "177                      [otherfeatures, core, rnaseq]   \n",
       "178                      [otherfeatures, core, rnaseq]   \n",
       "179                                             [core]   \n",
       "180                      [otherfeatures, rnaseq, core]   \n",
       "181                      [otherfeatures, core, rnaseq]   \n",
       "182                      [otherfeatures, rnaseq, core]   \n",
       "183                                     [rnaseq, core]   \n",
       "184  [core, rnaseq, funcgen, otherfeatures, variation]   \n",
       "185                                             [core]   \n",
       "186                      [core, rnaseq, otherfeatures]   \n",
       "187                      [otherfeatures, core, rnaseq]   \n",
       "188                      [rnaseq, core, otherfeatures]   \n",
       "189                      [otherfeatures, core, rnaseq]   \n",
       "190                      [core, rnaseq, otherfeatures]   \n",
       "191                      [core, rnaseq, otherfeatures]   \n",
       "192                                     [core, rnaseq]   \n",
       "193                              [core, otherfeatures]   \n",
       "194                      [core, rnaseq, otherfeatures]   \n",
       "195  [otherfeatures, variation, core, funcgen, rnaseq]   \n",
       "196                                    [core, funcgen]   \n",
       "197                     [otherfeatures, core, funcgen]   \n",
       "198                      [otherfeatures, core, rnaseq]   \n",
       "\n",
       "                             name  release               strain  \\\n",
       "0             chlorocebus_sabaeus       96                 None   \n",
       "1      lonchura_striata_domestica       96                 None   \n",
       "2          mandrillus_leucophaeus       96                 None   \n",
       "3       piliocolobus_tephrosceles       96                 None   \n",
       "4                 cebus_capucinus       96                 None   \n",
       "5            theropithecus_gelada       96                 None   \n",
       "6          chelonoidis_abingdonii       96                 None   \n",
       "7         panthera_tigris_altaica       96                 None   \n",
       "8                    mus_musculus       96  reference (CL57BL6)   \n",
       "9               poecilia_mexicana       96                 None   \n",
       "10            nomascus_leucogenys       96                 None   \n",
       "11                     sus_scrofa       96                 None   \n",
       "12          spermophilus_dauricus       96                 None   \n",
       "13                     ovis_aries       96                 None   \n",
       "14                   cavia_aperea       96                 None   \n",
       "15                 ciona_savignyi       96                 None   \n",
       "16             mus_musculus_dba2j       96               DBA/2J   \n",
       "17              pteropus_vampyrus       96                 None   \n",
       "18          propithecus_coquereli       96                 None   \n",
       "19            equus_asinus_asinus       96                 None   \n",
       "20        drosophila_melanogaster       96                 None   \n",
       "21               carlito_syrichta       96                 None   \n",
       "22   heterocephalus_glaber_female       96                 None   \n",
       "23        amphilophus_citrinellus       96                 None   \n",
       "24                 prolemur_simus       96                 None   \n",
       "25     heterocephalus_glaber_male       96                 None   \n",
       "26             manacus_vitellinus       96                 None   \n",
       "27                   homo_sapiens       96                 None   \n",
       "28                serinus_canaria       96                 None   \n",
       "29        marmota_marmota_marmota       96                 None   \n",
       "..                            ...      ...                  ...   \n",
       "169             macaca_nemestrina       96                 None   \n",
       "170               dipodomys_ordii       96                 None   \n",
       "171            petromyzon_marinus       96                 None   \n",
       "172          scleropages_formosus       96                 None   \n",
       "173         mastacembelus_armatus       96                 None   \n",
       "174              calidris_pygmaea       96                 None   \n",
       "175           meleagris_gallopavo       96                 None   \n",
       "176             mus_musculus_cbaj       96                CBA/J   \n",
       "177         meriones_unguiculatus       96                 None   \n",
       "178       kryptolebias_marmoratus       96                 None   \n",
       "179            loxodonta_africana       96                 None   \n",
       "180           chinchilla_lanigera       96                 None   \n",
       "181                   parus_major       96                 None   \n",
       "182            poecilia_latipinna       96                 None   \n",
       "183          oryzias_latipes_hsok       96                 None   \n",
       "184                    bos_taurus       96                 None   \n",
       "185           choloepus_hoffmanni       96                 None   \n",
       "186               ursus_maritimus       96                 None   \n",
       "187              monopterus_albus       96                 None   \n",
       "188             castor_canadensis       96                 None   \n",
       "189            astyanax_mexicanus       96                 None   \n",
       "190   acanthochromis_polyacanthus       96                 None   \n",
       "191            fukomys_damarensis       96                 None   \n",
       "192      paramormyrops_kingsleyae       96                 None   \n",
       "193             bison_bison_bison       96                 None   \n",
       "194      neolamprologus_brichardi       96                 None   \n",
       "195                equus_caballus       96                 None   \n",
       "196           mus_musculus_balbcj       96              BALB/cJ   \n",
       "197          mesocricetus_auratus       96                 None   \n",
       "198             takifugu_rubripes       96                 None   \n",
       "\n",
       "    strain_collection taxon_id  \n",
       "0                None    60711  \n",
       "1                None   299123  \n",
       "2                None     9568  \n",
       "3                None   591936  \n",
       "4                None  1737458  \n",
       "5                None     9565  \n",
       "6                None   106734  \n",
       "7                None    74533  \n",
       "8               mouse    10090  \n",
       "9                None    48701  \n",
       "10               None    61853  \n",
       "11               None     9823  \n",
       "12               None    99837  \n",
       "13               None     9940  \n",
       "14               None    37548  \n",
       "15               None    51511  \n",
       "16              mouse    10090  \n",
       "17               None   132908  \n",
       "18               None   379532  \n",
       "19               None    83772  \n",
       "20               None     7227  \n",
       "21               None  1868482  \n",
       "22               None    10181  \n",
       "23               None    61819  \n",
       "24               None  1328070  \n",
       "25               None    10181  \n",
       "26               None   328815  \n",
       "27               None     9606  \n",
       "28               None     9135  \n",
       "29               None     9994  \n",
       "..                ...      ...  \n",
       "169              None     9545  \n",
       "170              None    10020  \n",
       "171              None     7757  \n",
       "172              None   113540  \n",
       "173              None   205130  \n",
       "174              None   425635  \n",
       "175              None     9103  \n",
       "176             mouse    10090  \n",
       "177              None    10047  \n",
       "178              None    37003  \n",
       "179              None     9785  \n",
       "180              None    34839  \n",
       "181              None     9157  \n",
       "182              None    48699  \n",
       "183              None     8090  \n",
       "184              None     9913  \n",
       "185              None     9358  \n",
       "186              None    29073  \n",
       "187              None    43700  \n",
       "188              None    51338  \n",
       "189              None     7994  \n",
       "190              None    80966  \n",
       "191              None   885580  \n",
       "192              None  1676925  \n",
       "193              None    43346  \n",
       "194              None    32507  \n",
       "195              None     9796  \n",
       "196             mouse    10090  \n",
       "197              None    10036  \n",
       "198              None    31033  \n",
       "\n",
       "[199 rows x 12 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame.from_dict(decoded[\"species\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list = pd.DataFrame.from_dict(decoded[\"species\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on StringMethods in module pandas.core.strings object:\n",
      "\n",
      "class StringMethods(pandas.core.base.NoNewAttributesMixin)\n",
      " |  Vectorized string functions for Series and Index. NAs stay NA unless\n",
      " |  handled otherwise by a particular method. Patterned after Python's string\n",
      " |  methods, with some inspiration from R's stringr package.\n",
      " |  \n",
      " |  Examples\n",
      " |  --------\n",
      " |  >>> s.str.split('_')\n",
      " |  >>> s.str.replace('_', '')\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      StringMethods\n",
      " |      pandas.core.base.NoNewAttributesMixin\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __getitem__(self, key)\n",
      " |  \n",
      " |  __init__(self, data)\n",
      " |      Initialize self.  See help(type(self)) for accurate signature.\n",
      " |  \n",
      " |  __iter__(self)\n",
      " |  \n",
      " |  capitalize lambda self\n",
      " |      Convert strings in the Series/Index to be capitalized.\n",
      " |      \n",
      " |      Equivalent to :meth:`str.capitalize`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series/Index of objects\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.lower : Converts all characters to lowercase.\n",
      " |      Series.str.upper : Converts all characters to uppercase.\n",
      " |      Series.str.title : Converts first character of each word to uppercase and\n",
      " |          remaining to lowercase.\n",
      " |      Series.str.capitalize : Converts first character to uppercase and\n",
      " |          remaining to lowercase.\n",
      " |      Series.str.swapcase : Converts uppercase to lowercase and lowercase to\n",
      " |          uppercase.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series(['lower', 'CAPITALS', 'this is a sentence', 'SwApCaSe'])\n",
      " |      >>> s\n",
      " |      0                 lower\n",
      " |      1              CAPITALS\n",
      " |      2    this is a sentence\n",
      " |      3              SwApCaSe\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.lower()\n",
      " |      0                 lower\n",
      " |      1              capitals\n",
      " |      2    this is a sentence\n",
      " |      3              swapcase\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.upper()\n",
      " |      0                 LOWER\n",
      " |      1              CAPITALS\n",
      " |      2    THIS IS A SENTENCE\n",
      " |      3              SWAPCASE\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.title()\n",
      " |      0                 Lower\n",
      " |      1              Capitals\n",
      " |      2    This Is A Sentence\n",
      " |      3              Swapcase\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.capitalize()\n",
      " |      0                 Lower\n",
      " |      1              Capitals\n",
      " |      2    This is a sentence\n",
      " |      3              Swapcase\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.swapcase()\n",
      " |      0                 LOWER\n",
      " |      1              capitals\n",
      " |      2    THIS IS A SENTENCE\n",
      " |      3              sWaPcAsE\n",
      " |      dtype: object\n",
      " |  \n",
      " |  cat(self, others=None, sep=None, na_rep=None, join=None)\n",
      " |      Concatenate strings in the Series/Index with given separator.\n",
      " |      \n",
      " |      If `others` is specified, this function concatenates the Series/Index\n",
      " |      and elements of `others` element-wise.\n",
      " |      If `others` is not passed, then all values in the Series/Index are\n",
      " |      concatenated into a single string with a given `sep`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      others : Series, Index, DataFrame, np.ndarrary or list-like\n",
      " |          Series, Index, DataFrame, np.ndarray (one- or two-dimensional) and\n",
      " |          other list-likes of strings must have the same length as the\n",
      " |          calling Series/Index, with the exception of indexed objects (i.e.\n",
      " |          Series/Index/DataFrame) if `join` is not None.\n",
      " |      \n",
      " |          If others is a list-like that contains a combination of Series,\n",
      " |          Index or np.ndarray (1-dim), then all elements will be unpacked and\n",
      " |          must satisfy the above criteria individually.\n",
      " |      \n",
      " |          If others is None, the method returns the concatenation of all\n",
      " |          strings in the calling Series/Index.\n",
      " |      sep : str, default ''\n",
      " |          The separator between the different elements/columns. By default\n",
      " |          the empty string `''` is used.\n",
      " |      na_rep : str or None, default None\n",
      " |          Representation that is inserted for all missing values:\n",
      " |      \n",
      " |          - If `na_rep` is None, and `others` is None, missing values in the\n",
      " |            Series/Index are omitted from the result.\n",
      " |          - If `na_rep` is None, and `others` is not None, a row containing a\n",
      " |            missing value in any of the columns (before concatenation) will\n",
      " |            have a missing value in the result.\n",
      " |      join : {'left', 'right', 'outer', 'inner'}, default None\n",
      " |          Determines the join-style between the calling Series/Index and any\n",
      " |          Series/Index/DataFrame in `others` (objects without an index need\n",
      " |          to match the length of the calling Series/Index). If None,\n",
      " |          alignment is disabled, but this option will be removed in a future\n",
      " |          version of pandas and replaced with a default of `'left'`. To\n",
      " |          disable alignment, use `.values` on any Series/Index/DataFrame in\n",
      " |          `others`.\n",
      " |      \n",
      " |          .. versionadded:: 0.23.0\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      concat : str or Series/Index of objects\n",
      " |          If `others` is None, `str` is returned, otherwise a `Series/Index`\n",
      " |          (same type as caller) of objects is returned.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      split : Split each string in the Series/Index.\n",
      " |      join : Join lists contained as elements in the Series/Index.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      When not passing `others`, all values are concatenated into a single\n",
      " |      string:\n",
      " |      \n",
      " |      >>> s = pd.Series(['a', 'b', np.nan, 'd'])\n",
      " |      >>> s.str.cat(sep=' ')\n",
      " |      'a b d'\n",
      " |      \n",
      " |      By default, NA values in the Series are ignored. Using `na_rep`, they\n",
      " |      can be given a representation:\n",
      " |      \n",
      " |      >>> s.str.cat(sep=' ', na_rep='?')\n",
      " |      'a b ? d'\n",
      " |      \n",
      " |      If `others` is specified, corresponding values are concatenated with\n",
      " |      the separator. Result will be a Series of strings.\n",
      " |      \n",
      " |      >>> s.str.cat(['A', 'B', 'C', 'D'], sep=',')\n",
      " |      0    a,A\n",
      " |      1    b,B\n",
      " |      2    NaN\n",
      " |      3    d,D\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Missing values will remain missing in the result, but can again be\n",
      " |      represented using `na_rep`\n",
      " |      \n",
      " |      >>> s.str.cat(['A', 'B', 'C', 'D'], sep=',', na_rep='-')\n",
      " |      0    a,A\n",
      " |      1    b,B\n",
      " |      2    -,C\n",
      " |      3    d,D\n",
      " |      dtype: object\n",
      " |      \n",
      " |      If `sep` is not specified, the values are concatenated without\n",
      " |      separation.\n",
      " |      \n",
      " |      >>> s.str.cat(['A', 'B', 'C', 'D'], na_rep='-')\n",
      " |      0    aA\n",
      " |      1    bB\n",
      " |      2    -C\n",
      " |      3    dD\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Series with different indexes can be aligned before concatenation. The\n",
      " |      `join`-keyword works as in other methods.\n",
      " |      \n",
      " |      >>> t = pd.Series(['d', 'a', 'e', 'c'], index=[3, 0, 4, 2])\n",
      " |      >>> s.str.cat(t, join='left', na_rep='-')\n",
      " |      0    aa\n",
      " |      1    b-\n",
      " |      2    -c\n",
      " |      3    dd\n",
      " |      dtype: object\n",
      " |      >>>\n",
      " |      >>> s.str.cat(t, join='outer', na_rep='-')\n",
      " |      0    aa\n",
      " |      1    b-\n",
      " |      2    -c\n",
      " |      3    dd\n",
      " |      4    -e\n",
      " |      dtype: object\n",
      " |      >>>\n",
      " |      >>> s.str.cat(t, join='inner', na_rep='-')\n",
      " |      0    aa\n",
      " |      2    -c\n",
      " |      3    dd\n",
      " |      dtype: object\n",
      " |      >>>\n",
      " |      >>> s.str.cat(t, join='right', na_rep='-')\n",
      " |      3    dd\n",
      " |      0    aa\n",
      " |      4    -e\n",
      " |      2    -c\n",
      " |      dtype: object\n",
      " |      \n",
      " |      For more examples, see :ref:`here <text.concatenate>`.\n",
      " |  \n",
      " |  center(self, width, fillchar=' ')\n",
      " |      Filling left and right side of strings in the Series/Index with an\n",
      " |      additional character. Equivalent to :meth:`str.center`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      width : int\n",
      " |          Minimum width of resulting string; additional characters will be filled\n",
      " |          with ``fillchar``\n",
      " |      fillchar : str\n",
      " |          Additional character for filling, default is whitespace\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      filled : Series/Index of objects\n",
      " |  \n",
      " |  contains(self, pat, case=True, flags=0, na=nan, regex=True)\n",
      " |      Test if pattern or regex is contained within a string of a Series or Index.\n",
      " |      \n",
      " |      Return boolean Series or Index based on whether a given pattern or regex is\n",
      " |      contained within a string of a Series or Index.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      pat : str\n",
      " |          Character sequence or regular expression.\n",
      " |      case : bool, default True\n",
      " |          If True, case sensitive.\n",
      " |      flags : int, default 0 (no flags)\n",
      " |          Flags to pass through to the re module, e.g. re.IGNORECASE.\n",
      " |      na : default NaN\n",
      " |          Fill value for missing values.\n",
      " |      regex : bool, default True\n",
      " |          If True, assumes the pat is a regular expression.\n",
      " |      \n",
      " |          If False, treats the pat as a literal string.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series or Index of boolean values\n",
      " |          A Series or Index of boolean values indicating whether the\n",
      " |          given pattern is contained within the string of each element\n",
      " |          of the Series or Index.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      match : Analogous, but stricter, relying on re.match instead of re.search.\n",
      " |      Series.str.startswith : Test if the start of each string element matches a\n",
      " |          pattern.\n",
      " |      Series.str.endswith : Same as startswith, but tests the end of string.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      \n",
      " |      Returning a Series of booleans using only a literal pattern.\n",
      " |      \n",
      " |      >>> s1 = pd.Series(['Mouse', 'dog', 'house and parrot', '23', np.NaN])\n",
      " |      >>> s1.str.contains('og', regex=False)\n",
      " |      0    False\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      4      NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Returning an Index of booleans using only a literal pattern.\n",
      " |      \n",
      " |      >>> ind = pd.Index(['Mouse', 'dog', 'house and parrot', '23.0', np.NaN])\n",
      " |      >>> ind.str.contains('23', regex=False)\n",
      " |      Index([False, False, False, True, nan], dtype='object')\n",
      " |      \n",
      " |      Specifying case sensitivity using `case`.\n",
      " |      \n",
      " |      >>> s1.str.contains('oG', case=True, regex=True)\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      4      NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Specifying `na` to be `False` instead of `NaN` replaces NaN values\n",
      " |      with `False`. If Series or Index does not contain NaN values\n",
      " |      the resultant dtype will be `bool`, otherwise, an `object` dtype.\n",
      " |      \n",
      " |      >>> s1.str.contains('og', na=False, regex=True)\n",
      " |      0    False\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      4    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      Returning 'house' or 'dog' when either expression occurs in a string.\n",
      " |      \n",
      " |      >>> s1.str.contains('house|dog', regex=True)\n",
      " |      0    False\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      4      NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Ignoring case sensitivity using `flags` with regex.\n",
      " |      \n",
      " |      >>> import re\n",
      " |      >>> s1.str.contains('PARROT', flags=re.IGNORECASE, regex=True)\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      4      NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Returning any digit using regular expression.\n",
      " |      \n",
      " |      >>> s1.str.contains('\\d', regex=True)\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3     True\n",
      " |      4      NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Ensure `pat` is a not a literal pattern when `regex` is set to True.\n",
      " |      Note in the following example one might expect only `s2[1]` and `s2[3]` to\n",
      " |      return `True`. However, '.0' as a regex matches any character\n",
      " |      followed by a 0.\n",
      " |      \n",
      " |      >>> s2 = pd.Series(['40','40.0','41','41.0','35'])\n",
      " |      >>> s2.str.contains('.0', regex=True)\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3     True\n",
      " |      4    False\n",
      " |      dtype: bool\n",
      " |  \n",
      " |  count = str_count(self, pat, flags=0, **kwargs)\n",
      " |      Count occurrences of pattern in each string of the Series/Index.\n",
      " |      \n",
      " |      This function is used to count the number of times a particular regex\n",
      " |      pattern is repeated in each of the string elements of the\n",
      " |      :class:`~pandas.Series`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      pat : str\n",
      " |          Valid regular expression.\n",
      " |      flags : int, default 0, meaning no flags\n",
      " |          Flags for the `re` module. For a complete list, `see here\n",
      " |          <https://docs.python.org/3/howto/regex.html#compilation-flags>`_.\n",
      " |      **kwargs\n",
      " |          For compatibility with other string methods. Not used.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      counts : Series or Index\n",
      " |          Same type as the calling object containing the integer counts.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      re : Standard library module for regular expressions.\n",
      " |      str.count : Standard library version, without regular expression support.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      Some characters need to be escaped when passing in `pat`.\n",
      " |      eg. ``'$'`` has a special meaning in regex and must be escaped when\n",
      " |      finding this literal character.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series(['A', 'B', 'Aaba', 'Baca', np.nan, 'CABA', 'cat'])\n",
      " |      >>> s.str.count('a')\n",
      " |      0    0.0\n",
      " |      1    0.0\n",
      " |      2    2.0\n",
      " |      3    2.0\n",
      " |      4    NaN\n",
      " |      5    0.0\n",
      " |      6    1.0\n",
      " |      dtype: float64\n",
      " |      \n",
      " |      Escape ``'$'`` to find the literal dollar sign.\n",
      " |      \n",
      " |      >>> s = pd.Series(['$', 'B', 'Aab$', '$$ca', 'C$B$', 'cat'])\n",
      " |      >>> s.str.count('\\$')\n",
      " |      0    1\n",
      " |      1    0\n",
      " |      2    1\n",
      " |      3    2\n",
      " |      4    2\n",
      " |      5    0\n",
      " |      dtype: int64\n",
      " |      \n",
      " |      This is also available on Index\n",
      " |      \n",
      " |      >>> pd.Index(['A', 'A', 'Aaba', 'cat']).str.count('a')\n",
      " |      Int64Index([0, 0, 2, 1], dtype='int64')\n",
      " |  \n",
      " |  decode(self, encoding, errors='strict')\n",
      " |      Decode character string in the Series/Index using indicated encoding.\n",
      " |      Equivalent to :meth:`str.decode` in python2 and :meth:`bytes.decode` in\n",
      " |      python3.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      encoding : str\n",
      " |      errors : str, optional\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      decoded : Series/Index of objects\n",
      " |  \n",
      " |  encode(self, encoding, errors='strict')\n",
      " |      Encode character string in the Series/Index using indicated encoding.\n",
      " |      Equivalent to :meth:`str.encode`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      encoding : str\n",
      " |      errors : str, optional\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      encoded : Series/Index of objects\n",
      " |  \n",
      " |  endswith = str_endswith(self, pat, na=nan)\n",
      " |      Test if the end of each string element matches a pattern.\n",
      " |      \n",
      " |      Equivalent to :meth:`str.endswith`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      pat : str\n",
      " |          Character sequence. Regular expressions are not accepted.\n",
      " |      na : object, default NaN\n",
      " |          Object shown if element tested is not a string.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series or Index of bool\n",
      " |          A Series of booleans indicating whether the given pattern matches\n",
      " |          the end of each string element.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      str.endswith : Python standard library string method.\n",
      " |      Series.str.startswith : Same as endswith, but tests the start of string.\n",
      " |      Series.str.contains : Tests if string element contains a pattern.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series(['bat', 'bear', 'caT', np.nan])\n",
      " |      >>> s\n",
      " |      0     bat\n",
      " |      1    bear\n",
      " |      2     caT\n",
      " |      3     NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.endswith('t')\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3      NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Specifying `na` to be `False` instead of `NaN`.\n",
      " |      \n",
      " |      >>> s.str.endswith('t', na=False)\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |  \n",
      " |  extract(self, pat, flags=0, expand=True)\n",
      " |      Extract capture groups in the regex `pat` as columns in a DataFrame.\n",
      " |      \n",
      " |      For each subject string in the Series, extract groups from the\n",
      " |      first match of regular expression `pat`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      pat : string\n",
      " |          Regular expression pattern with capturing groups.\n",
      " |      flags : int, default 0 (no flags)\n",
      " |          Flags from the ``re`` module, e.g. ``re.IGNORECASE``, that\n",
      " |          modify regular expression matching for things like case,\n",
      " |          spaces, etc. For more details, see :mod:`re`.\n",
      " |      expand : bool, default True\n",
      " |          If True, return DataFrame with one column per capture group.\n",
      " |          If False, return a Series/Index if there is one capture group\n",
      " |          or DataFrame if there are multiple capture groups.\n",
      " |      \n",
      " |          .. versionadded:: 0.18.0\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      DataFrame or Series or Index\n",
      " |          A DataFrame with one row for each subject string, and one\n",
      " |          column for each group. Any capture group names in regular\n",
      " |          expression pat will be used for column names; otherwise\n",
      " |          capture group numbers will be used. The dtype of each result\n",
      " |          column is always object, even when no match is found. If\n",
      " |          ``expand=False`` and pat has only one capture group, then\n",
      " |          return a Series (if subject is a Series) or Index (if subject\n",
      " |          is an Index).\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      extractall : Returns all matches (not just the first match).\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      A pattern with two groups will return a DataFrame with two columns.\n",
      " |      Non-matches will be NaN.\n",
      " |      \n",
      " |      >>> s = pd.Series(['a1', 'b2', 'c3'])\n",
      " |      >>> s.str.extract(r'([ab])(\\d)')\n",
      " |           0    1\n",
      " |      0    a    1\n",
      " |      1    b    2\n",
      " |      2  NaN  NaN\n",
      " |      \n",
      " |      A pattern may contain optional groups.\n",
      " |      \n",
      " |      >>> s.str.extract(r'([ab])?(\\d)')\n",
      " |           0  1\n",
      " |      0    a  1\n",
      " |      1    b  2\n",
      " |      2  NaN  3\n",
      " |      \n",
      " |      Named groups will become column names in the result.\n",
      " |      \n",
      " |      >>> s.str.extract(r'(?P<letter>[ab])(?P<digit>\\d)')\n",
      " |        letter digit\n",
      " |      0      a     1\n",
      " |      1      b     2\n",
      " |      2    NaN   NaN\n",
      " |      \n",
      " |      A pattern with one group will return a DataFrame with one column\n",
      " |      if expand=True.\n",
      " |      \n",
      " |      >>> s.str.extract(r'[ab](\\d)', expand=True)\n",
      " |           0\n",
      " |      0    1\n",
      " |      1    2\n",
      " |      2  NaN\n",
      " |      \n",
      " |      A pattern with one group will return a Series if expand=False.\n",
      " |      \n",
      " |      >>> s.str.extract(r'[ab](\\d)', expand=False)\n",
      " |      0      1\n",
      " |      1      2\n",
      " |      2    NaN\n",
      " |      dtype: object\n",
      " |  \n",
      " |  extractall(self, pat, flags=0)\n",
      " |      For each subject string in the Series, extract groups from all\n",
      " |      matches of regular expression pat. When each subject string in the\n",
      " |      Series has exactly one match, extractall(pat).xs(0, level='match')\n",
      " |      is the same as extract(pat).\n",
      " |      \n",
      " |      .. versionadded:: 0.18.0\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      pat : str\n",
      " |          Regular expression pattern with capturing groups.\n",
      " |      flags : int, default 0 (no flags)\n",
      " |          A ``re`` module flag, for example ``re.IGNORECASE``. These allow\n",
      " |          to modify regular expression matching for things like case, spaces,\n",
      " |          etc. Multiple flags can be combined with the bitwise OR operator,\n",
      " |          for example ``re.IGNORECASE | re.MULTILINE``.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      DataFrame\n",
      " |          A ``DataFrame`` with one row for each match, and one column for each\n",
      " |          group. Its rows have a ``MultiIndex`` with first levels that come from\n",
      " |          the subject ``Series``. The last level is named 'match' and indexes the\n",
      " |          matches in each item of the ``Series``. Any capture group names in\n",
      " |          regular expression pat will be used for column names; otherwise capture\n",
      " |          group numbers will be used.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      extract : Returns first match only (not all matches).\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      A pattern with one group will return a DataFrame with one column.\n",
      " |      Indices with no matches will not appear in the result.\n",
      " |      \n",
      " |      >>> s = pd.Series([\"a1a2\", \"b1\", \"c1\"], index=[\"A\", \"B\", \"C\"])\n",
      " |      >>> s.str.extractall(r\"[ab](\\d)\")\n",
      " |               0\n",
      " |        match\n",
      " |      A 0      1\n",
      " |        1      2\n",
      " |      B 0      1\n",
      " |      \n",
      " |      Capture group names are used for column names of the result.\n",
      " |      \n",
      " |      >>> s.str.extractall(r\"[ab](?P<digit>\\d)\")\n",
      " |              digit\n",
      " |        match\n",
      " |      A 0         1\n",
      " |        1         2\n",
      " |      B 0         1\n",
      " |      \n",
      " |      A pattern with two groups will return a DataFrame with two columns.\n",
      " |      \n",
      " |      >>> s.str.extractall(r\"(?P<letter>[ab])(?P<digit>\\d)\")\n",
      " |              letter digit\n",
      " |        match\n",
      " |      A 0          a     1\n",
      " |        1          a     2\n",
      " |      B 0          b     1\n",
      " |      \n",
      " |      Optional groups that do not match are NaN in the result.\n",
      " |      \n",
      " |      >>> s.str.extractall(r\"(?P<letter>[ab])?(?P<digit>\\d)\")\n",
      " |              letter digit\n",
      " |        match\n",
      " |      A 0          a     1\n",
      " |        1          a     2\n",
      " |      B 0          b     1\n",
      " |      C 0        NaN     1\n",
      " |  \n",
      " |  find(self, sub, start=0, end=None)\n",
      " |      Return lowest indexes in each strings in the Series/Index\n",
      " |      where the substring is fully contained between [start:end].\n",
      " |      Return -1 on failure. Equivalent to standard :meth:`str.find`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      sub : str\n",
      " |          Substring being searched\n",
      " |      start : int\n",
      " |          Left edge index\n",
      " |      end : int\n",
      " |          Right edge index\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      found : Series/Index of integer values\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      rfind : Return highest indexes in each strings.\n",
      " |  \n",
      " |  findall = str_findall(self, pat, flags=0, **kwargs)\n",
      " |      Find all occurrences of pattern or regular expression in the Series/Index.\n",
      " |      \n",
      " |      Equivalent to applying :func:`re.findall` to all the elements in the\n",
      " |      Series/Index.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      pat : string\n",
      " |          Pattern or regular expression.\n",
      " |      flags : int, default 0\n",
      " |          ``re`` module flags, e.g. `re.IGNORECASE` (default is 0, which means\n",
      " |          no flags).\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series/Index of lists of strings\n",
      " |          All non-overlapping matches of pattern or regular expression in each\n",
      " |          string of this Series/Index.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      count : Count occurrences of pattern or regular expression in each string\n",
      " |          of the Series/Index.\n",
      " |      extractall : For each string in the Series, extract groups from all matches\n",
      " |          of regular expression and return a DataFrame with one row for each\n",
      " |          match and one column for each group.\n",
      " |      re.findall : The equivalent ``re`` function to all non-overlapping matches\n",
      " |          of pattern or regular expression in string, as a list of strings.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      \n",
      " |      >>> s = pd.Series(['Lion', 'Monkey', 'Rabbit'])\n",
      " |      \n",
      " |      The search for the pattern 'Monkey' returns one match:\n",
      " |      \n",
      " |      >>> s.str.findall('Monkey')\n",
      " |      0          []\n",
      " |      1    [Monkey]\n",
      " |      2          []\n",
      " |      dtype: object\n",
      " |      \n",
      " |      On the other hand, the search for the pattern 'MONKEY' doesn't return any\n",
      " |      match:\n",
      " |      \n",
      " |      >>> s.str.findall('MONKEY')\n",
      " |      0    []\n",
      " |      1    []\n",
      " |      2    []\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Flags can be added to the pattern or regular expression. For instance,\n",
      " |      to find the pattern 'MONKEY' ignoring the case:\n",
      " |      \n",
      " |      >>> import re\n",
      " |      >>> s.str.findall('MONKEY', flags=re.IGNORECASE)\n",
      " |      0          []\n",
      " |      1    [Monkey]\n",
      " |      2          []\n",
      " |      dtype: object\n",
      " |      \n",
      " |      When the pattern matches more than one string in the Series, all matches\n",
      " |      are returned:\n",
      " |      \n",
      " |      >>> s.str.findall('on')\n",
      " |      0    [on]\n",
      " |      1    [on]\n",
      " |      2      []\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Regular expressions are supported too. For instance, the search for all the\n",
      " |      strings ending with the word 'on' is shown next:\n",
      " |      \n",
      " |      >>> s.str.findall('on$')\n",
      " |      0    [on]\n",
      " |      1      []\n",
      " |      2      []\n",
      " |      dtype: object\n",
      " |      \n",
      " |      If the pattern is found more than once in the same string, then a list of\n",
      " |      multiple strings is returned:\n",
      " |      \n",
      " |      >>> s.str.findall('b')\n",
      " |      0        []\n",
      " |      1        []\n",
      " |      2    [b, b]\n",
      " |      dtype: object\n",
      " |  \n",
      " |  get(self, i)\n",
      " |      Extract element from each component at specified position.\n",
      " |      \n",
      " |      Extract element from lists, tuples, or strings in each element in the\n",
      " |      Series/Index.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      i : int\n",
      " |          Position of element to extract.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      items : Series/Index of objects\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series([\"String\",\n",
      " |                 (1, 2, 3),\n",
      " |                 [\"a\", \"b\", \"c\"],\n",
      " |                 123, -456,\n",
      " |                 {1:\"Hello\", \"2\":\"World\"}])\n",
      " |      >>> s\n",
      " |      0                        String\n",
      " |      1                     (1, 2, 3)\n",
      " |      2                     [a, b, c]\n",
      " |      3                           123\n",
      " |      4                          -456\n",
      " |      5    {1: 'Hello', '2': 'World'}\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.get(1)\n",
      " |      0        t\n",
      " |      1        2\n",
      " |      2        b\n",
      " |      3      NaN\n",
      " |      4      NaN\n",
      " |      5    Hello\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.get(-1)\n",
      " |      0      g\n",
      " |      1      3\n",
      " |      2      c\n",
      " |      3    NaN\n",
      " |      4    NaN\n",
      " |      5    NaN\n",
      " |      dtype: object\n",
      " |  \n",
      " |  get_dummies(self, sep='|')\n",
      " |      Split each string in the Series by sep and return a frame of\n",
      " |      dummy/indicator variables.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      sep : string, default \"|\"\n",
      " |          String to split on.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      dummies : DataFrame\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      get_dummies\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> pd.Series(['a|b', 'a', 'a|c']).str.get_dummies()\n",
      " |         a  b  c\n",
      " |      0  1  1  0\n",
      " |      1  1  0  0\n",
      " |      2  1  0  1\n",
      " |      \n",
      " |      >>> pd.Series(['a|b', np.nan, 'a|c']).str.get_dummies()\n",
      " |         a  b  c\n",
      " |      0  1  1  0\n",
      " |      1  0  0  0\n",
      " |      2  1  0  1\n",
      " |  \n",
      " |  index(self, sub, start=0, end=None)\n",
      " |      Return lowest indexes in each strings where the substring is\n",
      " |      fully contained between [start:end]. This is the same as\n",
      " |      ``str.find`` except instead of returning -1, it raises a ValueError\n",
      " |      when the substring is not found. Equivalent to standard ``str.index``.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      sub : str\n",
      " |          Substring being searched\n",
      " |      start : int\n",
      " |          Left edge index\n",
      " |      end : int\n",
      " |          Right edge index\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      found : Series/Index of objects\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      rindex : Return highest indexes in each strings.\n",
      " |  \n",
      " |  isalnum lambda self\n",
      " |      Check whether all characters in each string are alphanumeric.\n",
      " |      \n",
      " |      This is equivalent to running the Python string method\n",
      " |      :meth:`str.isalnum` for each element of the Series/Index. If a string\n",
      " |      has zero characters, ``False`` is returned for that check.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series or Index of bool\n",
      " |          Series or Index of boolean values with the same length as the original\n",
      " |          Series/Index.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.isalpha : Check whether all characters are alphabetic.\n",
      " |      Series.str.isnumeric : Check whether all characters are numeric.\n",
      " |      Series.str.isalnum : Check whether all characters are alphanumeric.\n",
      " |      Series.str.isdigit : Check whether all characters are digits.\n",
      " |      Series.str.isdecimal : Check whether all characters are decimal.\n",
      " |      Series.str.isspace : Check whether all characters are whitespace.\n",
      " |      Series.str.islower : Check whether all characters are lowercase.\n",
      " |      Series.str.isupper : Check whether all characters are uppercase.\n",
      " |      Series.str.istitle : Check whether all characters are titlecase.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      **Checks for Alphabetic and Numeric Characters**\n",
      " |      \n",
      " |      >>> s1 = pd.Series(['one', 'one1', '1', ''])\n",
      " |      \n",
      " |      >>> s1.str.isalpha()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s1.str.isnumeric()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s1.str.isalnum()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      Note that checks against characters mixed with any additional punctuation\n",
      " |      or whitespace will evaluate to false for an alphanumeric check.\n",
      " |      \n",
      " |      >>> s2 = pd.Series(['A B', '1.5', '3,000'])\n",
      " |      >>> s2.str.isalnum()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **More Detailed Checks for Numeric Characters**\n",
      " |      \n",
      " |      There are several different but overlapping sets of numeric characters that\n",
      " |      can be checked for.\n",
      " |      \n",
      " |      >>> s3 = pd.Series(['23', '³', '⅕', ''])\n",
      " |      \n",
      " |      The ``s3.str.isdecimal`` method checks for characters used to form numbers\n",
      " |      in base 10.\n",
      " |      \n",
      " |      >>> s3.str.isdecimal()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s.str.isdigit`` method is the same as ``s3.str.isdecimal`` but also\n",
      " |      includes special digits, like superscripted and subscripted digits in\n",
      " |      unicode.\n",
      " |      \n",
      " |      >>> s3.str.isdigit()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s.str.isnumeric`` method is the same as ``s3.str.isdigit`` but also\n",
      " |      includes other characters that can represent quantities such as unicode\n",
      " |      fractions.\n",
      " |      \n",
      " |      >>> s3.str.isnumeric()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **Checks for Whitespace**\n",
      " |      \n",
      " |      >>> s4 = pd.Series([' ', '\\t\\r\\n ', ''])\n",
      " |      >>> s4.str.isspace()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **Checks for Character Case**\n",
      " |      \n",
      " |      >>> s5 = pd.Series(['leopard', 'Golden Eagle', 'SNAKE', ''])\n",
      " |      \n",
      " |      >>> s5.str.islower()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s5.str.isupper()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s5.str.istitle`` method checks for whether all words are in title\n",
      " |      case (whether only the first letter of each word is capitalized). Words are\n",
      " |      assumed to be as any sequence of non-numeric characters seperated by\n",
      " |      whitespace characters.\n",
      " |      \n",
      " |      >>> s5.str.istitle()\n",
      " |      0    False\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |  \n",
      " |  isalpha lambda self\n",
      " |      Check whether all characters in each string are alphabetic.\n",
      " |      \n",
      " |      This is equivalent to running the Python string method\n",
      " |      :meth:`str.isalpha` for each element of the Series/Index. If a string\n",
      " |      has zero characters, ``False`` is returned for that check.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series or Index of bool\n",
      " |          Series or Index of boolean values with the same length as the original\n",
      " |          Series/Index.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.isalpha : Check whether all characters are alphabetic.\n",
      " |      Series.str.isnumeric : Check whether all characters are numeric.\n",
      " |      Series.str.isalnum : Check whether all characters are alphanumeric.\n",
      " |      Series.str.isdigit : Check whether all characters are digits.\n",
      " |      Series.str.isdecimal : Check whether all characters are decimal.\n",
      " |      Series.str.isspace : Check whether all characters are whitespace.\n",
      " |      Series.str.islower : Check whether all characters are lowercase.\n",
      " |      Series.str.isupper : Check whether all characters are uppercase.\n",
      " |      Series.str.istitle : Check whether all characters are titlecase.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      **Checks for Alphabetic and Numeric Characters**\n",
      " |      \n",
      " |      >>> s1 = pd.Series(['one', 'one1', '1', ''])\n",
      " |      \n",
      " |      >>> s1.str.isalpha()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s1.str.isnumeric()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s1.str.isalnum()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      Note that checks against characters mixed with any additional punctuation\n",
      " |      or whitespace will evaluate to false for an alphanumeric check.\n",
      " |      \n",
      " |      >>> s2 = pd.Series(['A B', '1.5', '3,000'])\n",
      " |      >>> s2.str.isalnum()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **More Detailed Checks for Numeric Characters**\n",
      " |      \n",
      " |      There are several different but overlapping sets of numeric characters that\n",
      " |      can be checked for.\n",
      " |      \n",
      " |      >>> s3 = pd.Series(['23', '³', '⅕', ''])\n",
      " |      \n",
      " |      The ``s3.str.isdecimal`` method checks for characters used to form numbers\n",
      " |      in base 10.\n",
      " |      \n",
      " |      >>> s3.str.isdecimal()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s.str.isdigit`` method is the same as ``s3.str.isdecimal`` but also\n",
      " |      includes special digits, like superscripted and subscripted digits in\n",
      " |      unicode.\n",
      " |      \n",
      " |      >>> s3.str.isdigit()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s.str.isnumeric`` method is the same as ``s3.str.isdigit`` but also\n",
      " |      includes other characters that can represent quantities such as unicode\n",
      " |      fractions.\n",
      " |      \n",
      " |      >>> s3.str.isnumeric()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **Checks for Whitespace**\n",
      " |      \n",
      " |      >>> s4 = pd.Series([' ', '\\t\\r\\n ', ''])\n",
      " |      >>> s4.str.isspace()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **Checks for Character Case**\n",
      " |      \n",
      " |      >>> s5 = pd.Series(['leopard', 'Golden Eagle', 'SNAKE', ''])\n",
      " |      \n",
      " |      >>> s5.str.islower()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s5.str.isupper()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s5.str.istitle`` method checks for whether all words are in title\n",
      " |      case (whether only the first letter of each word is capitalized). Words are\n",
      " |      assumed to be as any sequence of non-numeric characters seperated by\n",
      " |      whitespace characters.\n",
      " |      \n",
      " |      >>> s5.str.istitle()\n",
      " |      0    False\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |  \n",
      " |  isdecimal lambda self\n",
      " |      Check whether all characters in each string are decimal.\n",
      " |      \n",
      " |      This is equivalent to running the Python string method\n",
      " |      :meth:`str.isdecimal` for each element of the Series/Index. If a string\n",
      " |      has zero characters, ``False`` is returned for that check.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series or Index of bool\n",
      " |          Series or Index of boolean values with the same length as the original\n",
      " |          Series/Index.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.isalpha : Check whether all characters are alphabetic.\n",
      " |      Series.str.isnumeric : Check whether all characters are numeric.\n",
      " |      Series.str.isalnum : Check whether all characters are alphanumeric.\n",
      " |      Series.str.isdigit : Check whether all characters are digits.\n",
      " |      Series.str.isdecimal : Check whether all characters are decimal.\n",
      " |      Series.str.isspace : Check whether all characters are whitespace.\n",
      " |      Series.str.islower : Check whether all characters are lowercase.\n",
      " |      Series.str.isupper : Check whether all characters are uppercase.\n",
      " |      Series.str.istitle : Check whether all characters are titlecase.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      **Checks for Alphabetic and Numeric Characters**\n",
      " |      \n",
      " |      >>> s1 = pd.Series(['one', 'one1', '1', ''])\n",
      " |      \n",
      " |      >>> s1.str.isalpha()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s1.str.isnumeric()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s1.str.isalnum()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      Note that checks against characters mixed with any additional punctuation\n",
      " |      or whitespace will evaluate to false for an alphanumeric check.\n",
      " |      \n",
      " |      >>> s2 = pd.Series(['A B', '1.5', '3,000'])\n",
      " |      >>> s2.str.isalnum()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **More Detailed Checks for Numeric Characters**\n",
      " |      \n",
      " |      There are several different but overlapping sets of numeric characters that\n",
      " |      can be checked for.\n",
      " |      \n",
      " |      >>> s3 = pd.Series(['23', '³', '⅕', ''])\n",
      " |      \n",
      " |      The ``s3.str.isdecimal`` method checks for characters used to form numbers\n",
      " |      in base 10.\n",
      " |      \n",
      " |      >>> s3.str.isdecimal()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s.str.isdigit`` method is the same as ``s3.str.isdecimal`` but also\n",
      " |      includes special digits, like superscripted and subscripted digits in\n",
      " |      unicode.\n",
      " |      \n",
      " |      >>> s3.str.isdigit()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s.str.isnumeric`` method is the same as ``s3.str.isdigit`` but also\n",
      " |      includes other characters that can represent quantities such as unicode\n",
      " |      fractions.\n",
      " |      \n",
      " |      >>> s3.str.isnumeric()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **Checks for Whitespace**\n",
      " |      \n",
      " |      >>> s4 = pd.Series([' ', '\\t\\r\\n ', ''])\n",
      " |      >>> s4.str.isspace()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **Checks for Character Case**\n",
      " |      \n",
      " |      >>> s5 = pd.Series(['leopard', 'Golden Eagle', 'SNAKE', ''])\n",
      " |      \n",
      " |      >>> s5.str.islower()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s5.str.isupper()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s5.str.istitle`` method checks for whether all words are in title\n",
      " |      case (whether only the first letter of each word is capitalized). Words are\n",
      " |      assumed to be as any sequence of non-numeric characters seperated by\n",
      " |      whitespace characters.\n",
      " |      \n",
      " |      >>> s5.str.istitle()\n",
      " |      0    False\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |  \n",
      " |  isdigit lambda self\n",
      " |      Check whether all characters in each string are digits.\n",
      " |      \n",
      " |      This is equivalent to running the Python string method\n",
      " |      :meth:`str.isdigit` for each element of the Series/Index. If a string\n",
      " |      has zero characters, ``False`` is returned for that check.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series or Index of bool\n",
      " |          Series or Index of boolean values with the same length as the original\n",
      " |          Series/Index.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.isalpha : Check whether all characters are alphabetic.\n",
      " |      Series.str.isnumeric : Check whether all characters are numeric.\n",
      " |      Series.str.isalnum : Check whether all characters are alphanumeric.\n",
      " |      Series.str.isdigit : Check whether all characters are digits.\n",
      " |      Series.str.isdecimal : Check whether all characters are decimal.\n",
      " |      Series.str.isspace : Check whether all characters are whitespace.\n",
      " |      Series.str.islower : Check whether all characters are lowercase.\n",
      " |      Series.str.isupper : Check whether all characters are uppercase.\n",
      " |      Series.str.istitle : Check whether all characters are titlecase.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      **Checks for Alphabetic and Numeric Characters**\n",
      " |      \n",
      " |      >>> s1 = pd.Series(['one', 'one1', '1', ''])\n",
      " |      \n",
      " |      >>> s1.str.isalpha()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s1.str.isnumeric()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s1.str.isalnum()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      Note that checks against characters mixed with any additional punctuation\n",
      " |      or whitespace will evaluate to false for an alphanumeric check.\n",
      " |      \n",
      " |      >>> s2 = pd.Series(['A B', '1.5', '3,000'])\n",
      " |      >>> s2.str.isalnum()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **More Detailed Checks for Numeric Characters**\n",
      " |      \n",
      " |      There are several different but overlapping sets of numeric characters that\n",
      " |      can be checked for.\n",
      " |      \n",
      " |      >>> s3 = pd.Series(['23', '³', '⅕', ''])\n",
      " |      \n",
      " |      The ``s3.str.isdecimal`` method checks for characters used to form numbers\n",
      " |      in base 10.\n",
      " |      \n",
      " |      >>> s3.str.isdecimal()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s.str.isdigit`` method is the same as ``s3.str.isdecimal`` but also\n",
      " |      includes special digits, like superscripted and subscripted digits in\n",
      " |      unicode.\n",
      " |      \n",
      " |      >>> s3.str.isdigit()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s.str.isnumeric`` method is the same as ``s3.str.isdigit`` but also\n",
      " |      includes other characters that can represent quantities such as unicode\n",
      " |      fractions.\n",
      " |      \n",
      " |      >>> s3.str.isnumeric()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **Checks for Whitespace**\n",
      " |      \n",
      " |      >>> s4 = pd.Series([' ', '\\t\\r\\n ', ''])\n",
      " |      >>> s4.str.isspace()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **Checks for Character Case**\n",
      " |      \n",
      " |      >>> s5 = pd.Series(['leopard', 'Golden Eagle', 'SNAKE', ''])\n",
      " |      \n",
      " |      >>> s5.str.islower()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s5.str.isupper()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s5.str.istitle`` method checks for whether all words are in title\n",
      " |      case (whether only the first letter of each word is capitalized). Words are\n",
      " |      assumed to be as any sequence of non-numeric characters seperated by\n",
      " |      whitespace characters.\n",
      " |      \n",
      " |      >>> s5.str.istitle()\n",
      " |      0    False\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |  \n",
      " |  islower lambda self\n",
      " |      Check whether all characters in each string are lowercase.\n",
      " |      \n",
      " |      This is equivalent to running the Python string method\n",
      " |      :meth:`str.islower` for each element of the Series/Index. If a string\n",
      " |      has zero characters, ``False`` is returned for that check.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series or Index of bool\n",
      " |          Series or Index of boolean values with the same length as the original\n",
      " |          Series/Index.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.isalpha : Check whether all characters are alphabetic.\n",
      " |      Series.str.isnumeric : Check whether all characters are numeric.\n",
      " |      Series.str.isalnum : Check whether all characters are alphanumeric.\n",
      " |      Series.str.isdigit : Check whether all characters are digits.\n",
      " |      Series.str.isdecimal : Check whether all characters are decimal.\n",
      " |      Series.str.isspace : Check whether all characters are whitespace.\n",
      " |      Series.str.islower : Check whether all characters are lowercase.\n",
      " |      Series.str.isupper : Check whether all characters are uppercase.\n",
      " |      Series.str.istitle : Check whether all characters are titlecase.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      **Checks for Alphabetic and Numeric Characters**\n",
      " |      \n",
      " |      >>> s1 = pd.Series(['one', 'one1', '1', ''])\n",
      " |      \n",
      " |      >>> s1.str.isalpha()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s1.str.isnumeric()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s1.str.isalnum()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      Note that checks against characters mixed with any additional punctuation\n",
      " |      or whitespace will evaluate to false for an alphanumeric check.\n",
      " |      \n",
      " |      >>> s2 = pd.Series(['A B', '1.5', '3,000'])\n",
      " |      >>> s2.str.isalnum()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **More Detailed Checks for Numeric Characters**\n",
      " |      \n",
      " |      There are several different but overlapping sets of numeric characters that\n",
      " |      can be checked for.\n",
      " |      \n",
      " |      >>> s3 = pd.Series(['23', '³', '⅕', ''])\n",
      " |      \n",
      " |      The ``s3.str.isdecimal`` method checks for characters used to form numbers\n",
      " |      in base 10.\n",
      " |      \n",
      " |      >>> s3.str.isdecimal()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s.str.isdigit`` method is the same as ``s3.str.isdecimal`` but also\n",
      " |      includes special digits, like superscripted and subscripted digits in\n",
      " |      unicode.\n",
      " |      \n",
      " |      >>> s3.str.isdigit()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s.str.isnumeric`` method is the same as ``s3.str.isdigit`` but also\n",
      " |      includes other characters that can represent quantities such as unicode\n",
      " |      fractions.\n",
      " |      \n",
      " |      >>> s3.str.isnumeric()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **Checks for Whitespace**\n",
      " |      \n",
      " |      >>> s4 = pd.Series([' ', '\\t\\r\\n ', ''])\n",
      " |      >>> s4.str.isspace()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **Checks for Character Case**\n",
      " |      \n",
      " |      >>> s5 = pd.Series(['leopard', 'Golden Eagle', 'SNAKE', ''])\n",
      " |      \n",
      " |      >>> s5.str.islower()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s5.str.isupper()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s5.str.istitle`` method checks for whether all words are in title\n",
      " |      case (whether only the first letter of each word is capitalized). Words are\n",
      " |      assumed to be as any sequence of non-numeric characters seperated by\n",
      " |      whitespace characters.\n",
      " |      \n",
      " |      >>> s5.str.istitle()\n",
      " |      0    False\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |  \n",
      " |  isnumeric lambda self\n",
      " |      Check whether all characters in each string are numeric.\n",
      " |      \n",
      " |      This is equivalent to running the Python string method\n",
      " |      :meth:`str.isnumeric` for each element of the Series/Index. If a string\n",
      " |      has zero characters, ``False`` is returned for that check.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series or Index of bool\n",
      " |          Series or Index of boolean values with the same length as the original\n",
      " |          Series/Index.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.isalpha : Check whether all characters are alphabetic.\n",
      " |      Series.str.isnumeric : Check whether all characters are numeric.\n",
      " |      Series.str.isalnum : Check whether all characters are alphanumeric.\n",
      " |      Series.str.isdigit : Check whether all characters are digits.\n",
      " |      Series.str.isdecimal : Check whether all characters are decimal.\n",
      " |      Series.str.isspace : Check whether all characters are whitespace.\n",
      " |      Series.str.islower : Check whether all characters are lowercase.\n",
      " |      Series.str.isupper : Check whether all characters are uppercase.\n",
      " |      Series.str.istitle : Check whether all characters are titlecase.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      **Checks for Alphabetic and Numeric Characters**\n",
      " |      \n",
      " |      >>> s1 = pd.Series(['one', 'one1', '1', ''])\n",
      " |      \n",
      " |      >>> s1.str.isalpha()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s1.str.isnumeric()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s1.str.isalnum()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      Note that checks against characters mixed with any additional punctuation\n",
      " |      or whitespace will evaluate to false for an alphanumeric check.\n",
      " |      \n",
      " |      >>> s2 = pd.Series(['A B', '1.5', '3,000'])\n",
      " |      >>> s2.str.isalnum()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **More Detailed Checks for Numeric Characters**\n",
      " |      \n",
      " |      There are several different but overlapping sets of numeric characters that\n",
      " |      can be checked for.\n",
      " |      \n",
      " |      >>> s3 = pd.Series(['23', '³', '⅕', ''])\n",
      " |      \n",
      " |      The ``s3.str.isdecimal`` method checks for characters used to form numbers\n",
      " |      in base 10.\n",
      " |      \n",
      " |      >>> s3.str.isdecimal()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s.str.isdigit`` method is the same as ``s3.str.isdecimal`` but also\n",
      " |      includes special digits, like superscripted and subscripted digits in\n",
      " |      unicode.\n",
      " |      \n",
      " |      >>> s3.str.isdigit()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s.str.isnumeric`` method is the same as ``s3.str.isdigit`` but also\n",
      " |      includes other characters that can represent quantities such as unicode\n",
      " |      fractions.\n",
      " |      \n",
      " |      >>> s3.str.isnumeric()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **Checks for Whitespace**\n",
      " |      \n",
      " |      >>> s4 = pd.Series([' ', '\\t\\r\\n ', ''])\n",
      " |      >>> s4.str.isspace()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **Checks for Character Case**\n",
      " |      \n",
      " |      >>> s5 = pd.Series(['leopard', 'Golden Eagle', 'SNAKE', ''])\n",
      " |      \n",
      " |      >>> s5.str.islower()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s5.str.isupper()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s5.str.istitle`` method checks for whether all words are in title\n",
      " |      case (whether only the first letter of each word is capitalized). Words are\n",
      " |      assumed to be as any sequence of non-numeric characters seperated by\n",
      " |      whitespace characters.\n",
      " |      \n",
      " |      >>> s5.str.istitle()\n",
      " |      0    False\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |  \n",
      " |  isspace lambda self\n",
      " |      Check whether all characters in each string are whitespace.\n",
      " |      \n",
      " |      This is equivalent to running the Python string method\n",
      " |      :meth:`str.isspace` for each element of the Series/Index. If a string\n",
      " |      has zero characters, ``False`` is returned for that check.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series or Index of bool\n",
      " |          Series or Index of boolean values with the same length as the original\n",
      " |          Series/Index.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.isalpha : Check whether all characters are alphabetic.\n",
      " |      Series.str.isnumeric : Check whether all characters are numeric.\n",
      " |      Series.str.isalnum : Check whether all characters are alphanumeric.\n",
      " |      Series.str.isdigit : Check whether all characters are digits.\n",
      " |      Series.str.isdecimal : Check whether all characters are decimal.\n",
      " |      Series.str.isspace : Check whether all characters are whitespace.\n",
      " |      Series.str.islower : Check whether all characters are lowercase.\n",
      " |      Series.str.isupper : Check whether all characters are uppercase.\n",
      " |      Series.str.istitle : Check whether all characters are titlecase.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      **Checks for Alphabetic and Numeric Characters**\n",
      " |      \n",
      " |      >>> s1 = pd.Series(['one', 'one1', '1', ''])\n",
      " |      \n",
      " |      >>> s1.str.isalpha()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s1.str.isnumeric()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s1.str.isalnum()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      Note that checks against characters mixed with any additional punctuation\n",
      " |      or whitespace will evaluate to false for an alphanumeric check.\n",
      " |      \n",
      " |      >>> s2 = pd.Series(['A B', '1.5', '3,000'])\n",
      " |      >>> s2.str.isalnum()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **More Detailed Checks for Numeric Characters**\n",
      " |      \n",
      " |      There are several different but overlapping sets of numeric characters that\n",
      " |      can be checked for.\n",
      " |      \n",
      " |      >>> s3 = pd.Series(['23', '³', '⅕', ''])\n",
      " |      \n",
      " |      The ``s3.str.isdecimal`` method checks for characters used to form numbers\n",
      " |      in base 10.\n",
      " |      \n",
      " |      >>> s3.str.isdecimal()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s.str.isdigit`` method is the same as ``s3.str.isdecimal`` but also\n",
      " |      includes special digits, like superscripted and subscripted digits in\n",
      " |      unicode.\n",
      " |      \n",
      " |      >>> s3.str.isdigit()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s.str.isnumeric`` method is the same as ``s3.str.isdigit`` but also\n",
      " |      includes other characters that can represent quantities such as unicode\n",
      " |      fractions.\n",
      " |      \n",
      " |      >>> s3.str.isnumeric()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **Checks for Whitespace**\n",
      " |      \n",
      " |      >>> s4 = pd.Series([' ', '\\t\\r\\n ', ''])\n",
      " |      >>> s4.str.isspace()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **Checks for Character Case**\n",
      " |      \n",
      " |      >>> s5 = pd.Series(['leopard', 'Golden Eagle', 'SNAKE', ''])\n",
      " |      \n",
      " |      >>> s5.str.islower()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s5.str.isupper()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s5.str.istitle`` method checks for whether all words are in title\n",
      " |      case (whether only the first letter of each word is capitalized). Words are\n",
      " |      assumed to be as any sequence of non-numeric characters seperated by\n",
      " |      whitespace characters.\n",
      " |      \n",
      " |      >>> s5.str.istitle()\n",
      " |      0    False\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |  \n",
      " |  istitle lambda self\n",
      " |      Check whether all characters in each string are titlecase.\n",
      " |      \n",
      " |      This is equivalent to running the Python string method\n",
      " |      :meth:`str.istitle` for each element of the Series/Index. If a string\n",
      " |      has zero characters, ``False`` is returned for that check.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series or Index of bool\n",
      " |          Series or Index of boolean values with the same length as the original\n",
      " |          Series/Index.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.isalpha : Check whether all characters are alphabetic.\n",
      " |      Series.str.isnumeric : Check whether all characters are numeric.\n",
      " |      Series.str.isalnum : Check whether all characters are alphanumeric.\n",
      " |      Series.str.isdigit : Check whether all characters are digits.\n",
      " |      Series.str.isdecimal : Check whether all characters are decimal.\n",
      " |      Series.str.isspace : Check whether all characters are whitespace.\n",
      " |      Series.str.islower : Check whether all characters are lowercase.\n",
      " |      Series.str.isupper : Check whether all characters are uppercase.\n",
      " |      Series.str.istitle : Check whether all characters are titlecase.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      **Checks for Alphabetic and Numeric Characters**\n",
      " |      \n",
      " |      >>> s1 = pd.Series(['one', 'one1', '1', ''])\n",
      " |      \n",
      " |      >>> s1.str.isalpha()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s1.str.isnumeric()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s1.str.isalnum()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      Note that checks against characters mixed with any additional punctuation\n",
      " |      or whitespace will evaluate to false for an alphanumeric check.\n",
      " |      \n",
      " |      >>> s2 = pd.Series(['A B', '1.5', '3,000'])\n",
      " |      >>> s2.str.isalnum()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **More Detailed Checks for Numeric Characters**\n",
      " |      \n",
      " |      There are several different but overlapping sets of numeric characters that\n",
      " |      can be checked for.\n",
      " |      \n",
      " |      >>> s3 = pd.Series(['23', '³', '⅕', ''])\n",
      " |      \n",
      " |      The ``s3.str.isdecimal`` method checks for characters used to form numbers\n",
      " |      in base 10.\n",
      " |      \n",
      " |      >>> s3.str.isdecimal()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s.str.isdigit`` method is the same as ``s3.str.isdecimal`` but also\n",
      " |      includes special digits, like superscripted and subscripted digits in\n",
      " |      unicode.\n",
      " |      \n",
      " |      >>> s3.str.isdigit()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s.str.isnumeric`` method is the same as ``s3.str.isdigit`` but also\n",
      " |      includes other characters that can represent quantities such as unicode\n",
      " |      fractions.\n",
      " |      \n",
      " |      >>> s3.str.isnumeric()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **Checks for Whitespace**\n",
      " |      \n",
      " |      >>> s4 = pd.Series([' ', '\\t\\r\\n ', ''])\n",
      " |      >>> s4.str.isspace()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **Checks for Character Case**\n",
      " |      \n",
      " |      >>> s5 = pd.Series(['leopard', 'Golden Eagle', 'SNAKE', ''])\n",
      " |      \n",
      " |      >>> s5.str.islower()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s5.str.isupper()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s5.str.istitle`` method checks for whether all words are in title\n",
      " |      case (whether only the first letter of each word is capitalized). Words are\n",
      " |      assumed to be as any sequence of non-numeric characters seperated by\n",
      " |      whitespace characters.\n",
      " |      \n",
      " |      >>> s5.str.istitle()\n",
      " |      0    False\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |  \n",
      " |  isupper lambda self\n",
      " |      Check whether all characters in each string are uppercase.\n",
      " |      \n",
      " |      This is equivalent to running the Python string method\n",
      " |      :meth:`str.isupper` for each element of the Series/Index. If a string\n",
      " |      has zero characters, ``False`` is returned for that check.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series or Index of bool\n",
      " |          Series or Index of boolean values with the same length as the original\n",
      " |          Series/Index.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.isalpha : Check whether all characters are alphabetic.\n",
      " |      Series.str.isnumeric : Check whether all characters are numeric.\n",
      " |      Series.str.isalnum : Check whether all characters are alphanumeric.\n",
      " |      Series.str.isdigit : Check whether all characters are digits.\n",
      " |      Series.str.isdecimal : Check whether all characters are decimal.\n",
      " |      Series.str.isspace : Check whether all characters are whitespace.\n",
      " |      Series.str.islower : Check whether all characters are lowercase.\n",
      " |      Series.str.isupper : Check whether all characters are uppercase.\n",
      " |      Series.str.istitle : Check whether all characters are titlecase.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      **Checks for Alphabetic and Numeric Characters**\n",
      " |      \n",
      " |      >>> s1 = pd.Series(['one', 'one1', '1', ''])\n",
      " |      \n",
      " |      >>> s1.str.isalpha()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s1.str.isnumeric()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s1.str.isalnum()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      Note that checks against characters mixed with any additional punctuation\n",
      " |      or whitespace will evaluate to false for an alphanumeric check.\n",
      " |      \n",
      " |      >>> s2 = pd.Series(['A B', '1.5', '3,000'])\n",
      " |      >>> s2.str.isalnum()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **More Detailed Checks for Numeric Characters**\n",
      " |      \n",
      " |      There are several different but overlapping sets of numeric characters that\n",
      " |      can be checked for.\n",
      " |      \n",
      " |      >>> s3 = pd.Series(['23', '³', '⅕', ''])\n",
      " |      \n",
      " |      The ``s3.str.isdecimal`` method checks for characters used to form numbers\n",
      " |      in base 10.\n",
      " |      \n",
      " |      >>> s3.str.isdecimal()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s.str.isdigit`` method is the same as ``s3.str.isdecimal`` but also\n",
      " |      includes special digits, like superscripted and subscripted digits in\n",
      " |      unicode.\n",
      " |      \n",
      " |      >>> s3.str.isdigit()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s.str.isnumeric`` method is the same as ``s3.str.isdigit`` but also\n",
      " |      includes other characters that can represent quantities such as unicode\n",
      " |      fractions.\n",
      " |      \n",
      " |      >>> s3.str.isnumeric()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **Checks for Whitespace**\n",
      " |      \n",
      " |      >>> s4 = pd.Series([' ', '\\t\\r\\n ', ''])\n",
      " |      >>> s4.str.isspace()\n",
      " |      0     True\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      **Checks for Character Case**\n",
      " |      \n",
      " |      >>> s5 = pd.Series(['leopard', 'Golden Eagle', 'SNAKE', ''])\n",
      " |      \n",
      " |      >>> s5.str.islower()\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      >>> s5.str.isupper()\n",
      " |      0    False\n",
      " |      1    False\n",
      " |      2     True\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |      \n",
      " |      The ``s5.str.istitle`` method checks for whether all words are in title\n",
      " |      case (whether only the first letter of each word is capitalized). Words are\n",
      " |      assumed to be as any sequence of non-numeric characters seperated by\n",
      " |      whitespace characters.\n",
      " |      \n",
      " |      >>> s5.str.istitle()\n",
      " |      0    False\n",
      " |      1     True\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |  \n",
      " |  join(self, sep)\n",
      " |      Join lists contained as elements in the Series/Index with passed delimiter.\n",
      " |      \n",
      " |      If the elements of a Series are lists themselves, join the content of these\n",
      " |      lists using the delimiter passed to the function.\n",
      " |      This function is an equivalent to :meth:`str.join`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      sep : str\n",
      " |          Delimiter to use between list entries.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series/Index: object\n",
      " |          The list entries concatenated by intervening occurrences of the\n",
      " |          delimiter.\n",
      " |      \n",
      " |      Raises\n",
      " |      -------\n",
      " |      AttributeError\n",
      " |          If the supplied Series contains neither strings nor lists.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      str.join : Standard library version of this method.\n",
      " |      Series.str.split : Split strings around given separator/delimiter.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      If any of the list items is not a string object, the result of the join\n",
      " |      will be `NaN`.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      Example with a list that contains non-string elements.\n",
      " |      \n",
      " |      >>> s = pd.Series([['lion', 'elephant', 'zebra'],\n",
      " |      ...                [1.1, 2.2, 3.3],\n",
      " |      ...                ['cat', np.nan, 'dog'],\n",
      " |      ...                ['cow', 4.5, 'goat'],\n",
      " |      ...                ['duck', ['swan', 'fish'], 'guppy']])\n",
      " |      >>> s\n",
      " |      0        [lion, elephant, zebra]\n",
      " |      1                [1.1, 2.2, 3.3]\n",
      " |      2                [cat, nan, dog]\n",
      " |      3               [cow, 4.5, goat]\n",
      " |      4    [duck, [swan, fish], guppy]\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Join all lists using a '-'. The lists containing object(s) of types other\n",
      " |      than str will produce a NaN.\n",
      " |      \n",
      " |      >>> s.str.join('-')\n",
      " |      0    lion-elephant-zebra\n",
      " |      1                    NaN\n",
      " |      2                    NaN\n",
      " |      3                    NaN\n",
      " |      4                    NaN\n",
      " |      dtype: object\n",
      " |  \n",
      " |  len(self)\n",
      " |      Computes the length of each element in the Series/Index. The element may be\n",
      " |      a sequence (such as a string, tuple or list) or a collection\n",
      " |      (such as a dictionary).\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series or Index of int\n",
      " |          A Series or Index of integer values indicating the length of each\n",
      " |          element in the Series or Index.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      str.len : Python built-in function returning the length of an object.\n",
      " |      Series.size : Returns the length of the Series.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      Returns the length (number of characters) in a string. Returns the\n",
      " |      number of entries for dictionaries, lists or tuples.\n",
      " |      \n",
      " |      >>> s = pd.Series(['dog',\n",
      " |      ...                 '',\n",
      " |      ...                 5,\n",
      " |      ...                 {'foo' : 'bar'},\n",
      " |      ...                 [2, 3, 5, 7],\n",
      " |      ...                 ('one', 'two', 'three')])\n",
      " |      >>> s\n",
      " |      0                  dog\n",
      " |      1\n",
      " |      2                    5\n",
      " |      3       {'foo': 'bar'}\n",
      " |      4         [2, 3, 5, 7]\n",
      " |      5    (one, two, three)\n",
      " |      dtype: object\n",
      " |      >>> s.str.len()\n",
      " |      0    3.0\n",
      " |      1    0.0\n",
      " |      2    NaN\n",
      " |      3    1.0\n",
      " |      4    4.0\n",
      " |      5    3.0\n",
      " |      dtype: float64\n",
      " |  \n",
      " |  ljust(self, width, fillchar=' ')\n",
      " |      Filling right side of strings in the Series/Index with an\n",
      " |      additional character. Equivalent to :meth:`str.ljust`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      width : int\n",
      " |          Minimum width of resulting string; additional characters will be filled\n",
      " |          with ``fillchar``\n",
      " |      fillchar : str\n",
      " |          Additional character for filling, default is whitespace\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      filled : Series/Index of objects\n",
      " |  \n",
      " |  lower lambda self\n",
      " |      Convert strings in the Series/Index to lowercase.\n",
      " |      \n",
      " |      Equivalent to :meth:`str.lower`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series/Index of objects\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.lower : Converts all characters to lowercase.\n",
      " |      Series.str.upper : Converts all characters to uppercase.\n",
      " |      Series.str.title : Converts first character of each word to uppercase and\n",
      " |          remaining to lowercase.\n",
      " |      Series.str.capitalize : Converts first character to uppercase and\n",
      " |          remaining to lowercase.\n",
      " |      Series.str.swapcase : Converts uppercase to lowercase and lowercase to\n",
      " |          uppercase.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series(['lower', 'CAPITALS', 'this is a sentence', 'SwApCaSe'])\n",
      " |      >>> s\n",
      " |      0                 lower\n",
      " |      1              CAPITALS\n",
      " |      2    this is a sentence\n",
      " |      3              SwApCaSe\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.lower()\n",
      " |      0                 lower\n",
      " |      1              capitals\n",
      " |      2    this is a sentence\n",
      " |      3              swapcase\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.upper()\n",
      " |      0                 LOWER\n",
      " |      1              CAPITALS\n",
      " |      2    THIS IS A SENTENCE\n",
      " |      3              SWAPCASE\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.title()\n",
      " |      0                 Lower\n",
      " |      1              Capitals\n",
      " |      2    This Is A Sentence\n",
      " |      3              Swapcase\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.capitalize()\n",
      " |      0                 Lower\n",
      " |      1              Capitals\n",
      " |      2    This is a sentence\n",
      " |      3              Swapcase\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.swapcase()\n",
      " |      0                 LOWER\n",
      " |      1              capitals\n",
      " |      2    THIS IS A SENTENCE\n",
      " |      3              sWaPcAsE\n",
      " |      dtype: object\n",
      " |  \n",
      " |  lstrip(self, to_strip=None)\n",
      " |      Remove leading and trailing characters.\n",
      " |      \n",
      " |      Strip whitespaces (including newlines) or a set of specified characters\n",
      " |      from each string in the Series/Index from left side.\n",
      " |      Equivalent to :meth:`str.lstrip`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      to_strip : str or None, default None\n",
      " |          Specifying the set of characters to be removed.\n",
      " |          All combinations of this set of characters will be stripped.\n",
      " |          If None then whitespaces are removed.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series/Index of objects\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.strip : Remove leading and trailing characters in Series/Index.\n",
      " |      Series.str.lstrip : Remove leading characters in Series/Index.\n",
      " |      Series.str.rstrip : Remove trailing characters in Series/Index.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series(['1. Ant.  ', '2. Bee!\\n', '3. Cat?\\t', np.nan])\n",
      " |      >>> s\n",
      " |      0    1. Ant.\n",
      " |      1    2. Bee!\\n\n",
      " |      2    3. Cat?\\t\n",
      " |      3          NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.strip()\n",
      " |      0    1. Ant.\n",
      " |      1    2. Bee!\n",
      " |      2    3. Cat?\n",
      " |      3        NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.lstrip('123.')\n",
      " |      0    Ant.\n",
      " |      1    Bee!\\n\n",
      " |      2    Cat?\\t\n",
      " |      3       NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.rstrip('.!? \\n\\t')\n",
      " |      0    1. Ant\n",
      " |      1    2. Bee\n",
      " |      2    3. Cat\n",
      " |      3       NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.strip('123.!? \\n\\t')\n",
      " |      0    Ant\n",
      " |      1    Bee\n",
      " |      2    Cat\n",
      " |      3    NaN\n",
      " |      dtype: object\n",
      " |  \n",
      " |  match(self, pat, case=True, flags=0, na=nan)\n",
      " |      Determine if each string matches a regular expression.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      pat : string\n",
      " |          Character sequence or regular expression\n",
      " |      case : boolean, default True\n",
      " |          If True, case sensitive\n",
      " |      flags : int, default 0 (no flags)\n",
      " |          re module flags, e.g. re.IGNORECASE\n",
      " |      na : default NaN, fill value for missing values\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series/array of boolean values\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      contains : Analogous, but less strict, relying on re.search instead of\n",
      " |          re.match.\n",
      " |      extract : Extract matched groups.\n",
      " |  \n",
      " |  normalize(self, form)\n",
      " |      Return the Unicode normal form for the strings in the Series/Index.\n",
      " |      For more information on the forms, see the\n",
      " |      :func:`unicodedata.normalize`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      form : {'NFC', 'NFKC', 'NFD', 'NFKD'}\n",
      " |          Unicode form\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      normalized : Series/Index of objects\n",
      " |  \n",
      " |  pad(self, width, side='left', fillchar=' ')\n",
      " |      Pad strings in the Series/Index up to width.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      width : int\n",
      " |          Minimum width of resulting string; additional characters will be filled\n",
      " |          with character defined in `fillchar`.\n",
      " |      side : {'left', 'right', 'both'}, default 'left'\n",
      " |          Side from which to fill resulting string.\n",
      " |      fillchar : str, default ' '\n",
      " |          Additional character for filling, default is whitespace.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series or Index of object\n",
      " |          Returns Series or Index with minimum number of char in object.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.rjust : Fills the left side of strings with an arbitrary\n",
      " |          character. Equivalent to ``Series.str.pad(side='left')``.\n",
      " |      Series.str.ljust : Fills the right side of strings with an arbitrary\n",
      " |          character. Equivalent to ``Series.str.pad(side='right')``.\n",
      " |      Series.str.center : Fills boths sides of strings with an arbitrary\n",
      " |          character. Equivalent to ``Series.str.pad(side='both')``.\n",
      " |      Series.str.zfill :  Pad strings in the Series/Index by prepending '0'\n",
      " |          character. Equivalent to ``Series.str.pad(side='left', fillchar='0')``.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series([\"caribou\", \"tiger\"])\n",
      " |      >>> s\n",
      " |      0    caribou\n",
      " |      1      tiger\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.pad(width=10)\n",
      " |      0       caribou\n",
      " |      1         tiger\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.pad(width=10, side='right', fillchar='-')\n",
      " |      0    caribou---\n",
      " |      1    tiger-----\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.pad(width=10, side='both', fillchar='-')\n",
      " |      0    -caribou--\n",
      " |      1    --tiger---\n",
      " |      dtype: object\n",
      " |  \n",
      " |  partition(self, sep=' ', expand=True)\n",
      " |      Split the string at the first occurrence of `sep`.\n",
      " |      \n",
      " |      This method splits the string at the first occurrence of `sep`,\n",
      " |      and returns 3 elements containing the part before the separator,\n",
      " |      the separator itself, and the part after the separator.\n",
      " |      If the separator is not found, return 3 elements containing the string itself, followed by two empty strings.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      sep : str, default whitespace\n",
      " |          String to split on.\n",
      " |      pat : str, default whitespace\n",
      " |          .. deprecated:: 0.24.0\n",
      " |             Use ``sep`` instead\n",
      " |      expand : bool, default True\n",
      " |          If True, return DataFrame/MultiIndex expanding dimensionality.\n",
      " |          If False, return Series/Index.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      DataFrame/MultiIndex or Series/Index of objects\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      rpartition : Split the string at the last occurrence of `sep`.\n",
      " |      Series.str.split : Split strings around given separators.\n",
      " |      str.partition : Standard library version.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      \n",
      " |      >>> s = pd.Series(['Linda van der Berg', 'George Pitt-Rivers'])\n",
      " |      >>> s\n",
      " |      0    Linda van der Berg\n",
      " |      1    George Pitt-Rivers\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.partition()\n",
      " |              0  1             2\n",
      " |      0   Linda     van der Berg\n",
      " |      1  George      Pitt-Rivers\n",
      " |      \n",
      " |      To partition by the last space instead of the first one:\n",
      " |      \n",
      " |      >>> s.str.rpartition()\n",
      " |                     0  1            2\n",
      " |      0  Linda van der            Berg\n",
      " |      1         George     Pitt-Rivers\n",
      " |      \n",
      " |      To partition by something different than a space:\n",
      " |      \n",
      " |      >>> s.str.partition('-')\n",
      " |                          0  1       2\n",
      " |      0  Linda van der Berg\n",
      " |      1         George Pitt  -  Rivers\n",
      " |      \n",
      " |      To return a Series containining tuples instead of a DataFrame:\n",
      " |      \n",
      " |      >>> s.str.partition('-', expand=False)\n",
      " |      0    (Linda van der Berg, , )\n",
      " |      1    (George Pitt, -, Rivers)\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Also available on indices:\n",
      " |      \n",
      " |      >>> idx = pd.Index(['X 123', 'Y 999'])\n",
      " |      >>> idx\n",
      " |      Index(['X 123', 'Y 999'], dtype='object')\n",
      " |      \n",
      " |      Which will create a MultiIndex:\n",
      " |      \n",
      " |      >>> idx.str.partition()\n",
      " |      MultiIndex(levels=[['X', 'Y'], [' '], ['123', '999']],\n",
      " |                 codes=[[0, 1], [0, 0], [0, 1]])\n",
      " |      \n",
      " |      Or an index with tuples with ``expand=False``:\n",
      " |      \n",
      " |      >>> idx.str.partition(expand=False)\n",
      " |      Index([('X', ' ', '123'), ('Y', ' ', '999')], dtype='object')\n",
      " |  \n",
      " |  repeat(self, repeats)\n",
      " |      Duplicate each string in the Series or Index.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      repeats : int or sequence of int\n",
      " |          Same value for all (int) or different value per (sequence).\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series or Index of object\n",
      " |          Series or Index of repeated string objects specified by\n",
      " |          input parameter repeats.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series(['a', 'b', 'c'])\n",
      " |      >>> s\n",
      " |      0    a\n",
      " |      1    b\n",
      " |      2    c\n",
      " |      \n",
      " |      Single int repeats string in Series\n",
      " |      \n",
      " |      >>> s.str.repeat(repeats=2)\n",
      " |      0    aa\n",
      " |      1    bb\n",
      " |      2    cc\n",
      " |      \n",
      " |      Sequence of int repeats corresponding string in Series\n",
      " |      \n",
      " |      >>> s.str.repeat(repeats=[1, 2, 3])\n",
      " |      0      a\n",
      " |      1     bb\n",
      " |      2    ccc\n",
      " |  \n",
      " |  replace(self, pat, repl, n=-1, case=None, flags=0, regex=True)\n",
      " |      Replace occurrences of pattern/regex in the Series/Index with\n",
      " |      some other string. Equivalent to :meth:`str.replace` or\n",
      " |      :func:`re.sub`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      pat : string or compiled regex\n",
      " |          String can be a character sequence or regular expression.\n",
      " |      \n",
      " |          .. versionadded:: 0.20.0\n",
      " |              `pat` also accepts a compiled regex.\n",
      " |      \n",
      " |      repl : string or callable\n",
      " |          Replacement string or a callable. The callable is passed the regex\n",
      " |          match object and must return a replacement string to be used.\n",
      " |          See :func:`re.sub`.\n",
      " |      \n",
      " |          .. versionadded:: 0.20.0\n",
      " |              `repl` also accepts a callable.\n",
      " |      \n",
      " |      n : int, default -1 (all)\n",
      " |          Number of replacements to make from start\n",
      " |      case : boolean, default None\n",
      " |          - If True, case sensitive (the default if `pat` is a string)\n",
      " |          - Set to False for case insensitive\n",
      " |          - Cannot be set if `pat` is a compiled regex\n",
      " |      flags : int, default 0 (no flags)\n",
      " |          - re module flags, e.g. re.IGNORECASE\n",
      " |          - Cannot be set if `pat` is a compiled regex\n",
      " |      regex : boolean, default True\n",
      " |          - If True, assumes the passed-in pattern is a regular expression.\n",
      " |          - If False, treats the pattern as a literal string\n",
      " |          - Cannot be set to False if `pat` is a compiled regex or `repl` is\n",
      " |            a callable.\n",
      " |      \n",
      " |          .. versionadded:: 0.23.0\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series or Index of object\n",
      " |          A copy of the object with all matching occurrences of `pat` replaced by\n",
      " |          `repl`.\n",
      " |      \n",
      " |      Raises\n",
      " |      ------\n",
      " |      ValueError\n",
      " |          * if `regex` is False and `repl` is a callable or `pat` is a compiled\n",
      " |            regex\n",
      " |          * if `pat` is a compiled regex and `case` or `flags` is set\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      When `pat` is a compiled regex, all flags should be included in the\n",
      " |      compiled regex. Use of `case`, `flags`, or `regex=False` with a compiled\n",
      " |      regex will raise an error.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      When `pat` is a string and `regex` is True (the default), the given `pat`\n",
      " |      is compiled as a regex. When `repl` is a string, it replaces matching\n",
      " |      regex patterns as with :meth:`re.sub`. NaN value(s) in the Series are\n",
      " |      left as is:\n",
      " |      \n",
      " |      >>> pd.Series(['foo', 'fuz', np.nan]).str.replace('f.', 'ba', regex=True)\n",
      " |      0    bao\n",
      " |      1    baz\n",
      " |      2    NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      When `pat` is a string and `regex` is False, every `pat` is replaced with\n",
      " |      `repl` as with :meth:`str.replace`:\n",
      " |      \n",
      " |      >>> pd.Series(['f.o', 'fuz', np.nan]).str.replace('f.', 'ba', regex=False)\n",
      " |      0    bao\n",
      " |      1    fuz\n",
      " |      2    NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      When `repl` is a callable, it is called on every `pat` using\n",
      " |      :func:`re.sub`. The callable should expect one positional argument\n",
      " |      (a regex object) and return a string.\n",
      " |      \n",
      " |      To get the idea:\n",
      " |      \n",
      " |      >>> pd.Series(['foo', 'fuz', np.nan]).str.replace('f', repr)\n",
      " |      0    <_sre.SRE_Match object; span=(0, 1), match='f'>oo\n",
      " |      1    <_sre.SRE_Match object; span=(0, 1), match='f'>uz\n",
      " |      2                                                  NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Reverse every lowercase alphabetic word:\n",
      " |      \n",
      " |      >>> repl = lambda m: m.group(0)[::-1]\n",
      " |      >>> pd.Series(['foo 123', 'bar baz', np.nan]).str.replace(r'[a-z]+', repl)\n",
      " |      0    oof 123\n",
      " |      1    rab zab\n",
      " |      2        NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Using regex groups (extract second group and swap case):\n",
      " |      \n",
      " |      >>> pat = r\"(?P<one>\\w+) (?P<two>\\w+) (?P<three>\\w+)\"\n",
      " |      >>> repl = lambda m: m.group('two').swapcase()\n",
      " |      >>> pd.Series(['One Two Three', 'Foo Bar Baz']).str.replace(pat, repl)\n",
      " |      0    tWO\n",
      " |      1    bAR\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Using a compiled regex with flags\n",
      " |      \n",
      " |      >>> regex_pat = re.compile(r'FUZ', flags=re.IGNORECASE)\n",
      " |      >>> pd.Series(['foo', 'fuz', np.nan]).str.replace(regex_pat, 'bar')\n",
      " |      0    foo\n",
      " |      1    bar\n",
      " |      2    NaN\n",
      " |      dtype: object\n",
      " |  \n",
      " |  rfind(self, sub, start=0, end=None)\n",
      " |      Return highest indexes in each strings in the Series/Index\n",
      " |      where the substring is fully contained between [start:end].\n",
      " |      Return -1 on failure. Equivalent to standard :meth:`str.rfind`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      sub : str\n",
      " |          Substring being searched\n",
      " |      start : int\n",
      " |          Left edge index\n",
      " |      end : int\n",
      " |          Right edge index\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      found : Series/Index of integer values\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      find : Return lowest indexes in each strings.\n",
      " |  \n",
      " |  rindex(self, sub, start=0, end=None)\n",
      " |      Return highest indexes in each strings where the substring is\n",
      " |      fully contained between [start:end]. This is the same as\n",
      " |      ``str.rfind`` except instead of returning -1, it raises a ValueError\n",
      " |      when the substring is not found. Equivalent to standard ``str.rindex``.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      sub : str\n",
      " |          Substring being searched\n",
      " |      start : int\n",
      " |          Left edge index\n",
      " |      end : int\n",
      " |          Right edge index\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      found : Series/Index of objects\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      index : Return lowest indexes in each strings.\n",
      " |  \n",
      " |  rjust(self, width, fillchar=' ')\n",
      " |      Filling left side of strings in the Series/Index with an\n",
      " |      additional character. Equivalent to :meth:`str.rjust`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      width : int\n",
      " |          Minimum width of resulting string; additional characters will be filled\n",
      " |          with ``fillchar``\n",
      " |      fillchar : str\n",
      " |          Additional character for filling, default is whitespace\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      filled : Series/Index of objects\n",
      " |  \n",
      " |  rpartition(self, sep=' ', expand=True)\n",
      " |      Split the string at the last occurrence of `sep`.\n",
      " |      \n",
      " |      This method splits the string at the last occurrence of `sep`,\n",
      " |      and returns 3 elements containing the part before the separator,\n",
      " |      the separator itself, and the part after the separator.\n",
      " |      If the separator is not found, return 3 elements containing two empty strings, followed by the string itself.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      sep : str, default whitespace\n",
      " |          String to split on.\n",
      " |      pat : str, default whitespace\n",
      " |          .. deprecated:: 0.24.0\n",
      " |             Use ``sep`` instead\n",
      " |      expand : bool, default True\n",
      " |          If True, return DataFrame/MultiIndex expanding dimensionality.\n",
      " |          If False, return Series/Index.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      DataFrame/MultiIndex or Series/Index of objects\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      partition : Split the string at the first occurrence of `sep`.\n",
      " |      Series.str.split : Split strings around given separators.\n",
      " |      str.partition : Standard library version.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      \n",
      " |      >>> s = pd.Series(['Linda van der Berg', 'George Pitt-Rivers'])\n",
      " |      >>> s\n",
      " |      0    Linda van der Berg\n",
      " |      1    George Pitt-Rivers\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.partition()\n",
      " |              0  1             2\n",
      " |      0   Linda     van der Berg\n",
      " |      1  George      Pitt-Rivers\n",
      " |      \n",
      " |      To partition by the last space instead of the first one:\n",
      " |      \n",
      " |      >>> s.str.rpartition()\n",
      " |                     0  1            2\n",
      " |      0  Linda van der            Berg\n",
      " |      1         George     Pitt-Rivers\n",
      " |      \n",
      " |      To partition by something different than a space:\n",
      " |      \n",
      " |      >>> s.str.partition('-')\n",
      " |                          0  1       2\n",
      " |      0  Linda van der Berg\n",
      " |      1         George Pitt  -  Rivers\n",
      " |      \n",
      " |      To return a Series containining tuples instead of a DataFrame:\n",
      " |      \n",
      " |      >>> s.str.partition('-', expand=False)\n",
      " |      0    (Linda van der Berg, , )\n",
      " |      1    (George Pitt, -, Rivers)\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Also available on indices:\n",
      " |      \n",
      " |      >>> idx = pd.Index(['X 123', 'Y 999'])\n",
      " |      >>> idx\n",
      " |      Index(['X 123', 'Y 999'], dtype='object')\n",
      " |      \n",
      " |      Which will create a MultiIndex:\n",
      " |      \n",
      " |      >>> idx.str.partition()\n",
      " |      MultiIndex(levels=[['X', 'Y'], [' '], ['123', '999']],\n",
      " |                 codes=[[0, 1], [0, 0], [0, 1]])\n",
      " |      \n",
      " |      Or an index with tuples with ``expand=False``:\n",
      " |      \n",
      " |      >>> idx.str.partition(expand=False)\n",
      " |      Index([('X', ' ', '123'), ('Y', ' ', '999')], dtype='object')\n",
      " |  \n",
      " |  rsplit(self, pat=None, n=-1, expand=False)\n",
      " |      Split strings around given separator/delimiter.\n",
      " |      \n",
      " |      Splits the string in the Series/Index from the end,\n",
      " |      at the specified delimiter string. Equivalent to :meth:`str.rsplit`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      pat : str, optional\n",
      " |          String or regular expression to split on.\n",
      " |          If not specified, split on whitespace.\n",
      " |      n : int, default -1 (all)\n",
      " |          Limit number of splits in output.\n",
      " |          ``None``, 0 and -1 will be interpreted as return all splits.\n",
      " |      expand : bool, default False\n",
      " |          Expand the splitted strings into separate columns.\n",
      " |      \n",
      " |          * If ``True``, return DataFrame/MultiIndex expanding dimensionality.\n",
      " |          * If ``False``, return Series/Index, containing lists of strings.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series, Index, DataFrame or MultiIndex\n",
      " |          Type matches caller unless ``expand=True`` (see Notes).\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |       Series.str.split : Split strings around given separator/delimiter.\n",
      " |       Series.str.rsplit : Splits string around given separator/delimiter,\n",
      " |       starting from the right.\n",
      " |       Series.str.join : Join lists contained as elements in the Series/Index\n",
      " |       with passed delimiter.\n",
      " |       str.split : Standard library version for split.\n",
      " |       str.rsplit : Standard library version for rsplit.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      The handling of the `n` keyword depends on the number of found splits:\n",
      " |      \n",
      " |      - If found splits > `n`,  make first `n` splits only\n",
      " |      - If found splits <= `n`, make all splits\n",
      " |      - If for a certain row the number of found splits < `n`,\n",
      " |        append `None` for padding up to `n` if ``expand=True``\n",
      " |      \n",
      " |      If using ``expand=True``, Series and Index callers return DataFrame and\n",
      " |      MultiIndex objects, respectively.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series([\"this is a regular sentence\",\n",
      " |      \"https://docs.python.org/3/tutorial/index.html\", np.nan])\n",
      " |      \n",
      " |      In the default setting, the string is split by whitespace.\n",
      " |      \n",
      " |      >>> s.str.split()\n",
      " |      0                   [this, is, a, regular, sentence]\n",
      " |      1    [https://docs.python.org/3/tutorial/index.html]\n",
      " |      2                                                NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Without the `n` parameter, the outputs of `rsplit` and `split`\n",
      " |      are identical.\n",
      " |      \n",
      " |      >>> s.str.rsplit()\n",
      " |      0                   [this, is, a, regular, sentence]\n",
      " |      1    [https://docs.python.org/3/tutorial/index.html]\n",
      " |      2                                                NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      The `n` parameter can be used to limit the number of splits on the\n",
      " |      delimiter. The outputs of `split` and `rsplit` are different.\n",
      " |      \n",
      " |      >>> s.str.split(n=2)\n",
      " |      0                     [this, is, a regular sentence]\n",
      " |      1    [https://docs.python.org/3/tutorial/index.html]\n",
      " |      2                                                NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.rsplit(n=2)\n",
      " |      0                     [this is a, regular, sentence]\n",
      " |      1    [https://docs.python.org/3/tutorial/index.html]\n",
      " |      2                                                NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      The `pat` parameter can be used to split by other characters.\n",
      " |      \n",
      " |      >>> s.str.split(pat = \"/\")\n",
      " |      0                         [this is a regular sentence]\n",
      " |      1    [https:, , docs.python.org, 3, tutorial, index...\n",
      " |      2                                                  NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      When using ``expand=True``, the split elements will expand out into\n",
      " |      separate columns. If NaN is present, it is propagated throughout\n",
      " |      the columns during the split.\n",
      " |      \n",
      " |      >>> s.str.split(expand=True)\n",
      " |                                                     0     1     2        3\n",
      " |      0                                           this    is     a  regular\n",
      " |      1  https://docs.python.org/3/tutorial/index.html  None  None     None\n",
      " |      2                                            NaN   NaN   NaN      NaN \n",
      " |                   4\n",
      " |      0     sentence\n",
      " |      1         None\n",
      " |      2          NaN\n",
      " |      \n",
      " |      For slightly more complex use cases like splitting the html document name\n",
      " |      from a url, a combination of parameter settings can be used.\n",
      " |      \n",
      " |      >>> s.str.rsplit(\"/\", n=1, expand=True)\n",
      " |                                          0           1\n",
      " |      0          this is a regular sentence        None\n",
      " |      1  https://docs.python.org/3/tutorial  index.html\n",
      " |      2                                 NaN         NaN\n",
      " |  \n",
      " |  rstrip(self, to_strip=None)\n",
      " |      Remove leading and trailing characters.\n",
      " |      \n",
      " |      Strip whitespaces (including newlines) or a set of specified characters\n",
      " |      from each string in the Series/Index from right side.\n",
      " |      Equivalent to :meth:`str.rstrip`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      to_strip : str or None, default None\n",
      " |          Specifying the set of characters to be removed.\n",
      " |          All combinations of this set of characters will be stripped.\n",
      " |          If None then whitespaces are removed.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series/Index of objects\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.strip : Remove leading and trailing characters in Series/Index.\n",
      " |      Series.str.lstrip : Remove leading characters in Series/Index.\n",
      " |      Series.str.rstrip : Remove trailing characters in Series/Index.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series(['1. Ant.  ', '2. Bee!\\n', '3. Cat?\\t', np.nan])\n",
      " |      >>> s\n",
      " |      0    1. Ant.\n",
      " |      1    2. Bee!\\n\n",
      " |      2    3. Cat?\\t\n",
      " |      3          NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.strip()\n",
      " |      0    1. Ant.\n",
      " |      1    2. Bee!\n",
      " |      2    3. Cat?\n",
      " |      3        NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.lstrip('123.')\n",
      " |      0    Ant.\n",
      " |      1    Bee!\\n\n",
      " |      2    Cat?\\t\n",
      " |      3       NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.rstrip('.!? \\n\\t')\n",
      " |      0    1. Ant\n",
      " |      1    2. Bee\n",
      " |      2    3. Cat\n",
      " |      3       NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.strip('123.!? \\n\\t')\n",
      " |      0    Ant\n",
      " |      1    Bee\n",
      " |      2    Cat\n",
      " |      3    NaN\n",
      " |      dtype: object\n",
      " |  \n",
      " |  slice(self, start=None, stop=None, step=None)\n",
      " |      Slice substrings from each element in the Series or Index.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      start : int, optional\n",
      " |          Start position for slice operation.\n",
      " |      stop : int, optional\n",
      " |          Stop position for slice operation.\n",
      " |      step : int, optional\n",
      " |          Step size for slice operation.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series or Index of object\n",
      " |          Series or Index from sliced substring from original string object.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.slice_replace : Replace a slice with a string.\n",
      " |      Series.str.get : Return element at position.\n",
      " |          Equivalent to `Series.str.slice(start=i, stop=i+1)` with `i`\n",
      " |          being the position.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series([\"koala\", \"fox\", \"chameleon\"])\n",
      " |      >>> s\n",
      " |      0        koala\n",
      " |      1          fox\n",
      " |      2    chameleon\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.slice(start=1)\n",
      " |      0        oala\n",
      " |      1          ox\n",
      " |      2    hameleon\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.slice(stop=2)\n",
      " |      0    ko\n",
      " |      1    fo\n",
      " |      2    ch\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.slice(step=2)\n",
      " |      0      kaa\n",
      " |      1       fx\n",
      " |      2    caeen\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.slice(start=0, stop=5, step=3)\n",
      " |      0    kl\n",
      " |      1     f\n",
      " |      2    cm\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Equivalent behaviour to:\n",
      " |      \n",
      " |      >>> s.str[0:5:3]\n",
      " |      0    kl\n",
      " |      1     f\n",
      " |      2    cm\n",
      " |      dtype: object\n",
      " |  \n",
      " |  slice_replace(self, start=None, stop=None, repl=None)\n",
      " |      Replace a positional slice of a string with another value.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      start : int, optional\n",
      " |          Left index position to use for the slice. If not specified (None),\n",
      " |          the slice is unbounded on the left, i.e. slice from the start\n",
      " |          of the string.\n",
      " |      stop : int, optional\n",
      " |          Right index position to use for the slice. If not specified (None),\n",
      " |          the slice is unbounded on the right, i.e. slice until the\n",
      " |          end of the string.\n",
      " |      repl : str, optional\n",
      " |          String for replacement. If not specified (None), the sliced region\n",
      " |          is replaced with an empty string.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      replaced : Series or Index\n",
      " |          Same type as the original object.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.slice : Just slicing without replacement.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series(['a', 'ab', 'abc', 'abdc', 'abcde'])\n",
      " |      >>> s\n",
      " |      0        a\n",
      " |      1       ab\n",
      " |      2      abc\n",
      " |      3     abdc\n",
      " |      4    abcde\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Specify just `start`, meaning replace `start` until the end of the\n",
      " |      string with `repl`.\n",
      " |      \n",
      " |      >>> s.str.slice_replace(1, repl='X')\n",
      " |      0    aX\n",
      " |      1    aX\n",
      " |      2    aX\n",
      " |      3    aX\n",
      " |      4    aX\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Specify just `stop`, meaning the start of the string to `stop` is replaced\n",
      " |      with `repl`, and the rest of the string is included.\n",
      " |      \n",
      " |      >>> s.str.slice_replace(stop=2, repl='X')\n",
      " |      0       X\n",
      " |      1       X\n",
      " |      2      Xc\n",
      " |      3     Xdc\n",
      " |      4    Xcde\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Specify `start` and `stop`, meaning the slice from `start` to `stop` is\n",
      " |      replaced with `repl`. Everything before or after `start` and `stop` is\n",
      " |      included as is.\n",
      " |      \n",
      " |      >>> s.str.slice_replace(start=1, stop=3, repl='X')\n",
      " |      0      aX\n",
      " |      1      aX\n",
      " |      2      aX\n",
      " |      3     aXc\n",
      " |      4    aXde\n",
      " |      dtype: object\n",
      " |  \n",
      " |  split(self, pat=None, n=-1, expand=False)\n",
      " |      Split strings around given separator/delimiter.\n",
      " |      \n",
      " |      Splits the string in the Series/Index from the beginning,\n",
      " |      at the specified delimiter string. Equivalent to :meth:`str.split`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      pat : str, optional\n",
      " |          String or regular expression to split on.\n",
      " |          If not specified, split on whitespace.\n",
      " |      n : int, default -1 (all)\n",
      " |          Limit number of splits in output.\n",
      " |          ``None``, 0 and -1 will be interpreted as return all splits.\n",
      " |      expand : bool, default False\n",
      " |          Expand the splitted strings into separate columns.\n",
      " |      \n",
      " |          * If ``True``, return DataFrame/MultiIndex expanding dimensionality.\n",
      " |          * If ``False``, return Series/Index, containing lists of strings.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series, Index, DataFrame or MultiIndex\n",
      " |          Type matches caller unless ``expand=True`` (see Notes).\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |       Series.str.split : Split strings around given separator/delimiter.\n",
      " |       Series.str.rsplit : Splits string around given separator/delimiter,\n",
      " |       starting from the right.\n",
      " |       Series.str.join : Join lists contained as elements in the Series/Index\n",
      " |       with passed delimiter.\n",
      " |       str.split : Standard library version for split.\n",
      " |       str.rsplit : Standard library version for rsplit.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      The handling of the `n` keyword depends on the number of found splits:\n",
      " |      \n",
      " |      - If found splits > `n`,  make first `n` splits only\n",
      " |      - If found splits <= `n`, make all splits\n",
      " |      - If for a certain row the number of found splits < `n`,\n",
      " |        append `None` for padding up to `n` if ``expand=True``\n",
      " |      \n",
      " |      If using ``expand=True``, Series and Index callers return DataFrame and\n",
      " |      MultiIndex objects, respectively.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series([\"this is a regular sentence\",\n",
      " |      \"https://docs.python.org/3/tutorial/index.html\", np.nan])\n",
      " |      \n",
      " |      In the default setting, the string is split by whitespace.\n",
      " |      \n",
      " |      >>> s.str.split()\n",
      " |      0                   [this, is, a, regular, sentence]\n",
      " |      1    [https://docs.python.org/3/tutorial/index.html]\n",
      " |      2                                                NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Without the `n` parameter, the outputs of `rsplit` and `split`\n",
      " |      are identical.\n",
      " |      \n",
      " |      >>> s.str.rsplit()\n",
      " |      0                   [this, is, a, regular, sentence]\n",
      " |      1    [https://docs.python.org/3/tutorial/index.html]\n",
      " |      2                                                NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      The `n` parameter can be used to limit the number of splits on the\n",
      " |      delimiter. The outputs of `split` and `rsplit` are different.\n",
      " |      \n",
      " |      >>> s.str.split(n=2)\n",
      " |      0                     [this, is, a regular sentence]\n",
      " |      1    [https://docs.python.org/3/tutorial/index.html]\n",
      " |      2                                                NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.rsplit(n=2)\n",
      " |      0                     [this is a, regular, sentence]\n",
      " |      1    [https://docs.python.org/3/tutorial/index.html]\n",
      " |      2                                                NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      The `pat` parameter can be used to split by other characters.\n",
      " |      \n",
      " |      >>> s.str.split(pat = \"/\")\n",
      " |      0                         [this is a regular sentence]\n",
      " |      1    [https:, , docs.python.org, 3, tutorial, index...\n",
      " |      2                                                  NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      When using ``expand=True``, the split elements will expand out into\n",
      " |      separate columns. If NaN is present, it is propagated throughout\n",
      " |      the columns during the split.\n",
      " |      \n",
      " |      >>> s.str.split(expand=True)\n",
      " |                                                     0     1     2        3\n",
      " |      0                                           this    is     a  regular\n",
      " |      1  https://docs.python.org/3/tutorial/index.html  None  None     None\n",
      " |      2                                            NaN   NaN   NaN      NaN \n",
      " |                   4\n",
      " |      0     sentence\n",
      " |      1         None\n",
      " |      2          NaN\n",
      " |      \n",
      " |      For slightly more complex use cases like splitting the html document name\n",
      " |      from a url, a combination of parameter settings can be used.\n",
      " |      \n",
      " |      >>> s.str.rsplit(\"/\", n=1, expand=True)\n",
      " |                                          0           1\n",
      " |      0          this is a regular sentence        None\n",
      " |      1  https://docs.python.org/3/tutorial  index.html\n",
      " |      2                                 NaN         NaN\n",
      " |  \n",
      " |  startswith = str_startswith(self, pat, na=nan)\n",
      " |      Test if the start of each string element matches a pattern.\n",
      " |      \n",
      " |      Equivalent to :meth:`str.startswith`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      pat : str\n",
      " |          Character sequence. Regular expressions are not accepted.\n",
      " |      na : object, default NaN\n",
      " |          Object shown if element tested is not a string.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series or Index of bool\n",
      " |          A Series of booleans indicating whether the given pattern matches\n",
      " |          the start of each string element.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      str.startswith : Python standard library string method.\n",
      " |      Series.str.endswith : Same as startswith, but tests the end of string.\n",
      " |      Series.str.contains : Tests if string element contains a pattern.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series(['bat', 'Bear', 'cat', np.nan])\n",
      " |      >>> s\n",
      " |      0     bat\n",
      " |      1    Bear\n",
      " |      2     cat\n",
      " |      3     NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.startswith('b')\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3      NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Specifying `na` to be `False` instead of `NaN`.\n",
      " |      \n",
      " |      >>> s.str.startswith('b', na=False)\n",
      " |      0     True\n",
      " |      1    False\n",
      " |      2    False\n",
      " |      3    False\n",
      " |      dtype: bool\n",
      " |  \n",
      " |  strip(self, to_strip=None)\n",
      " |      Remove leading and trailing characters.\n",
      " |      \n",
      " |      Strip whitespaces (including newlines) or a set of specified characters\n",
      " |      from each string in the Series/Index from left and right sides.\n",
      " |      Equivalent to :meth:`str.strip`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      to_strip : str or None, default None\n",
      " |          Specifying the set of characters to be removed.\n",
      " |          All combinations of this set of characters will be stripped.\n",
      " |          If None then whitespaces are removed.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series/Index of objects\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.strip : Remove leading and trailing characters in Series/Index.\n",
      " |      Series.str.lstrip : Remove leading characters in Series/Index.\n",
      " |      Series.str.rstrip : Remove trailing characters in Series/Index.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series(['1. Ant.  ', '2. Bee!\\n', '3. Cat?\\t', np.nan])\n",
      " |      >>> s\n",
      " |      0    1. Ant.\n",
      " |      1    2. Bee!\\n\n",
      " |      2    3. Cat?\\t\n",
      " |      3          NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.strip()\n",
      " |      0    1. Ant.\n",
      " |      1    2. Bee!\n",
      " |      2    3. Cat?\n",
      " |      3        NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.lstrip('123.')\n",
      " |      0    Ant.\n",
      " |      1    Bee!\\n\n",
      " |      2    Cat?\\t\n",
      " |      3       NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.rstrip('.!? \\n\\t')\n",
      " |      0    1. Ant\n",
      " |      1    2. Bee\n",
      " |      2    3. Cat\n",
      " |      3       NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.strip('123.!? \\n\\t')\n",
      " |      0    Ant\n",
      " |      1    Bee\n",
      " |      2    Cat\n",
      " |      3    NaN\n",
      " |      dtype: object\n",
      " |  \n",
      " |  swapcase lambda self\n",
      " |      Convert strings in the Series/Index to be swapcased.\n",
      " |      \n",
      " |      Equivalent to :meth:`str.swapcase`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series/Index of objects\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.lower : Converts all characters to lowercase.\n",
      " |      Series.str.upper : Converts all characters to uppercase.\n",
      " |      Series.str.title : Converts first character of each word to uppercase and\n",
      " |          remaining to lowercase.\n",
      " |      Series.str.capitalize : Converts first character to uppercase and\n",
      " |          remaining to lowercase.\n",
      " |      Series.str.swapcase : Converts uppercase to lowercase and lowercase to\n",
      " |          uppercase.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series(['lower', 'CAPITALS', 'this is a sentence', 'SwApCaSe'])\n",
      " |      >>> s\n",
      " |      0                 lower\n",
      " |      1              CAPITALS\n",
      " |      2    this is a sentence\n",
      " |      3              SwApCaSe\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.lower()\n",
      " |      0                 lower\n",
      " |      1              capitals\n",
      " |      2    this is a sentence\n",
      " |      3              swapcase\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.upper()\n",
      " |      0                 LOWER\n",
      " |      1              CAPITALS\n",
      " |      2    THIS IS A SENTENCE\n",
      " |      3              SWAPCASE\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.title()\n",
      " |      0                 Lower\n",
      " |      1              Capitals\n",
      " |      2    This Is A Sentence\n",
      " |      3              Swapcase\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.capitalize()\n",
      " |      0                 Lower\n",
      " |      1              Capitals\n",
      " |      2    This is a sentence\n",
      " |      3              Swapcase\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.swapcase()\n",
      " |      0                 LOWER\n",
      " |      1              capitals\n",
      " |      2    THIS IS A SENTENCE\n",
      " |      3              sWaPcAsE\n",
      " |      dtype: object\n",
      " |  \n",
      " |  title lambda self\n",
      " |      Convert strings in the Series/Index to titlecase.\n",
      " |      \n",
      " |      Equivalent to :meth:`str.title`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series/Index of objects\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.lower : Converts all characters to lowercase.\n",
      " |      Series.str.upper : Converts all characters to uppercase.\n",
      " |      Series.str.title : Converts first character of each word to uppercase and\n",
      " |          remaining to lowercase.\n",
      " |      Series.str.capitalize : Converts first character to uppercase and\n",
      " |          remaining to lowercase.\n",
      " |      Series.str.swapcase : Converts uppercase to lowercase and lowercase to\n",
      " |          uppercase.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series(['lower', 'CAPITALS', 'this is a sentence', 'SwApCaSe'])\n",
      " |      >>> s\n",
      " |      0                 lower\n",
      " |      1              CAPITALS\n",
      " |      2    this is a sentence\n",
      " |      3              SwApCaSe\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.lower()\n",
      " |      0                 lower\n",
      " |      1              capitals\n",
      " |      2    this is a sentence\n",
      " |      3              swapcase\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.upper()\n",
      " |      0                 LOWER\n",
      " |      1              CAPITALS\n",
      " |      2    THIS IS A SENTENCE\n",
      " |      3              SWAPCASE\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.title()\n",
      " |      0                 Lower\n",
      " |      1              Capitals\n",
      " |      2    This Is A Sentence\n",
      " |      3              Swapcase\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.capitalize()\n",
      " |      0                 Lower\n",
      " |      1              Capitals\n",
      " |      2    This is a sentence\n",
      " |      3              Swapcase\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.swapcase()\n",
      " |      0                 LOWER\n",
      " |      1              capitals\n",
      " |      2    THIS IS A SENTENCE\n",
      " |      3              sWaPcAsE\n",
      " |      dtype: object\n",
      " |  \n",
      " |  translate(self, table, deletechars=None)\n",
      " |      Map all characters in the string through the given mapping table.\n",
      " |      Equivalent to standard :meth:`str.translate`. Note that the optional\n",
      " |      argument deletechars is only valid if you are using python 2. For python 3,\n",
      " |      character deletion should be specified via the table argument.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      table : dict (python 3), str or None (python 2)\n",
      " |          In python 3, table is a mapping of Unicode ordinals to Unicode\n",
      " |          ordinals, strings, or None. Unmapped characters are left untouched.\n",
      " |          Characters mapped to None are deleted. :meth:`str.maketrans` is a\n",
      " |          helper function for making translation tables.\n",
      " |          In python 2, table is either a string of length 256 or None. If the\n",
      " |          table argument is None, no translation is applied and the operation\n",
      " |          simply removes the characters in deletechars. :func:`string.maketrans`\n",
      " |          is a helper function for making translation tables.\n",
      " |      deletechars : str, optional (python 2)\n",
      " |          A string of characters to delete. This argument is only valid\n",
      " |          in python 2.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      translated : Series/Index of objects\n",
      " |  \n",
      " |  upper lambda self\n",
      " |      Convert strings in the Series/Index to uppercase.\n",
      " |      \n",
      " |      Equivalent to :meth:`str.upper`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series/Index of objects\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.lower : Converts all characters to lowercase.\n",
      " |      Series.str.upper : Converts all characters to uppercase.\n",
      " |      Series.str.title : Converts first character of each word to uppercase and\n",
      " |          remaining to lowercase.\n",
      " |      Series.str.capitalize : Converts first character to uppercase and\n",
      " |          remaining to lowercase.\n",
      " |      Series.str.swapcase : Converts uppercase to lowercase and lowercase to\n",
      " |          uppercase.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series(['lower', 'CAPITALS', 'this is a sentence', 'SwApCaSe'])\n",
      " |      >>> s\n",
      " |      0                 lower\n",
      " |      1              CAPITALS\n",
      " |      2    this is a sentence\n",
      " |      3              SwApCaSe\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.lower()\n",
      " |      0                 lower\n",
      " |      1              capitals\n",
      " |      2    this is a sentence\n",
      " |      3              swapcase\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.upper()\n",
      " |      0                 LOWER\n",
      " |      1              CAPITALS\n",
      " |      2    THIS IS A SENTENCE\n",
      " |      3              SWAPCASE\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.title()\n",
      " |      0                 Lower\n",
      " |      1              Capitals\n",
      " |      2    This Is A Sentence\n",
      " |      3              Swapcase\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.capitalize()\n",
      " |      0                 Lower\n",
      " |      1              Capitals\n",
      " |      2    This is a sentence\n",
      " |      3              Swapcase\n",
      " |      dtype: object\n",
      " |      \n",
      " |      >>> s.str.swapcase()\n",
      " |      0                 LOWER\n",
      " |      1              capitals\n",
      " |      2    THIS IS A SENTENCE\n",
      " |      3              sWaPcAsE\n",
      " |      dtype: object\n",
      " |  \n",
      " |  wrap(self, width, **kwargs)\n",
      " |      Wrap long strings in the Series/Index to be formatted in\n",
      " |      paragraphs with length less than a given width.\n",
      " |      \n",
      " |      This method has the same keyword parameters and defaults as\n",
      " |      :class:`textwrap.TextWrapper`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      width : int\n",
      " |          Maximum line-width\n",
      " |      expand_tabs : bool, optional\n",
      " |          If true, tab characters will be expanded to spaces (default: True)\n",
      " |      replace_whitespace : bool, optional\n",
      " |          If true, each whitespace character (as defined by string.whitespace)\n",
      " |          remaining after tab expansion will be replaced by a single space\n",
      " |          (default: True)\n",
      " |      drop_whitespace : bool, optional\n",
      " |          If true, whitespace that, after wrapping, happens to end up at the\n",
      " |          beginning or end of a line is dropped (default: True)\n",
      " |      break_long_words : bool, optional\n",
      " |          If true, then words longer than width will be broken in order to ensure\n",
      " |          that no lines are longer than width. If it is false, long words will\n",
      " |          not be broken, and some lines may be longer than width. (default: True)\n",
      " |      break_on_hyphens : bool, optional\n",
      " |          If true, wrapping will occur preferably on whitespace and right after\n",
      " |          hyphens in compound words, as it is customary in English. If false,\n",
      " |          only whitespaces will be considered as potentially good places for line\n",
      " |          breaks, but you need to set break_long_words to false if you want truly\n",
      " |          insecable words. (default: True)\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      wrapped : Series/Index of objects\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      Internally, this method uses a :class:`textwrap.TextWrapper` instance with\n",
      " |      default settings. To achieve behavior matching R's stringr library str_wrap\n",
      " |      function, use the arguments:\n",
      " |      \n",
      " |      - expand_tabs = False\n",
      " |      - replace_whitespace = True\n",
      " |      - drop_whitespace = True\n",
      " |      - break_long_words = False\n",
      " |      - break_on_hyphens = False\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      \n",
      " |      >>> s = pd.Series(['line to be wrapped', 'another line to be wrapped'])\n",
      " |      >>> s.str.wrap(12)\n",
      " |      0             line to be\\nwrapped\n",
      " |      1    another line\\nto be\\nwrapped\n",
      " |  \n",
      " |  zfill(self, width)\n",
      " |      Pad strings in the Series/Index by prepending '0' characters.\n",
      " |      \n",
      " |      Strings in the Series/Index are padded with '0' characters on the\n",
      " |      left of the string to reach a total string length  `width`. Strings\n",
      " |      in the Series/Index with length greater or equal to `width` are\n",
      " |      unchanged.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      width : int\n",
      " |          Minimum length of resulting string; strings with length less\n",
      " |          than `width` be prepended with '0' characters.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Series/Index of objects\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.str.rjust : Fills the left side of strings with an arbitrary\n",
      " |          character.\n",
      " |      Series.str.ljust : Fills the right side of strings with an arbitrary\n",
      " |          character.\n",
      " |      Series.str.pad : Fills the specified sides of strings with an arbitrary\n",
      " |          character.\n",
      " |      Series.str.center : Fills boths sides of strings with an arbitrary\n",
      " |          character.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      Differs from :meth:`str.zfill` which has special handling\n",
      " |      for '+'/'-' in the string.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> s = pd.Series(['-1', '1', '1000', 10, np.nan])\n",
      " |      >>> s\n",
      " |      0      -1\n",
      " |      1       1\n",
      " |      2    1000\n",
      " |      3      10\n",
      " |      4     NaN\n",
      " |      dtype: object\n",
      " |      \n",
      " |      Note that ``10`` and ``NaN`` are not strings, therefore they are\n",
      " |      converted to ``NaN``. The minus sign in ``'-1'`` is treated as a\n",
      " |      regular character and the zero is added to the left of it\n",
      " |      (:meth:`str.zfill` would have moved it to the left). ``1000``\n",
      " |      remains unchanged as it is longer than `width`.\n",
      " |      \n",
      " |      >>> s.str.zfill(3)\n",
      " |      0     0-1\n",
      " |      1     001\n",
      " |      2    1000\n",
      " |      3     NaN\n",
      " |      4     NaN\n",
      " |      dtype: object\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from pandas.core.base.NoNewAttributesMixin:\n",
      " |  \n",
      " |  __setattr__(self, key, value)\n",
      " |      Implement setattr(self, name, value).\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from pandas.core.base.NoNewAttributesMixin:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(species_list['display_name'].str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "alpha = species_list.iloc[[11]]['name']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
