

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Usage &mdash; MerryCRISPR 2.1 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="merrycrispr" href="merrycrispr.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> MerryCRISPR
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#new-species">new-species</a></li>
<li class="toctree-l2"><a class="reference internal" href="#prep-sequences">prep-sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="#create-library">create-library</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#docker">Docker</a></li>
<li class="toctree-l1"><a class="reference internal" href="merrycrispr.html">merrycrispr</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MerryCRISPR</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Usage</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/usage.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h1>
<p>MerryCRISPR has three subfunctions: <a class="reference internal" href="#new-species">new-species</a>, <a class="reference internal" href="#prep-sequences">prep-sequences</a>, and <a class="reference internal" href="#create-library">create-library</a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Usage</span><span class="p">:</span> <span class="n">merrycrispr</span> <span class="p">[</span><span class="n">OPTIONS</span><span class="p">]</span> <span class="n">COMMAND</span> <span class="p">[</span><span class="n">ARGS</span><span class="p">]</span><span class="o">...</span>

<span class="n">Options</span><span class="p">:</span>
  <span class="o">--</span><span class="n">help</span>  <span class="n">Show</span> <span class="n">this</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span><span class="o">.</span>

<span class="n">Commands</span><span class="p">:</span>
  <span class="n">create</span><span class="o">-</span><span class="n">library</span>  <span class="n">Build</span> <span class="n">a</span> <span class="n">CRISPR</span> <span class="n">library</span>
  <span class="n">new</span><span class="o">-</span><span class="n">species</span>     <span class="n">Import</span> <span class="n">a</span> <span class="n">new</span> <span class="n">species</span> <span class="kn">from</span> <span class="nn">Ensembl.</span>
  <span class="n">prep</span><span class="o">-</span><span class="n">sequences</span>  <span class="n">Generate</span> <span class="n">target</span> <span class="n">sequences</span> <span class="n">to</span> <span class="n">search</span> <span class="k">for</span> <span class="n">spacers</span>
</pre></div>
</div>
<div class="section" id="new-species">
<h2>new-species<a class="headerlink" href="#new-species" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Usage</span><span class="p">:</span> <span class="n">merrycrispr</span> <span class="n">new</span><span class="o">-</span><span class="n">species</span> <span class="p">[</span><span class="n">OPTIONS</span><span class="p">]</span>

  <span class="n">Import</span> <span class="n">a</span> <span class="n">new</span> <span class="n">species</span> <span class="kn">from</span> <span class="nn">Ensembl.</span>  <span class="n">Optionally</span><span class="p">,</span> <span class="n">build</span> <span class="n">a</span> <span class="n">Bowtie</span> <span class="n">index</span> <span class="k">for</span>
  <span class="n">the</span> <span class="n">files</span><span class="o">.</span>

<span class="n">Options</span><span class="p">:</span>
  <span class="o">-</span><span class="n">s</span><span class="p">,</span> <span class="o">--</span><span class="n">species_value</span> <span class="n">TEXT</span>      <span class="n">Species</span> <span class="n">value</span> <span class="n">to</span> <span class="n">retrieve</span> <span class="n">sequences</span> <span class="k">for</span><span class="o">.</span>  <span class="n">Must</span>
                                <span class="n">be</span> <span class="n">a</span> <span class="n">value</span> <span class="k">for</span> <span class="n">the</span> <span class="n">attribute</span> <span class="n">chosen</span><span class="p">,</span> <span class="n">i</span><span class="o">.</span><span class="n">e</span><span class="o">.</span>
                                <span class="s1">&#39;canis_familiaris&#39;</span> <span class="k">if</span> <span class="n">using</span> <span class="n">the</span> <span class="s1">&#39;name&#39;</span>
                                <span class="n">attribute</span>
  <span class="o">-</span><span class="n">a</span><span class="p">,</span> <span class="o">--</span><span class="n">species_attribute</span> <span class="n">TEXT</span>  <span class="n">Species</span> <span class="n">attribute</span> <span class="n">to</span> <span class="n">search</span> <span class="n">when</span> <span class="n">retrieving</span>
                                <span class="n">sequences</span><span class="o">.</span> <span class="n">Acceptable</span> <span class="n">values</span> <span class="n">include</span>
                                <span class="s1">&#39;taxon_id&#39;</span><span class="p">,</span> <span class="s1">&#39;accession&#39;</span><span class="p">,</span> <span class="s1">&#39;aliases&#39;</span><span class="p">,</span>
                                <span class="s1">&#39;division&#39;</span><span class="p">,</span> <span class="s1">&#39;groups&#39;</span><span class="p">,</span> <span class="s1">&#39;release&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">,</span>
                                <span class="s1">&#39;strain&#39;</span><span class="p">,</span> <span class="s1">&#39;strain_collection&#39;</span><span class="p">,</span> <span class="s1">&#39;display_name&#39;</span><span class="p">,</span>
                                <span class="s1">&#39;assembly&#39;</span><span class="p">,</span> <span class="ow">and</span> <span class="s1">&#39;common_name&#39;</span>
  <span class="o">-</span><span class="n">d</span><span class="p">,</span> <span class="o">--</span><span class="n">dest</span> <span class="n">TEXT</span>               <span class="n">Location</span> <span class="ow">in</span> <span class="n">which</span> <span class="n">to</span> <span class="n">place</span> <span class="n">downloaded</span> <span class="n">files</span>
                                <span class="ow">and</span><span class="p">,</span> <span class="k">if</span> <span class="n">selected</span><span class="p">,</span> <span class="n">new</span> <span class="n">Bowtie</span> <span class="n">index</span><span class="o">.</span>
  <span class="o">-</span><span class="n">b</span><span class="p">,</span> <span class="o">--</span><span class="n">build_bowtie</span>            <span class="n">Build</span> <span class="n">a</span> <span class="n">bowtie</span> <span class="n">index</span> <span class="k">for</span> <span class="n">the</span> <span class="n">retrieved</span> <span class="n">species</span>
  <span class="o">-</span><span class="n">w</span><span class="p">,</span> <span class="o">--</span><span class="n">show_available_builds</span>   <span class="n">Display</span> <span class="n">a</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">the</span> <span class="n">genome</span> <span class="n">builds</span> <span class="n">available</span>
                                <span class="kn">from</span> <span class="nn">Ensembl</span>
  <span class="o">--</span><span class="n">help</span>                        <span class="n">Show</span> <span class="n">this</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span><span class="o">.</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">new-species</span></code> subcommand is used to import sequences from Ensembl. Currently, <code class="docutils literal notranslate"><span class="pre">new-species</span></code> is only able to
retrieve the most current release ofa genome build (which is version 97 as of this writting). To import files, you will
need to know the value for one of the attributes for that genome’s release.  In plain English, it is probably best if
you use <code class="docutils literal notranslate"><span class="pre">accession</span></code> or <code class="docutils literal notranslate"><span class="pre">assembly</span></code> for the <code class="docutils literal notranslate"><span class="pre">--species_attribute</span></code> argument, look up the value on <a class="reference external" href="https://uswest.ensembl.org/info/about/species.html">Ensembl</a>, and use that for the <code class="docutils literal notranslate"><span class="pre">--species_value</span></code> argument.
Alternatively, you can run <code class="docutils literal notranslate"><span class="pre">new-species</span></code> with just the <code class="docutils literal notranslate"><span class="pre">--show_available</span></code> argument to see possible values. While you can
use the <code class="docutils literal notranslate"><span class="pre">common_name</span></code> (i.e “dog”) or species <code class="docutils literal notranslate"><span class="pre">name</span></code> (i.e. “Canis familiaris”), those are non-unique for some values; if
that is the case, <code class="docutils literal notranslate"><span class="pre">new-species</span></code> will exit and ask you to make a more precise request.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">--dest</span></code> argument is optional.  If you do not provide it, the files will be placed in a subdirectory of the current
working directory called “/mc_resources”.</p>
<p>Because it can take a up to several hours to build the index, by default <code class="docutils literal notranslate"><span class="pre">new-species</span></code> does not build one on species
import.  Passing the <code class="docutils literal notranslate"><span class="pre">--build_bowtie</span></code> flag when you run the command will tell it to build the index.</p>
<p>If there are already matching files in the <code class="docutils literal notranslate"><span class="pre">dest</span></code> directory, <code class="docutils literal notranslate"><span class="pre">new-species</span></code> will check to see if the file sizes match
those on Ensembl and, if they do, will not redownload the files.  This allows you to run the program some time after
downloading to build the index.</p>
<p>An example of using <code class="docutils literal notranslate"><span class="pre">new-species</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">merrycrispr</span> <span class="n">new</span><span class="o">-</span><span class="n">species</span> <span class="o">--</span><span class="n">species_value</span> <span class="n">homo_sapiens</span> \
    <span class="o">--</span><span class="n">species_attribute</span> <span class="n">name</span> \
    <span class="o">--</span><span class="n">dest</span> <span class="o">~/</span><span class="n">workdir</span><span class="o">/</span><span class="n">mc_human_files</span> \
    <span class="o">--</span><span class="n">build_bowtie</span>
</pre></div>
</div>
<p>Note that this is all provided as a convenience - you can download the sequence and annotation files and/or build
the Bowtie index manually.</p>
</div>
<div class="section" id="prep-sequences">
<h2>prep-sequences<a class="headerlink" href="#prep-sequences" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Usage</span><span class="p">:</span> <span class="n">merrycrispr</span> <span class="n">prep</span><span class="o">-</span><span class="n">sequences</span> <span class="p">[</span><span class="n">OPTIONS</span><span class="p">]</span>

  <span class="n">Generate</span> <span class="n">target</span> <span class="n">sequences</span> <span class="n">to</span> <span class="n">search</span> <span class="k">for</span> <span class="n">spacers</span>

  <span class="n">Generally</span><span class="p">,</span> <span class="n">a</span> <span class="n">utility</span> <span class="k">for</span> <span class="n">extracting</span> <span class="n">sequences</span> <span class="kn">from</span> <span class="nn">a</span> <span class="n">FASTA</span> <span class="n">file</span> <span class="k">for</span> <span class="n">a</span>
  <span class="n">given</span> <span class="n">GTF</span> <span class="n">annotation</span><span class="o">.</span>

<span class="n">Options</span><span class="p">:</span>
  <span class="o">-</span><span class="n">l</span><span class="p">,</span> <span class="o">--</span><span class="n">library_type</span> <span class="n">TEXT</span>  <span class="n">target</span> <span class="n">library</span> <span class="nb">type</span><span class="o">.</span> <span class="n">Accepted</span> <span class="n">values</span> <span class="n">are</span>
                           <span class="s1">&#39;knockout&#39;</span><span class="p">,</span> <span class="s1">&#39;repressor/activator&#39;</span><span class="p">,</span> <span class="s1">&#39;excision&#39;</span><span class="p">,</span> <span class="ow">or</span>
                           <span class="s1">&#39;Cas13&#39;</span>
  <span class="o">-</span><span class="n">g</span><span class="p">,</span> <span class="o">--</span><span class="n">gtf</span> <span class="n">TEXT</span>           <span class="nb">input</span> <span class="n">GTF</span><span class="o">/</span><span class="n">GFF</span> <span class="n">file</span>
  <span class="o">-</span><span class="n">f</span><span class="p">,</span> <span class="o">--</span><span class="n">fasta</span> <span class="n">TEXT</span>         <span class="n">FASTA</span> <span class="n">sequence</span> <span class="n">file</span>
  <span class="o">-</span><span class="n">o</span><span class="p">,</span> <span class="o">--</span><span class="n">output</span> <span class="n">TEXT</span>        <span class="n">output</span> <span class="n">file</span>
  <span class="o">-</span><span class="n">t</span><span class="p">,</span> <span class="o">--</span><span class="n">feature_type</span> <span class="n">TEXT</span>  <span class="n">feature</span> <span class="nb">type</span>
  <span class="o">-</span><span class="n">n</span><span class="p">,</span> <span class="o">--</span><span class="n">gene_name</span> <span class="n">TEXT</span>     <span class="n">gene</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="n">to</span> <span class="n">extract</span>
  <span class="o">--</span><span class="n">bound</span> <span class="n">INTEGER</span>          <span class="n">Retrieve</span> <span class="n">a</span> <span class="n">given</span> <span class="n">number</span> <span class="n">of</span> <span class="n">bases</span> <span class="n">on</span> <span class="n">either</span> <span class="n">side</span> <span class="n">of</span>
                           <span class="n">the</span> <span class="n">feature</span> <span class="n">instead</span> <span class="n">of</span> <span class="n">the</span> <span class="n">sequence</span> <span class="n">corresponding</span>
                           <span class="n">to</span> <span class="n">a</span> <span class="n">feature</span>
  <span class="o">--</span><span class="n">show_features</span>          <span class="n">Scan</span> <span class="n">an</span> <span class="n">annotation</span> <span class="n">file</span> <span class="n">to</span> <span class="n">identify</span> <span class="n">the</span> <span class="n">features</span> <span class="n">present</span>
  <span class="o">--</span><span class="n">show_genes</span>             <span class="n">Scan</span> <span class="n">an</span> <span class="n">annotation</span> <span class="n">file</span> <span class="n">to</span> <span class="n">identify</span> <span class="n">the</span> <span class="n">genes</span> <span class="n">present</span>
  <span class="o">--</span><span class="n">show_geneids</span>           <span class="n">Scan</span> <span class="n">an</span> <span class="n">annotation</span> <span class="n">file</span> <span class="n">to</span> <span class="n">identify</span> <span class="n">the</span> <span class="n">geneIDs</span> <span class="n">present</span>
  <span class="o">--</span><span class="n">help</span>                   <span class="n">Show</span> <span class="n">this</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span><span class="o">.</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">prep-sequences</span></code> subcommand is used to match sequences in a FASTA file to those described in a GTF or GFF annotation
file for downstream use by the <code class="docutils literal notranslate"><span class="pre">create-library</span></code> command. <code class="docutils literal notranslate"><span class="pre">prep-sequences</span></code> can, for example, be used to select one or
more particular genes, a certain biotype, or regions surrounding a gene.</p>
<p>Usage is fairly straightforward: give <code class="docutils literal notranslate"><span class="pre">prep-sequences</span></code> a FASTA and GTF, tell it the library_type, and give it the name
of an output file to write to.  The <code class="docutils literal notranslate"><span class="pre">--library_type</span></code> options are setup to produce spacers that are appropriate for the
desired function - the “knockout” option extracts exonic sequences; the “repressor/activator” option extracts sequences
-/+ 100 nucleotides surrounding the TSS (or, more precisely, the start of exon 1); the excision option excises a 100
nucleotide block upstream and 100 nucleotides downstream of a feature (“gene”, by default), though that can be changed
by passing a number to the <code class="docutils literal notranslate"><span class="pre">--bound</span></code> argument; and “Cas13” returns sequences matching the processed mRNA. Note that the
<code class="docutils literal notranslate"><span class="pre">--gene_name</span></code> argument can take multiple values with just a space between the gene names.</p>
<p>For example, to extract all of the exons from the most current human assembly for use in a knockout library:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">merrycrispr</span> <span class="n">prep</span><span class="o">-</span><span class="n">sequences</span> <span class="o">--</span><span class="n">library_type</span> <span class="n">knockout</span> \
    <span class="o">--</span><span class="n">gtf</span> <span class="o">~/</span><span class="n">workdir</span><span class="o">/</span><span class="n">mc_human_files</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="mf">97.</span><span class="n">gtf</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">fasta</span> <span class="o">~/</span><span class="n">workdir</span><span class="o">/</span><span class="n">mc_human_files</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="o">.</span><span class="n">dna</span><span class="o">.</span><span class="n">primary_assembly</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">output</span> <span class="o">~/</span><span class="n">workdir</span><span class="o">/</span><span class="n">mc_human_files</span><span class="o">/</span><span class="n">human_exons</span><span class="o">.</span><span class="n">fa</span>
</pre></div>
</div>
</div>
<div class="section" id="create-library">
<h2>create-library<a class="headerlink" href="#create-library" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Usage: merrycrispr create-library [OPTIONS]

  Build a CRISPR library

Options:
  -i, --input TEXT                Input FASTA file containing sequences to
                                  target.
  -p, --output TEXT               Name of file to write library to (in CSV
                                  format).
  -n, --nuclease TEXT             Cas nuclease to design for.  Current options
                                  include SpCas9, Cpf1, Cas13a, Csc2, SaCas9
  -r, --reference TEXT            Path to the directory containing the
                                  appropriate Bowtie index.
  --largeindex BOOLEAN
  --rule_set TEXT                 On-target score rule set to use. Current
                                  options include &#39;1&#39;, &#39;2&#39;, and &#39;Azimuth&#39;
  -on, --ontarget_score_threshold INTEGER
                                  Spacers with an on-target score below this
                                  will be ignored.
  -off, --offtarget_score_threshold INTEGER
                                  Spacers with an off-target score below this
                                  will be ignored.
  --offtarget_count_threshold INTEGER
                                  Spacers with more than this many off-targets
                                  will be ignored.
  --spacers_per_feature INTEGER   Number of spacers to find for each feature.
  --paired BOOLEAN                Should spacers be designed to work as pairs
                                  (e.g. for excision)?
  --number_upstream_spacers INTEGER
                                  If designing paired spacers, number of
                                  spacers to design that target upstream of
                                  the feature.
  --number_downstream_spacers INTEGER
                                  If designing paired spacers, number of
                                  spacers to design that target downstream of
                                  the feature.
  --min_paired_distance INTEGER   If designing paired spacers, minimum space
                                  required between the up- and downstream
                                  spacers.
  -c, --cores INTEGER             Number of processors to use. By default,
                                  will use all available.
  --help                          Show this message and exit.
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">create-library</span></code> subcommand takes the sequences from <code class="docutils literal notranslate"><span class="pre">prep-sequences</span></code> and returns a series of protospacers according
to the given specifications. At current, <code class="docutils literal notranslate"><span class="pre">create-library</span></code> is able to find spacers for SpCas9, Cpf1, Cas13a, Csc2,
and SaCas9, though the ability to add any nuclease is forthcoming.</p>
<p>At current, there are only on-target and off-target scoring functions for SpCas9-based spacers, mainly because I am
unaware of algorithms for the other nucleases.  For Cas9, the on-target scoring is based on the rules from the papers
<a class="reference external" href="https://doi.org/10.1038/nbt.3026">Donch et al. Nat Biotechnol. 2014</a> (Rule Set 1) or
<a class="reference external" href="https://doi.org/10.1038/nbt.3437">Donch et al. Nat Biotech 2006</a> (Rule Set 2), with the latter as implemented by my
heavily updated version of <a class="reference external" href="https://github.com/milescsmith/Azimuth/">Azimuth</a>. Unless you have a good reason to not do
so, Rules Set 2 should be used. Off-target scoring is a simple algorithm from
<a class="reference external" href="https://doi.org/10.1038/nbt.2647">Hsu el al. Nat Biotechnol. 2014</a>.  Scores range from 0 to 100, with higher numbers
being more desirable.</p>
<p>Following from the example in <a class="reference internal" href="#prep-sequences">prep-sequences</a> above, the following would find 6 spacers per gene for Cas9 with on- and
off-target scores above 50 using Rule Set 2:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">merrycrispr</span> <span class="n">create</span><span class="o">-</span><span class="n">library</span> <span class="o">--</span><span class="nb">input</span> <span class="o">~/</span><span class="n">workdir</span><span class="o">/</span><span class="n">mc_human_files</span><span class="o">/</span><span class="n">human_exons</span><span class="o">.</span><span class="n">fa</span> \
    <span class="o">--</span><span class="n">output</span> <span class="o">~/</span><span class="n">workdir</span><span class="o">/</span><span class="n">mc_human_files</span><span class="o">/</span><span class="n">human_knockout</span><span class="o">.</span><span class="n">csv</span> \
    <span class="o">--</span><span class="n">nuclease</span> <span class="n">SpCas9</span> \
    <span class="o">--</span><span class="n">reference</span> <span class="o">~/</span><span class="n">workdir</span><span class="o">/</span><span class="n">mc_human_files</span><span class="o">/</span><span class="n">Homo_sapiens</span> \
    <span class="o">--</span><span class="n">rule_set</span> <span class="mi">2</span> \
    <span class="o">--</span><span class="n">ontarget_score_threshold</span> <span class="mi">50</span>
    <span class="o">--</span><span class="n">offtarget_score_threshold</span> <span class="mi">50</span>
    <span class="o">--</span><span class="n">spacers_per_feature</span> <span class="mi">6</span>
</pre></div>
</div>
<p>Of the options, probably the <code class="docutils literal notranslate"><span class="pre">--reference</span></code> is the easiest to mistake how to use.  Bowtie creates a series of 4-7 files
with the suffix <code class="docutils literal notranslate"><span class="pre">.ewbt</span></code> (or, it the genome is exceptionally large, <code class="docutils literal notranslate"><span class="pre">.ewbtl</span></code>, thus the <code class="docutils literal notranslate"><span class="pre">--largeindex</span></code> flag). So, for
instance, when <code class="docutils literal notranslate"><span class="pre">new-species</span></code> above created the index, the files produced were:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>.
├── Homo_sapiens.1.ebwt
├── Homo_sapiens.2.ebwt
├── Homo_sapiens.3.ebwt
├── Homo_sapiens.4.ebwt
├── Homo_sapiens.rev.1.ebwt
└── Homo_sapiens.rev.2.ebwt
</pre></div>
</div>
<p>When specifying the index to use, Bowtie needs to know the common prefix - that is, everything up to the <code class="docutils literal notranslate"><span class="pre">.1.ebwt</span></code>.</p>
</div>
</div>
<div class="section" id="docker">
<h1>Docker<a class="headerlink" href="#docker" title="Permalink to this headline">¶</a></h1>
<p>While the method of installation via <a class="reference internal" href="installation.html#docker"><span class="std std-ref">Docker</span></a> is simpler, its self-contained nature means that
by default a Docker container cannot see files outside the container.  To make it work, we need to run the container
with a “bind mount” to an existing directory outside the container - this serves as a link between a real folder and
one inside the container.</p>
<p>To run a container with such a mount:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">docker</span> <span class="n">run</span> <span class="o">-</span><span class="n">v</span> <span class="n">OUTSIDE_PATH</span><span class="p">:</span><span class="n">INSIDE_PATH</span> <span class="n">IMAGE</span> <span class="n">COMMAND</span> <span class="p">[</span><span class="n">COMMAND</span> <span class="n">OPTIONS</span><span class="p">]</span>
</pre></div>
</div>
<p>So to run the <a class="reference internal" href="#new-species">new-species</a> command from above:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>docker run -v &quot;$(pwd)&quot;/:$HOME/workspace merrycrispr merrycrispr new-species \
    --species_value homo_sapiens \
    --species_attribute name \
    --dest $HOME/workspace/
</pre></div>
</div>
<p>The bind mount flag <code class="docutils literal notranslate"><span class="pre">-v</span> <span class="pre">&quot;$(pwd)&quot;/:$HOME/workspace</span></code> flag above instructs Docker to set up a link from the <em>p*resent
*w*orking *d*irectory to a directory named “workspace”in the user’s home folder inside the docker. Note that later in
the command, the location passed to the `–dest` argument is *inside</em> the container - this is fine as the link is
two-way, meaning that any files written to “$HOME/workspace” are also written to your local $PWD/workspace directory.</p>
<p>Also note that the double <code class="docutils literal notranslate"><span class="pre">merrycrispr</span> <span class="pre">merrycrispr</span></code> is not a typo - the first is to create a container from the
merrycrispr image, the second is to run the merrycrispr command inside the new container.</p>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="merrycrispr.html" class="btn btn-neutral float-right" title="merrycrispr" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Miles Smith.

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>